"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[8234],{3905:(e,t,n)=>{n.d(t,{Zo:()=>m,kt:()=>h});var a=n(7294);function i(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function s(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var a=Object.getOwnPropertySymbols(e);t&&(a=a.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,a)}return n}function o(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?s(Object(n),!0).forEach((function(t){i(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):s(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function r(e,t){if(null==e)return{};var n,a,i=function(e,t){if(null==e)return{};var n,a,i={},s=Object.keys(e);for(a=0;a<s.length;a++)n=s[a],t.indexOf(n)>=0||(i[n]=e[n]);return i}(e,t);if(Object.getOwnPropertySymbols){var s=Object.getOwnPropertySymbols(e);for(a=0;a<s.length;a++)n=s[a],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(i[n]=e[n])}return i}var l=a.createContext({}),p=function(e){var t=a.useContext(l),n=t;return e&&(n="function"==typeof e?e(t):o(o({},t),e)),n},m=function(e){var t=p(e.components);return a.createElement(l.Provider,{value:t},e.children)},d={inlineCode:"code",wrapper:function(e){var t=e.children;return a.createElement(a.Fragment,{},t)}},c=a.forwardRef((function(e,t){var n=e.components,i=e.mdxType,s=e.originalType,l=e.parentName,m=r(e,["components","mdxType","originalType","parentName"]),c=p(n),h=i,k=c["".concat(l,".").concat(h)]||c[h]||d[h]||s;return n?a.createElement(k,o(o({ref:t},m),{},{components:n})):a.createElement(k,o({ref:t},m))}));function h(e,t){var n=arguments,i=t&&t.mdxType;if("string"==typeof e||i){var s=n.length,o=new Array(s);o[0]=c;var r={};for(var l in t)hasOwnProperty.call(t,l)&&(r[l]=t[l]);r.originalType=e,r.mdxType="string"==typeof e?e:i,o[1]=r;for(var p=2;p<s;p++)o[p]=n[p];return a.createElement.apply(null,o)}return a.createElement.apply(null,n)}c.displayName="MDXCreateElement"},5735:(e,t,n)=>{n.r(t),n.d(t,{assets:()=>l,contentTitle:()=>o,default:()=>d,frontMatter:()=>s,metadata:()=>r,toc:()=>p});var a=n(7462),i=(n(7294),n(3905));const s={sidebar_position:1},o="Gene annotations tutorial",r={unversionedId:"bioinformatics/exploring_gene_annotations_in_bash/exploring_genes_in_bash",id:"bioinformatics/exploring_gene_annotations_in_bash/exploring_genes_in_bash",title:"Gene annotations tutorial",description:"Author: Gavin Band",source:"@site/docs/bioinformatics/exploring_gene_annotations_in_bash/exploring_genes_in_bash.md",sourceDirName:"bioinformatics/exploring_gene_annotations_in_bash",slug:"/bioinformatics/exploring_gene_annotations_in_bash/exploring_genes_in_bash",permalink:"/whg-training-resources/bioinformatics/exploring_gene_annotations_in_bash/exploring_genes_in_bash",draft:!1,editUrl:"https://github.com/whg-training/whg-training-resources/edit/main/docs/bioinformatics/exploring_gene_annotations_in_bash/exploring_genes_in_bash.md",tags:[],version:"current",sidebarPosition:1,frontMatter:{sidebar_position:1},sidebar:"tutorialSidebar",previous:{title:"Exploring gene annotations on the command-line",permalink:"/whg-training-resources/bioinformatics/exploring_gene_annotations_in_bash/"},next:{title:"Quick tips and tricks",permalink:"/whg-training-resources/bioinformatics/tips_and_tricks/"}},l={},p=[{value:"Getting started",id:"getting-started",level:2},{value:"Exploring the annotations",id:"exploring-the-annotations",level:2},{value:"Decompressing the file",id:"decompressing-the-file",level:3},{value:"Viewing the file",id:"viewing-the-file",level:3},{value:"Counting genes",id:"counting-genes",level:3},{value:"Investigating specific genes.",id:"investigating-specific-genes",level:3},{value:"Finding transcripts",id:"finding-transcripts",level:3},{value:"Finding exons",id:"finding-exons",level:3},{value:"Browsing the genome",id:"browsing-the-genome",level:2},{value:"Challenge questions",id:"challenge-questions",level:2}],m={toc:p};function d(e){let{components:t,...s}=e;return(0,i.kt)("wrapper",(0,a.Z)({},m,s,{components:t,mdxType:"MDXLayout"}),(0,i.kt)("h1",{id:"gene-annotations-tutorial"},"Gene annotations tutorial"),(0,i.kt)("p",null,(0,i.kt)("strong",{parentName:"p"},"Author"),": Gavin Band"),(0,i.kt)("p",null,"Welcome! In this tutorial we will show how to use the UNIX command line to explore the human gene annotations.\nThis tutorial has two main objectives:"),(0,i.kt)("ul",null,(0,i.kt)("li",{parentName:"ul"},"To demonstrate some useful ways of working in the UNIX command line;"),(0,i.kt)("li",{parentName:"ul"},"And to get you to start understanding gene annotation data - that is, the core data\nfiles which represent our knowledge of human genes.")),(0,i.kt)("p",null,"The information in these files includes such things as where genes are in the genome, how they are transcribed,\nwhich bits gets turned into proteins and so on.  They are pretty important files!"),(0,i.kt)("p",null,"This is not a full tutorial on command-line processing, but here is a table of some of the UNIX commands we'll\nuse. If you're not familiar with these, don't worry: there are a lot of commands and it takes a while to learn\nthem. Try the example commands now in your terminal:"),(0,i.kt)("table",null,(0,i.kt)("thead",{parentName:"table"},(0,i.kt)("tr",{parentName:"thead"},(0,i.kt)("th",{parentName:"tr",align:null},"Command"),(0,i.kt)("th",{parentName:"tr",align:null},"What it does"),(0,i.kt)("th",{parentName:"tr",align:null},"Example"))),(0,i.kt)("tbody",{parentName:"table"},(0,i.kt)("tr",{parentName:"tbody"},(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"ls")),(0,i.kt)("td",{parentName:"tr",align:null},"Lists files in a directory"),(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"ls"))),(0,i.kt)("tr",{parentName:"tbody"},(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"mkdir")),(0,i.kt)("td",{parentName:"tr",align:null},"Make a new directory"),(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"mkdir genes_tutorial"))),(0,i.kt)("tr",{parentName:"tbody"},(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"cd")),(0,i.kt)("td",{parentName:"tr",align:null},"Changes the current directory"),(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"cd genes_tutorial"))),(0,i.kt)("tr",{parentName:"tbody"},(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"echo")),(0,i.kt)("td",{parentName:"tr",align:null},"Print some text (that cna be redirected to a file.)"),(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},'echo "Hello all\\ngenes" > file.txt'))),(0,i.kt)("tr",{parentName:"tbody"},(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"cat")),(0,i.kt)("td",{parentName:"tr",align:null},"Print the output of one or more files."),(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"cat file.txt"))),(0,i.kt)("tr",{parentName:"tbody"},(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"less")),(0,i.kt)("td",{parentName:"tr",align:null},"Interactively explore a file (press ",(0,i.kt)("inlineCode",{parentName:"td"},"q")," to quit)"),(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"less file.txt"))),(0,i.kt)("tr",{parentName:"tbody"},(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"cut")),(0,i.kt)("td",{parentName:"tr",align:null},"Extract specific columns from a file"),(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"cut -d' ' -f1 file.txt"),".")),(0,i.kt)("tr",{parentName:"tbody"},(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"grep")),(0,i.kt)("td",{parentName:"tr",align:null},"Search for a string (or ",(0,i.kt)("em",{parentName:"td"},"regulare expression")),(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},'grep "Hello" file.txt'))),(0,i.kt)("tr",{parentName:"tbody"},(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"awk")),(0,i.kt)("td",{parentName:"tr",align:null},"General-purpose tool"),(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},'awk \'$1 == "Hello" file.txt'))),(0,i.kt)("tr",{parentName:"tbody"},(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"sort")),(0,i.kt)("td",{parentName:"tr",align:null},"Sort rows alphabetically"),(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"sort file.txt"))),(0,i.kt)("tr",{parentName:"tbody"},(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"uniq")),(0,i.kt)("td",{parentName:"tr",align:null},"Gather and count unique values"),(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"uniq -c file.txt"))),(0,i.kt)("tr",{parentName:"tbody"},(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"gzip"),"/",(0,i.kt)("inlineCode",{parentName:"td"},"gunzip")),(0,i.kt)("td",{parentName:"tr",align:null},"General-purpose compression/decompression."),(0,i.kt)("td",{parentName:"tr",align:null},(0,i.kt)("inlineCode",{parentName:"td"},"gzip file.txt"))))),(0,i.kt)("p",null,"Here are some tips that will make life easier:"),(0,i.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Tips and tricks")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("ul",{parentName:"div"},(0,i.kt)("li",{parentName:"ul"},"The command line will ",(0,i.kt)("em",{parentName:"li"},"auto-complete")," filenames for you if you press the ",(0,i.kt)("inlineCode",{parentName:"li"},"tab")," key - this saves a lot of typing."),(0,i.kt)("li",{parentName:"ul"},"Press the up arrow to go back in your command history - you can then edit/rerun the same command."),(0,i.kt)("li",{parentName:"ul"},"In filenames, ",(0,i.kt)("inlineCode",{parentName:"li"},"./")," indicates the current directory, while ",(0,i.kt)("inlineCode",{parentName:"li"},"../")," indicates the parent directory (i.e. one higher up.)\nso for example ",(0,i.kt)("inlineCode",{parentName:"li"},"cd ../")," takes you one level higher.")),(0,i.kt)("p",{parentName:"div"},(0,i.kt)("inlineCode",{parentName:"p"},"ls")," is particularly useful for looking around - for example"),(0,i.kt)("ul",{parentName:"div"},(0,i.kt)("li",{parentName:"ul"},(0,i.kt)("inlineCode",{parentName:"li"},"ls")," on its own prints a simple listing."),(0,i.kt)("li",{parentName:"ul"},(0,i.kt)("inlineCode",{parentName:"li"},"ls -a")," will also include ",(0,i.kt)("em",{parentName:"li"},"hidden files")," - these are filenames starting with a ",(0,i.kt)("inlineCode",{parentName:"li"},".")," that are usually excluded."),(0,i.kt)("li",{parentName:"ul"},(0,i.kt)("inlineCode",{parentName:"li"},"ls -l")," will print a long listing - dates, file owners, file sizes, etc. (A command I use a lot is ",(0,i.kt)("inlineCode",{parentName:"li"},"ls -lht")," which lists files ordered by\nmodification time with human-readable file sizes.)")))),(0,i.kt)("p",null,"When you're ready, move on to download the tutorial data."),(0,i.kt)("h2",{id:"getting-started"},"Getting started"),(0,i.kt)("p",null,"To get started, create a new folder for the tutorial and change dir into it:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"mkdir cmdline_tutorial\ncd cmdline_tutorial\n")),(0,i.kt)("p",null,"Now download the gene annotation file from gencode and place it in that folder.  You can either:"),(0,i.kt)("ul",null,(0,i.kt)("li",{parentName:"ul"},(0,i.kt)("p",{parentName:"li"},"Download it from the ",(0,i.kt)("a",{parentName:"p",href:"https://www.gencodegenes.org/human/"},"gencode download page")," for human gene annotations -\nyou want the 'Comprehensive gene annotation' file in ",(0,i.kt)("inlineCode",{parentName:"p"},"GFF3")," format.")),(0,i.kt)("li",{parentName:"ul"},(0,i.kt)("p",{parentName:"li"},"Or download the copy of the gencode file that I have placed in\n",(0,i.kt)("a",{parentName:"p",href:"https://www.well.ox.ac.uk/~gav/projects/whg-training-resources/data/programming/programming_with_gene_annotations/"},"this folder"),"."))),(0,i.kt)("p",null,"For example this command should work to do the download:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"curl -O https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_41/gencode.v41.annotation.gff3.gz\n")),(0,i.kt)("div",{className:"admonition admonition-note alert alert--secondary"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"14",height:"16",viewBox:"0 0 14 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.3 5.69a.942.942 0 0 1-.28-.7c0-.28.09-.52.28-.7.19-.18.42-.28.7-.28.28 0 .52.09.7.28.18.19.28.42.28.7 0 .28-.09.52-.28.7a1 1 0 0 1-.7.3c-.28 0-.52-.11-.7-.3zM8 7.99c-.02-.25-.11-.48-.31-.69-.2-.19-.42-.3-.69-.31H6c-.27.02-.48.13-.69.31-.2.2-.3.44-.31.69h1v3c.02.27.11.5.31.69.2.2.42.31.69.31h1c.27 0 .48-.11.69-.31.2-.19.3-.42.31-.69H8V7.98v.01zM7 2.3c-3.14 0-5.7 2.54-5.7 5.68 0 3.14 2.56 5.7 5.7 5.7s5.7-2.55 5.7-5.7c0-3.15-2.56-5.69-5.7-5.69v.01zM7 .98c3.86 0 7 3.14 7 7s-3.14 7-7 7-7-3.12-7-7 3.14-7 7-7z"}))),"Note")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("p",{parentName:"div"},"The file I'll work with below has ",(0,i.kt)("inlineCode",{parentName:"p"},"v41")," in the name - as above it's called ",(0,i.kt)("inlineCode",{parentName:"p"},"gencode.v41.annotation.gff3.gz"),". If\nyou have a different version of this file that's fine - you may get slightly different results below but they\nshould be very similar."))),(0,i.kt)("h2",{id:"exploring-the-annotations"},"Exploring the annotations"),(0,i.kt)("h3",{id:"decompressing-the-file"},"Decompressing the file"),(0,i.kt)("p",null,"Does your file have ",(0,i.kt)("inlineCode",{parentName:"p"},".gz")," on the end of its filename?  And how big is it?  You can ",(0,i.kt)("inlineCode",{parentName:"p"},"ls -lh")," to find out:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"ls -lh ./\n")),(0,i.kt)("p",null,"You'll see something like:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"-rw-r--r--  1 user  group    57M 11 Oct 14:09 gencode.v41.annotation.gff3.gz\n")),(0,i.kt)("p",null,"This file is 57 megabytes big and ends with ",(0,i.kt)("inlineCode",{parentName:"p"},".gz"),". (If you're doens't have ",(0,i.kt)("inlineCode",{parentName:"p"},".gz")," on the end and looks bigger -\ndon't worry. It's likely your operating system has decompressed it for you.)"),(0,i.kt)("p",null,"If you have that ",(0,i.kt)("inlineCode",{parentName:"p"},".gz")," ending the file has been compressed with ",(0,i.kt)("inlineCode",{parentName:"p"},"gzip"),". We could work directly with it, but to\nmake life simple let's decompress it now:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"gunzip gencode.v41.annotation.gff3.gz\n")),(0,i.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Note")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("p",{parentName:"div"},"If you're bored typing the filename - just type the first few letter and press ",(0,i.kt)("inlineCode",{parentName:"p"},"tab")," to auto-complete it."))),(0,i.kt)("p",null,"Use ",(0,i.kt)("inlineCode",{parentName:"p"},"ls -lh")," to see how big it is now. You should see that it's lost the ",(0,i.kt)("inlineCode",{parentName:"p"},".gz")," ending and now is about 1.4\ngigabytes (i.e. ",(0,i.kt)("span",{parentName:"p",className:"math math-inline"},(0,i.kt)("span",{parentName:"span",className:"katex"},(0,i.kt)("span",{parentName:"span",className:"katex-mathml"},(0,i.kt)("math",{parentName:"span",xmlns:"http://www.w3.org/1998/Math/MathML"},(0,i.kt)("semantics",{parentName:"math"},(0,i.kt)("mrow",{parentName:"semantics"},(0,i.kt)("mn",{parentName:"mrow"},"1.4"),(0,i.kt)("mo",{parentName:"mrow"},"\xd7"),(0,i.kt)("mn",{parentName:"mrow"},"1"),(0,i.kt)("msup",{parentName:"mrow"},(0,i.kt)("mn",{parentName:"msup"},"0"),(0,i.kt)("mn",{parentName:"msup"},"9"))),(0,i.kt)("annotation",{parentName:"semantics",encoding:"application/x-tex"},"1.4 \\times 10^9")))),(0,i.kt)("span",{parentName:"span",className:"katex-html","aria-hidden":"true"},(0,i.kt)("span",{parentName:"span",className:"base"},(0,i.kt)("span",{parentName:"span",className:"strut",style:{height:"0.7278em",verticalAlign:"-0.0833em"}}),(0,i.kt)("span",{parentName:"span",className:"mord"},"1.4"),(0,i.kt)("span",{parentName:"span",className:"mspace",style:{marginRight:"0.2222em"}}),(0,i.kt)("span",{parentName:"span",className:"mbin"},"\xd7"),(0,i.kt)("span",{parentName:"span",className:"mspace",style:{marginRight:"0.2222em"}})),(0,i.kt)("span",{parentName:"span",className:"base"},(0,i.kt)("span",{parentName:"span",className:"strut",style:{height:"0.8141em"}}),(0,i.kt)("span",{parentName:"span",className:"mord"},"1"),(0,i.kt)("span",{parentName:"span",className:"mord"},(0,i.kt)("span",{parentName:"span",className:"mord"},"0"),(0,i.kt)("span",{parentName:"span",className:"msupsub"},(0,i.kt)("span",{parentName:"span",className:"vlist-t"},(0,i.kt)("span",{parentName:"span",className:"vlist-r"},(0,i.kt)("span",{parentName:"span",className:"vlist",style:{height:"0.8141em"}},(0,i.kt)("span",{parentName:"span",style:{top:"-3.063em",marginRight:"0.05em"}},(0,i.kt)("span",{parentName:"span",className:"pstrut",style:{height:"2.7em"}}),(0,i.kt)("span",{parentName:"span",className:"sizing reset-size6 size3 mtight"},(0,i.kt)("span",{parentName:"span",className:"mord mtight"},"9"))))))))))))," bytes) big.  So ",(0,i.kt)("inlineCode",{parentName:"p"},"gzip")," compressed it by about 24 times!"),(0,i.kt)("h3",{id:"viewing-the-file"},"Viewing the file"),(0,i.kt)("p",null,"Use the ",(0,i.kt)("inlineCode",{parentName:"p"},"less")," command to view the file:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"less -S gencode.v41.annotation.gff3\n")),(0,i.kt)("p",null,"You can scroll around and have a look in there.  You should see some ",(0,i.kt)("strong",{parentName:"p"},"metadata")," lines at the top (they start with ",(0,i.kt)("inlineCode",{parentName:"p"},"#"),").  They look like this:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"##gff-version 3\n#description: evidence-based annotation of the human genome (GRCh38), version 41 (Ensembl 107)\n#provider: GENCODE\n#contact: gencode-help@ebi.ac.uk\n#format: gff3\n#date: 2022-05-12\n##sequence-region chr1 1 248956422\n")),(0,i.kt)("p",null,"...and include information on the human genome assembly used (GRCh38, what's known as 'build 38') and other\nthings."),(0,i.kt)("p",null,"This is followed by some ",(0,i.kt)("em",{parentName:"p"},"data rows"),", that look like:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"chr1    HAVANA  gene    11869   14409   .       +       .       ID=ENSG00000223972.5;gene_id=ENSG00000223972.>\nchr1    HAVANA  transcript      11869   14409   .       +       .       ID=ENST00000456328.2;Parent=ENSG00000>\nchr1    HAVANA  exon    11869   12227   .       +       .       ID=exon:ENST00000456328.2:1;Parent=ENST000004>\nchr1    HAVANA  exon    12613   12721   .       +       .       ID=exon:ENST00000456328.2:2;Parent=ENST000004>\n")),(0,i.kt)("p",null,"When you want to quit ",(0,i.kt)("inlineCode",{parentName:"p"},"less"),", press ",(0,i.kt)("inlineCode",{parentName:"p"},"q"),"."),(0,i.kt)("div",{className:"admonition admonition-note alert alert--secondary"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"14",height:"16",viewBox:"0 0 14 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.3 5.69a.942.942 0 0 1-.28-.7c0-.28.09-.52.28-.7.19-.18.42-.28.7-.28.28 0 .52.09.7.28.18.19.28.42.28.7 0 .28-.09.52-.28.7a1 1 0 0 1-.7.3c-.28 0-.52-.11-.7-.3zM8 7.99c-.02-.25-.11-.48-.31-.69-.2-.19-.42-.3-.69-.31H6c-.27.02-.48.13-.69.31-.2.2-.3.44-.31.69h1v3c.02.27.11.5.31.69.2.2.42.31.69.31h1c.27 0 .48-.11.69-.31.2-.19.3-.42.31-.69H8V7.98v.01zM7 2.3c-3.14 0-5.7 2.54-5.7 5.68 0 3.14 2.56 5.7 5.7 5.7s5.7-2.55 5.7-5.7c0-3.15-2.56-5.69-5.7-5.69v.01zM7 .98c3.86 0 7 3.14 7 7s-3.14 7-7 7-7-3.12-7-7 3.14-7 7-7z"}))),"Tip")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("p",{parentName:"div"},"What's the ",(0,i.kt)("inlineCode",{parentName:"p"},"-S")," for in that less command?  Well try it without and you'll see:"),(0,i.kt)("pre",{parentName:"div"},(0,i.kt)("code",{parentName:"pre"},"less gencode.v41.annotation.gff3\n")),(0,i.kt)("p",{parentName:"div"},"The ",(0,i.kt)("inlineCode",{parentName:"p"},"-S")," tells ",(0,i.kt)("inlineCode",{parentName:"p"},"less")," to extend all the lines off the right of the screen - without it they wrap around which\nmakes reading the file pretty difficult."))),(0,i.kt)("p",null,"So, what does all that data mean? This file format is one of those annoying ones that ",(0,i.kt)("strong",{parentName:"p"},"includes no column\nnames"),". To figure out what they mean, you have to look at the GFF3 specification. You can find this ",(0,i.kt)("a",{parentName:"p",href:"https://www.gencodegenes.org/pages/data_format.html"},"on the\nGENCODE site")," or a similar description ",(0,i.kt)("a",{parentName:"p",href:"https://www.google.com/search?client=safari&rls=en&q=ensembl.org%2F+GFF3&ie=UTF-8&oe=UTF-8"},"on\nEnsembl")),(0,i.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Question")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("p",{parentName:"div"},"Look at the first 'gene' in the file. By manually looking at the file and comparing to the file specification,\ncan you figure out:"),(0,i.kt)("ul",{parentName:"div"},(0,i.kt)("li",{parentName:"ul"},"which chromosome is it on?"),(0,i.kt)("li",{parentName:"ul"},"which strand is it transcribed on?"),(0,i.kt)("li",{parentName:"ul"},"what type of gene is it - is it protein-coding? (Hint: look for the ",(0,i.kt)("inlineCode",{parentName:"li"},"gene_type")," attribute.  It can be looked up in the ",(0,i.kt)("a",{parentName:"li",href:"https://www.gencodegenes.org/pages/biotypes.html"},"list of biotypes"),".)"),(0,i.kt)("li",{parentName:"ul"},"how many transcripts does the gene have?")))),(0,i.kt)("p",null,"Note that to answer this last question, you'll need to look at how the different rows in the file are related to\neach other.  In short:"),(0,i.kt)("ul",null,(0,i.kt)("li",{parentName:"ul"},"each row has an ",(0,i.kt)("inlineCode",{parentName:"li"},"ID")," attribute"),(0,i.kt)("li",{parentName:"ul"},"some rows also have a ",(0,i.kt)("inlineCode",{parentName:"li"},"Parent")," attribute")),(0,i.kt)("p",null,"these attributes make the records in the file into a tree.  So conceptually the structure looks something like"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"            gene\n           /    \\      \n transcript1   transcript2  ...\n")),(0,i.kt)("p",null,"i.e. each transcript has a parent gene - which means that it represents an observed or predicted RNA transcript of that\ngene.  Transcripts themselves have exons - the parts of the transcript that actually make it to mature messenger RNA - so actually it is more like this:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"             gene\n            /    \\      \n  transcript1   transcript2  ...\n   /      |      |       \\\nexon1  exon2    exon1    exon2\n")),(0,i.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Question")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("p",{parentName:"div"},"There are also ",(0,i.kt)("em",{parentName:"p"},"coding sequence")," records (",(0,i.kt)("inlineCode",{parentName:"p"},"type=CDS"),"). Can you tell what these have as parents - exons,\ntranscripts or genes?"))),(0,i.kt)("h3",{id:"counting-genes"},"Counting genes"),(0,i.kt)("p",null,"Quit less (by pressing ",(0,i.kt)("inlineCode",{parentName:"p"},"q"),") and let's generate some basic statistics. "),(0,i.kt)("p",null,"First, how many genes and other things are in the file? For this, we can use the ",(0,i.kt)("inlineCode",{parentName:"p"},"cut")," command to cut out the\nthird column (which contains the ",(0,i.kt)("inlineCode",{parentName:"p"},"type"),"). Then we'll ",(0,i.kt)("em",{parentName:"p"},"pipe the output")," into the sort command (which sorts the\nrows).  And finally we will ask the ",(0,i.kt)("inlineCode",{parentName:"p"},"uniq")," command to count:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"cut -f3 gencode.v41.annotation.gff3 | sort | uniq -c\n")),(0,i.kt)("p",null,"This will take a minute or two to run - it's a big file!"),(0,i.kt)("p",null,":::"),(0,i.kt)("p",null,"Ok - the output is not really useful because of all the metadata.  Let's use ",(0,i.kt)("inlineCode",{parentName:"p"},"grep -v")," to get rid of it:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"grep -v '#' gencode.v41.annotation.gff3 | cut -f3 | sort | uniq -c\n")),(0,i.kt)("p",null,"This ",(0,i.kt)("strong",{parentName:"p"},"finds lines")," that don't contain ",(0,i.kt)("inlineCode",{parentName:"p"},"#"),", ",(0,i.kt)("strong",{parentName:"p"},"extracts the third column")," from them, ",(0,i.kt)("strong",{parentName:"p"},"sorts them"),", and ",(0,i.kt)("strong",{parentName:"p"},"counts\nthe unique values"),"."),(0,i.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Picking apart the pipeline")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("p",{parentName:"div"},"If this command isn't making sense to you, a good idea is to look at what each step does.\nTry running these commands one by one to parse it apart:"),(0,i.kt)("p",{parentName:"div"},"View the whole file:"),(0,i.kt)("pre",{parentName:"div"},(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"less -S gencode.v41.annotation.gff3\n")),(0,i.kt)("p",{parentName:"div"},"Just the data rows:"),(0,i.kt)("pre",{parentName:"div"},(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"grep -v '#' gencode.v41.annotation.gff3 | less -S\n")),(0,i.kt)("p",{parentName:"div"},"Just the third column of the data rows:"),(0,i.kt)("pre",{parentName:"div"},(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"grep -v '#' gencode.v41.annotation.gff3 | cut -f3 | less -S\n")),(0,i.kt)("p",{parentName:"div"},"The third column sorted:"),(0,i.kt)("pre",{parentName:"div"},(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"grep -v '#' gencode.v41.annotation.gff3 | cut -f3 | sort | less -S\n")),(0,i.kt)("p",{parentName:"div"},"The sorted unique values in the third column...."),(0,i.kt)("pre",{parentName:"div"},(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"grep -v '#' gencode.v41.annotation.gff3 | cut -f3 | sort | uniq | less -S\n")),(0,i.kt)("p",{parentName:"div"},"...and the same thing with counts:"),(0,i.kt)("pre",{parentName:"div"},(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"grep -v '#' gencode.v41.annotation.gff3 | cut -f3 | sort | uniq -c | less -S\n")),(0,i.kt)("p",{parentName:"div"},"Hopefully by this point it is clear(er) what each step is doing."))),(0,i.kt)("p",null,"It prints:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"872459 CDS\n1625321 exon\n171599 five_prime_UTR\n 61852 gene\n 97009 start_codon\n 90749 stop_codon\n   119 stop_codon_redefined_as_selenocysteine\n203260 three_prime_UTR\n251236 transcript\n")),(0,i.kt)("p",null,"So - there are 1.6 million exons in the file and... ",(0,i.kt)("em",{parentName:"p"},"wait a moment"),", are there really 60,000 genes in the human genome?"),(0,i.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Question")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("p",{parentName:"div"},"The number 60,000 is way too big - why?"))),(0,i.kt)("p",null,"Correct! As we saw above, not all of the genes in this file are protein-coding (the first one said it was a\n'transcribed unprocessed pseudogene'.) Let's try to count just the protein-coding ones. To do this we will use a\ncouple of commands - ",(0,i.kt)("inlineCode",{parentName:"p"},"awk"),' which we are here using just to select rows with "gene" in the ',(0,i.kt)("inlineCode",{parentName:"p"},"type")," column, and\n",(0,i.kt)("inlineCode",{parentName:"p"},"wc")," which will count the number of lines:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"cat gencode.v41.annotation.gff3  | awk '$3==\"gene\"' | grep 'gene_type=protein_coding' | wc -l\n")),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"20017\n")),(0,i.kt)("p",null,"This is a much more sensible number - there are about 20,000 protein-coding genes in the human genome.\nThat\u2019s a lot but we are big animals!"),(0,i.kt)("h3",{id:"investigating-specific-genes"},"Investigating specific genes."),(0,i.kt)("p",null,"Let's switch track and try to dig out info about a specific gene - ",(0,i.kt)("a",{parentName:"p",href:"https://en.wikipedia.org/wiki/FUT2"},(0,i.kt)("em",{parentName:"a"},"FUT2")),".\nThat's an interesting gene because it encodes a fucosyltransferase is involved the synthesis pathway for\n'soluble' A and B antigens - that is A and B antigens found in blood plasma and other fluids. Mutations in FUT2\naffect whether these antigens are secreted. Because norovirus binds to these antigens, these mutations can\nconfer protection against norovirus."),(0,i.kt)("p",null,"A simple way to look this up is to just to grep (i.e. conduct a text search) for ",(0,i.kt)("inlineCode",{parentName:"p"},"FUT2"),":"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"grep FUT2 gencode.v41.annotation.gff3 | less -S\n")),(0,i.kt)("p",null,"Unfortunately that returns a lot of rows - let's just get genes:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"grep FUT2 gencode.v41.annotation.gff3 | awk '$3 == \"gene\"' | less -S\n")),(0,i.kt)("p",null,"Ok, this returns two records. If you look at the ",(0,i.kt)("inlineCode",{parentName:"p"},"gene_name")," attribute you'll see one, on chromosome 19, is\n",(0,i.kt)("em",{parentName:"p"},"FUT2"),", while the other is a different gene called ",(0,i.kt)("inlineCode",{parentName:"p"},"POFUT2"),". Let's use that to do a bit better:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"grep 'gene_name=FUT2' gencode.v41.annotation.gff3 | awk '$3 == \"gene\"' | less -S\n")),(0,i.kt)("p",null,"We got it!  Copy its ID to the clipboard - in my file it is ",(0,i.kt)("inlineCode",{parentName:"p"},"ENSG00000176920.13"),"."),(0,i.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Questions")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("ul",{parentName:"div"},(0,i.kt)("li",{parentName:"ul"},"How long is FUT2 on the chromosome?")),(0,i.kt)("p",{parentName:"div"},(0,i.kt)("strong",{parentName:"p"},"Note.")," to get the answer 100% right, you actually have to take the formula"),(0,i.kt)("div",{parentName:"div",className:"math math-display"},(0,i.kt)("span",{parentName:"div",className:"katex-display"},(0,i.kt)("span",{parentName:"span",className:"katex"},(0,i.kt)("span",{parentName:"span",className:"katex-mathml"},(0,i.kt)("math",{parentName:"span",xmlns:"http://www.w3.org/1998/Math/MathML",display:"block"},(0,i.kt)("semantics",{parentName:"math"},(0,i.kt)("mrow",{parentName:"semantics"},(0,i.kt)("mtext",{parentName:"mrow"},"end\xa0coordinate"),(0,i.kt)("mo",{parentName:"mrow"},"\u2212"),(0,i.kt)("mtext",{parentName:"mrow"},"start\xa0coordinate"),(0,i.kt)("mo",{parentName:"mrow"},"+"),(0,i.kt)("mn",{parentName:"mrow"},"1")),(0,i.kt)("annotation",{parentName:"semantics",encoding:"application/x-tex"},"\\text{end coordinate} - \\text{start coordinate} + 1")))),(0,i.kt)("span",{parentName:"span",className:"katex-html","aria-hidden":"true"},(0,i.kt)("span",{parentName:"span",className:"base"},(0,i.kt)("span",{parentName:"span",className:"strut",style:{height:"0.7778em",verticalAlign:"-0.0833em"}}),(0,i.kt)("span",{parentName:"span",className:"mord text"},(0,i.kt)("span",{parentName:"span",className:"mord"},"end\xa0coordinate")),(0,i.kt)("span",{parentName:"span",className:"mspace",style:{marginRight:"0.2222em"}}),(0,i.kt)("span",{parentName:"span",className:"mbin"},"\u2212"),(0,i.kt)("span",{parentName:"span",className:"mspace",style:{marginRight:"0.2222em"}})),(0,i.kt)("span",{parentName:"span",className:"base"},(0,i.kt)("span",{parentName:"span",className:"strut",style:{height:"0.7778em",verticalAlign:"-0.0833em"}}),(0,i.kt)("span",{parentName:"span",className:"mord text"},(0,i.kt)("span",{parentName:"span",className:"mord"},"start\xa0coordinate")),(0,i.kt)("span",{parentName:"span",className:"mspace",style:{marginRight:"0.2222em"}}),(0,i.kt)("span",{parentName:"span",className:"mbin"},"+"),(0,i.kt)("span",{parentName:"span",className:"mspace",style:{marginRight:"0.2222em"}})),(0,i.kt)("span",{parentName:"span",className:"base"},(0,i.kt)("span",{parentName:"span",className:"strut",style:{height:"0.6444em"}}),(0,i.kt)("span",{parentName:"span",className:"mord"},"1")))))),(0,i.kt)("p",{parentName:"div"},"This is because both start and end are expressed in a ",(0,i.kt)("em",{parentName:"p"},"1-based, closed coordinate system")," i.e. they both point\nat bases included in the gene.  (Think of a gene with only two bases in it to see why this is.)"))),(0,i.kt)("h3",{id:"finding-transcripts"},"Finding transcripts"),(0,i.kt)("p",null,"So how many transcripts does ",(0,i.kt)("em",{parentName:"p"},"FUT2")," have? Well we know how to do this - look for ",(0,i.kt)("em",{parentName:"p"},"transcript")," records with the\n",(0,i.kt)("em",{parentName:"p"},"FUT2")," gene as parent:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"grep 'Parent=ENSG00000176920.13' gencode.v41.annotation.gff3 | awk '$3 == \"transcript\"' | less -S\n")),(0,i.kt)("p",null,"So it has 4 transcripts - that is, the file suggests the gene may be transcribed to mRNA in 4 different ways.\nScroll around a bit to look at the attributes of these transcripts.  If you look closely you'll see there is some more information in there.\nFor example a ",(0,i.kt)("strong",{parentName:"p"},"transcript support\nlevel")," which reflects how confident GENCODE is about the transcript. See\n",(0,i.kt)("a",{parentName:"p",href:"https://www.ensembl.org/info/genome/genebuild/transcript_quality_tags.html"},"the Ensembl page"),"\nfor a description of these."),(0,i.kt)("p",null,'One of these transcripts (ENST00000425340.3) is also marked as \u2018Ensemble canonical\u2019 which means\n"',(0,i.kt)("a",{parentName:"p",href:"https://www.ensembl.org/info/genome/genebuild/canonical.html"},"a single, representative transcript identified at every locus"),"\".\nSo let's focus on that transcript and dig a bit deeper"),(0,i.kt)("h3",{id:"finding-exons"},"Finding exons"),(0,i.kt)("p",null,"This is easy now:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"grep 'Parent=ENST00000425340.3' gencode.v41.annotation.gff3 | awk '$3 == \"exon\"' | less -S\n")),(0,i.kt)("p",null,"Aha, it has two exons."),(0,i.kt)("p",null,"So, how long are these exons?  To make that easier let's use ",(0,i.kt)("inlineCode",{parentName:"p"},"cut")," to get rid of the noise:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"grep 'Parent=ENST00000425340.3' gencode.v41.annotation.gff3 | awk '$3 == \"exon\"' | cut -f1,3-5\n")),(0,i.kt)("p",null,"Adding that up, the two exons have length 119 and 2,997 - so only about 30% of the gene is actually transcribed into RNA!"),(0,i.kt)("p",null,"What about the bit that codes for protein? We can find that by looking for the ",(0,i.kt)("em",{parentName:"p"},"coding sequence")," records - they\nhave ",(0,i.kt)("inlineCode",{parentName:"p"},"type=CDS"),":"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"grep 'Parent=ENST00000425340.3' gencode.v41.annotation.gff3 | awk '$3 == \"exon\"' | cut -f1,3-5\n")),(0,i.kt)("p",null,"If you look at this you'll see the gene has one annotated coding sequence, and it lives entirely inside the\nsecond exon. Its length is 1032 base pairs. So **only about ",(0,i.kt)("span",{parentName:"p",className:"math math-inline"},(0,i.kt)("span",{parentName:"span",className:"katex"},(0,i.kt)("span",{parentName:"span",className:"katex-mathml"},(0,i.kt)("math",{parentName:"span",xmlns:"http://www.w3.org/1998/Math/MathML"},(0,i.kt)("semantics",{parentName:"math"},(0,i.kt)("mrow",{parentName:"semantics"},(0,i.kt)("mfrac",{parentName:"mrow"},(0,i.kt)("mn",{parentName:"mfrac"},"1"),(0,i.kt)("mn",{parentName:"mfrac"},"10"))),(0,i.kt)("annotation",{parentName:"semantics",encoding:"application/x-tex"},"\\tfrac{1}{10}")))),(0,i.kt)("span",{parentName:"span",className:"katex-html","aria-hidden":"true"},(0,i.kt)("span",{parentName:"span",className:"base"},(0,i.kt)("span",{parentName:"span",className:"strut",style:{height:"1.1901em",verticalAlign:"-0.345em"}}),(0,i.kt)("span",{parentName:"span",className:"mord"},(0,i.kt)("span",{parentName:"span",className:"mopen nulldelimiter"}),(0,i.kt)("span",{parentName:"span",className:"mfrac"},(0,i.kt)("span",{parentName:"span",className:"vlist-t vlist-t2"},(0,i.kt)("span",{parentName:"span",className:"vlist-r"},(0,i.kt)("span",{parentName:"span",className:"vlist",style:{height:"0.8451em"}},(0,i.kt)("span",{parentName:"span",style:{top:"-2.655em"}},(0,i.kt)("span",{parentName:"span",className:"pstrut",style:{height:"3em"}}),(0,i.kt)("span",{parentName:"span",className:"sizing reset-size6 size3 mtight"},(0,i.kt)("span",{parentName:"span",className:"mord mtight"},(0,i.kt)("span",{parentName:"span",className:"mord mtight"},"10")))),(0,i.kt)("span",{parentName:"span",style:{top:"-3.23em"}},(0,i.kt)("span",{parentName:"span",className:"pstrut",style:{height:"3em"}}),(0,i.kt)("span",{parentName:"span",className:"frac-line",style:{borderBottomWidth:"0.04em"}})),(0,i.kt)("span",{parentName:"span",style:{top:"-3.394em"}},(0,i.kt)("span",{parentName:"span",className:"pstrut",style:{height:"3em"}}),(0,i.kt)("span",{parentName:"span",className:"sizing reset-size6 size3 mtight"},(0,i.kt)("span",{parentName:"span",className:"mord mtight"},(0,i.kt)("span",{parentName:"span",className:"mord mtight"},"1"))))),(0,i.kt)("span",{parentName:"span",className:"vlist-s"},"\u200b")),(0,i.kt)("span",{parentName:"span",className:"vlist-r"},(0,i.kt)("span",{parentName:"span",className:"vlist",style:{height:"0.345em"}},(0,i.kt)("span",{parentName:"span"}))))),(0,i.kt)("span",{parentName:"span",className:"mclose nulldelimiter"})))))),"th of the gene codes for protein."),(0,i.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Note")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("p",{parentName:"div"},"If we've got this right then the nucleotide length of the coding sequence should be a multiple of something - what?\nIs 1032 an appropriate multiple?"))),(0,i.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Challenge question Now repeat the above process for another gene and see if things look similar. For")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("p",{parentName:"div"},"example, try the genes that encode ",(0,i.kt)("a",{parentName:"p",href:"https://en.wikipedia.org/wiki/Hemoglobin_subunit_alpha"},"alpha globin"),",\nnamed ",(0,i.kt)("inlineCode",{parentName:"p"},"HBA1")," and ",(0,i.kt)("inlineCode",{parentName:"p"},"HBA2"),"."))),(0,i.kt)("h2",{id:"browsing-the-genome"},"Browsing the genome"),(0,i.kt)("p",null,"Can it be right that only a small fraction of these genes is coding? To confirm our results, let\u2019s look up on\nthe UCSC genome browser:"),(0,i.kt)("ol",null,(0,i.kt)("li",{parentName:"ol"},(0,i.kt)("p",{parentName:"li"},"Visit ",(0,i.kt)("a",{parentName:"p",href:"https://genome.ucsc.edu"},"https://genome.ucsc.edu")," and click \u2018Genome Browser\u2019 (choose the Euro mirror)")),(0,i.kt)("li",{parentName:"ol"},(0,i.kt)("p",{parentName:"li"},"In \u2018Position/Search Term\u2019 type the gene name - say FUT2. (Make sure the 'Human Assembly' is set to\n",(0,i.kt)("inlineCode",{parentName:"p"},"GRCh38/hg38")," in the dropdown above). This may give you a list of genes - if so click on the one with the right\nname.")),(0,i.kt)("li",{parentName:"ol"},(0,i.kt)("p",{parentName:"li"},"You'll see the gene in its location on the genome. Try zooming out a little to see the gene in its\ncontext.  It should look something like this:"))),(0,i.kt)("p",null,(0,i.kt)("img",{alt:"img",src:n(1212).Z,width:"1024",height:"564"})),(0,i.kt)("p",null,"Sure enough, most of the gene is in introns."),(0,i.kt)("p",null,"You can click on the gene to read more information about it. Repeat this for the other genes you looked up e.g.\n",(0,i.kt)("inlineCode",{parentName:"p"},"HBA1")," - did you get it right?"),(0,i.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Note")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("p",{parentName:"div"},"As you can probably see, the genome browser contains an incredible wealth of information about the human genome,\nwith data representing many thousands of experiments done by researchers worldwide. Feel free to explore the browser to see what you can glean from the data presented - try clicking on things for\nmore information. (But watch out - it can be a bit overwhelming at first!)"),(0,i.kt)("p",{parentName:"div"},"For example, the bottom track in the image above shows ",(0,i.kt)("em",{parentName:"p"},"common genetic variants")," - some of them are coloured.\nYou can click on them for more information.  Can you find a SNP that encodes a change to the protein?"))),(0,i.kt)("h2",{id:"challenge-questions"},"Challenge questions"),(0,i.kt)("p",null,"The SNP ",(0,i.kt)("a",{parentName:"p",href:"https://www.ncbi.nlm.nih.gov/snp/rs601338"},"rs601338")," controls \u2019secretor status\u2019 (i.e. whether A/B\nantigens are secreted into bodily fluids) Individuals with GG or GA genotypes at this SNP are secretors, and\nindividuals with homozygous AA genotypes are non-secretors - as described in ",(0,i.kt)("a",{parentName:"p",href:"10.12688/wellcomeopenres.14636.2"},"this paper based on the ALSPAC\ncohort"),"."),(0,i.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Can you solve?")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("p",{parentName:"div"},"Q1: Where is rs601338 in the gene?  What change does it make to the encoded protein?"),(0,i.kt)("p",{parentName:"div"},"Q2: Find an individual from the 1000 Genomes Project data with secretor genotype, and an individual with\nnon-secretor genotype."),(0,i.kt)("p",{parentName:"div"},"(Hint: you can use the ",(0,i.kt)("a",{parentName:"p",href:"https://genome.ucsc.edu"},"UCSC genome browser")," and the ",(0,i.kt)("a",{parentName:"p",href:"http://www.ensembl.org/index.html"},"Ensembl genome\nbrowser")," to answer these questions.)"))))}d.isMDXComponent=!0},1212:(e,t,n)=>{n.d(t,{Z:()=>a});const a=n.p+"assets/images/ucsc_genome_browser-2afe80c12dcf4ce74f497a13f8506fa6.png"}}]);