"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[7363],{3905:(e,t,n)=>{n.d(t,{Zo:()=>g,kt:()=>c});var a=n(7294);function r(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function i(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var a=Object.getOwnPropertySymbols(e);t&&(a=a.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,a)}return n}function o(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?i(Object(n),!0).forEach((function(t){r(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):i(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function s(e,t){if(null==e)return{};var n,a,r=function(e,t){if(null==e)return{};var n,a,r={},i=Object.keys(e);for(a=0;a<i.length;a++)n=i[a],t.indexOf(n)>=0||(r[n]=e[n]);return r}(e,t);if(Object.getOwnPropertySymbols){var i=Object.getOwnPropertySymbols(e);for(a=0;a<i.length;a++)n=i[a],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(r[n]=e[n])}return r}var l=a.createContext({}),p=function(e){var t=a.useContext(l),n=t;return e&&(n="function"==typeof e?e(t):o(o({},t),e)),n},g=function(e){var t=p(e.components);return a.createElement(l.Provider,{value:t},e.children)},h={inlineCode:"code",wrapper:function(e){var t=e.children;return a.createElement(a.Fragment,{},t)}},m=a.forwardRef((function(e,t){var n=e.components,r=e.mdxType,i=e.originalType,l=e.parentName,g=s(e,["components","mdxType","originalType","parentName"]),m=p(n),c=r,u=m["".concat(l,".").concat(c)]||m[c]||h[c]||i;return n?a.createElement(u,o(o({ref:t},g),{},{components:n})):a.createElement(u,o({ref:t},g))}));function c(e,t){var n=arguments,r=t&&t.mdxType;if("string"==typeof e||r){var i=n.length,o=new Array(i);o[0]=m;var s={};for(var l in t)hasOwnProperty.call(t,l)&&(s[l]=t[l]);s.originalType=e,s.mdxType="string"==typeof e?e:r,o[1]=s;for(var p=2;p<i;p++)o[p]=n[p];return a.createElement.apply(null,o)}return a.createElement.apply(null,n)}m.displayName="MDXCreateElement"},9407:(e,t,n)=>{n.r(t),n.d(t,{assets:()=>l,contentTitle:()=>o,default:()=>h,frontMatter:()=>i,metadata:()=>s,toc:()=>p});var a=n(7462),r=(n(7294),n(3905));const i={sidebar_position:9},o="Extremely big, small and complex genes",s={unversionedId:"programming/programming_with_gene_annotations/Counting_genes_3",id:"programming/programming_with_gene_annotations/Counting_genes_3",title:"Extremely big, small and complex genes",description:"What is the largest gene, the gene with the most exons, or the gene with the most transcripts in each species?",source:"@site/docs/programming/programming_with_gene_annotations/Counting_genes_3.md",sourceDirName:"programming/programming_with_gene_annotations",slug:"/programming/programming_with_gene_annotations/Counting_genes_3",permalink:"/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_3",draft:!1,editUrl:"https://github.com/whg-training/whg-training-resources/edit/main/docs/programming/programming_with_gene_annotations/Counting_genes_3.md",tags:[],version:"current",sidebarPosition:9,frontMatter:{sidebar_position:9},sidebar:"tutorialSidebar",previous:{title:"Counting genes II: types of protein-coding genes",permalink:"/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_2"},next:{title:"Getting sequence lengths",permalink:"/whg-training-resources/programming/programming_with_gene_annotations/Getting_sequence_lengths"}},l={},p=[{value:"Single-exon genes",id:"single-exon-genes",level:3},{value:"Visualising",id:"visualising",level:3},{value:"Getting sequence lengths",id:"getting-sequence-lengths",level:3}],g={toc:p};function h(e){let{components:t,...i}=e;return(0,r.kt)("wrapper",(0,a.Z)({},g,i,{components:t,mdxType:"MDXLayout"}),(0,r.kt)("h1",{id:"extremely-big-small-and-complex-genes"},"Extremely big, small and complex genes"),(0,r.kt)("p",null,"What is the largest gene, the gene with the most exons, or the gene with the most transcripts in each species?"),(0,r.kt)("p",null,"You can work this out, but my solutions are in ",(0,r.kt)("a",{parentName:"p",href:"https://github.com/whg-training/whg-training-resources/blob/main/docs/programming/programming_with_gene_annotations/solutions/part3"},"this folder")," for reference:\nOr download as: ",(0,r.kt)("a",{target:"_blank",href:n(5397).Z},"sql solution")," and ",(0,r.kt)("a",{target:"_blank",href:n(2350).Z},"python solution"),"."),(0,r.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,r.kt)("div",{parentName:"div",className:"admonition-heading"},(0,r.kt)("h5",{parentName:"div"},(0,r.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,r.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,r.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Note")),(0,r.kt)("div",{parentName:"div",className:"admonition-content"},(0,r.kt)("p",{parentName:"div"},"To count exons per gene, I have used a not-very-good approach: taking the average across transcripts for each gene. This isn't\nvery sensible as there are often one or a few 'major' transcripts, but also sometimes some very short (or very long) transcripts\nlisted for a gene that will skew results. Ideally we should take a 'canonical' transcript."),(0,r.kt)("p",{parentName:"div"},"How to do that? Well, Ensembl have files containing a set of representative transcripts for each species ",(0,r.kt)("a",{parentName:"p",href:"https://ftp.ensembl.org/pub/current_tsv/"},"in this\nfolder"),".\nYou can also read ",(0,r.kt)("a",{parentName:"p",href:"https://www.ensembl.org/info/genome/genebuild/canonical.html"},"more details on what thse contain"),"."),(0,r.kt)("p",{parentName:"div"},"An ",(0,r.kt)("strong",{parentName:"p"},"extra challenge")," is therefore to incorporate these 'canonical transcripts' into your program - then you can focus on these\ntranscripts in your analysis."))),(0,r.kt)("p",null,"Here are some things I noticed:"),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},(0,r.kt)("p",{parentName:"li"},"In all the data I looked at, there's only one gene with > 100 transcripts. It is the human gene ",(0,r.kt)("a",{parentName:"p",href:"https://www.uniprot.org/uniprot/P53779"},(0,r.kt)("em",{parentName:"a"},"MAPK10"),', "Mitogen-activated\nprotein kinase 10"')," and it has 151 transcripts! It has so many that it takes a while to load in\n",(0,r.kt)("a",{parentName:"p",href:"http://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000109339;r=4:85990007-86594625"},"Ensembl")," and the\n",(0,r.kt)("a",{parentName:"p",href:"https://genome-euro.ucsc.edu/cgi-bin/hgTracks?db=hg38&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chr4%3A86016491%2D86594110&hgsid=275896231_HieWdPQTMOsgYQUFAnwTALgAECs0"},"UCSC Genome browser"),".\nIf you explore these genome browsers you will find a wealth of information on this gene including its expression patterns and function. Is there much published literature about it?\nWhat does it look like in other species?")),(0,r.kt)("li",{parentName:"ul"},(0,r.kt)("p",{parentName:"li"},"Quite a few genes have lots of exons, but one that really stands out is ",(0,r.kt)("a",{parentName:"p",href:"https://en.wikipedia.org/wiki/Titin"},(0,r.kt)("em",{parentName:"a"},"Titin")),".\nIt has a slightly incredible 184 exons in Chimpanzees (",(0,r.kt)("em",{parentName:"p"},"Pan troglodytes"),") and ~113 in humans.\nAnother is ",(0,r.kt)("a",{parentName:"p",href:"https://en.wikipedia.org/wiki/Nebulin"},(0,r.kt)("em",{parentName:"a"},"Nebulin")),".\nThese genes seem especially huge in Chimpanzees. Are they big in all species? In all great apes?\nIs there literature that might explain the results for these genes, or other genes with lots of exons?\n")),(0,r.kt)("li",{parentName:"ul"},(0,r.kt)("p",{parentName:"li"},"The genes with lots of exons aren't the largest though (by length). In humans and chimpanzees, that accolade goes\nto ",(0,r.kt)("a",{parentName:"p",href:"http://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000078328;r=16:5239802-7713340"},(0,r.kt)("inlineCode",{parentName:"a"},"RBFOX1")),"\nwith a whopping length of around 2 and a half megabases, or 0.08% of the genome. What does this look\nlike across species? Where is it expressed? (Interestingly, this gene has been previously associated ",(0,r.kt)("a",{parentName:"p",href:"https://www.nature.com/articles/s41380-018-0068-7"},"with aggressive\nbehaviour"),").")),(0,r.kt)("li",{parentName:"ul"},(0,r.kt)("p",{parentName:"li"},"There are also some pretty tiny genes around. In humans the smallest seems to be\n",(0,r.kt)("a",{parentName:"p",href:"http://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000288608;r=7:151061928-151061978;t=ENST00000674552"},(0,r.kt)("em",{parentName:"a"},"ENSG00000288608")),",\nwhich is only 50 base pairs long. (However, it seems to be contained in an exon of the much\nlonger ",(0,r.kt)("a",{parentName:"p",href:"https://genome-euro.ucsc.edu/cgi-bin/hgTracks?db=hg38&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chr7%3A151061877%2D151062029&hgsid=275886241_mxgWOGDr4elcf2SdW0zGz6ukwVJW"},(0,r.kt)("em",{parentName:"a"},"SLC4A2")," Anion exhange protein 2"),".\nI'm not quite sure what this means...)"))),(0,r.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,r.kt)("div",{parentName:"div",className:"admonition-heading"},(0,r.kt)("h5",{parentName:"div"},(0,r.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,r.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,r.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Challenge")),(0,r.kt)("div",{parentName:"div",className:"admonition-content"},(0,r.kt)("p",{parentName:"div"},"What we might really want to know is: the length of the encoded protein.  Can you find the genes with the longest exon sequence or the longest\nprotein-coding sequence?  (This will involve adding up the lengths of exons or CDS records in each transcript.)"))),(0,r.kt)("p",null,"What other extreme genes are there?"),(0,r.kt)("h3",{id:"single-exon-genes"},"Single-exon genes"),(0,r.kt)("p",null,"What proportion of protein-coding genes in each species have a single exon? (My solution is also in\n",(0,r.kt)("a",{target:"_blank",href:n(5397).Z},(0,r.kt)("code",null,"solutions/part3/solutions.sql"))," and\n",(0,r.kt)("a",{target:"_blank",href:n(2350).Z},(0,r.kt)("code",null,"solutions/part3/solutions.py")),"). Almost 44% of ",(0,r.kt)("a",{parentName:"p",href:"https://en.wikipedia.org/wiki/Plasmodium_falciparum"},"malaria\nparasite")," genes have a single exon, but the number is much lower\nin other species I've looked at."),(0,r.kt)("h3",{id:"visualising"},"Visualising"),(0,r.kt)("p",null,"Visualisation (plotting) is beyond the scope of this tutorial - but you probably know how to do it already. It would be\ninteresting to visualise some of these statistics, for example"),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},"histogram #transcripts, gene length, # of exons in each species"),(0,r.kt)("li",{parentName:"ul"},"plot the number of exons (or number of transcripts) versus gene length by species")),(0,r.kt)("p",null,"and so on. The sky is the limit in terms of what you could investigate here - every gene has its own story. We'll talk\nthrough your findings in the discussion session."),(0,r.kt)("p",null,(0,r.kt)("strong",{parentName:"p"},"Note.")," Plotting is one thing you can't do in SQL of course - load the data into python or R for this."),(0,r.kt)("pre",null,(0,r.kt)("code",{parentName:"pre"},'library( RSQLite )\ndb = dbConnect( dbDriver( "SQLite" ), "genes.sqlite" )\ndata = dbGetQuery( db, "SELECT * FROM gene_summary_view" )\n')),(0,r.kt)("h3",{id:"getting-sequence-lengths"},"Getting sequence lengths"),(0,r.kt)("p",null,"To go further than this and compute the amount of genome that is covered by genes, we will need to compute genome lengths.\nThe ",(0,r.kt)("a",{parentName:"p",href:"/whg-training-resources/programming/programming_with_gene_annotations/Getting_sequence_lengths"},"next page")," shows one way to do this."))}h.isMDXComponent=!0},2350:(e,t,n)=>{n.d(t,{Z:()=>a});const a=n.p+"assets/files/solutions-ec50e4e517140870f43279535b47e66f.py"},5397:(e,t,n)=>{n.d(t,{Z:()=>a});const a=n.p+"assets/files/solutions-3795ac42e16132bd667f971c8e746408.sql"}}]);