"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[9883],{3905:(e,n,t)=>{t.d(n,{Zo:()=>l,kt:()=>u});var r=t(7294);function a(e,n,t){return n in e?Object.defineProperty(e,n,{value:t,enumerable:!0,configurable:!0,writable:!0}):e[n]=t,e}function o(e,n){var t=Object.keys(e);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);n&&(r=r.filter((function(n){return Object.getOwnPropertyDescriptor(e,n).enumerable}))),t.push.apply(t,r)}return t}function i(e){for(var n=1;n<arguments.length;n++){var t=null!=arguments[n]?arguments[n]:{};n%2?o(Object(t),!0).forEach((function(n){a(e,n,t[n])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(t)):o(Object(t)).forEach((function(n){Object.defineProperty(e,n,Object.getOwnPropertyDescriptor(t,n))}))}return e}function s(e,n){if(null==e)return{};var t,r,a=function(e,n){if(null==e)return{};var t,r,a={},o=Object.keys(e);for(r=0;r<o.length;r++)t=o[r],n.indexOf(t)>=0||(a[t]=e[t]);return a}(e,n);if(Object.getOwnPropertySymbols){var o=Object.getOwnPropertySymbols(e);for(r=0;r<o.length;r++)t=o[r],n.indexOf(t)>=0||Object.prototype.propertyIsEnumerable.call(e,t)&&(a[t]=e[t])}return a}var g=r.createContext({}),c=function(e){var n=r.useContext(g),t=n;return e&&(t="function"==typeof e?e(n):i(i({},n),e)),t},l=function(e){var n=c(e.components);return r.createElement(g.Provider,{value:n},e.children)},m={inlineCode:"code",wrapper:function(e){var n=e.children;return r.createElement(r.Fragment,{},n)}},p=r.forwardRef((function(e,n){var t=e.components,a=e.mdxType,o=e.originalType,g=e.parentName,l=s(e,["components","mdxType","originalType","parentName"]),p=c(t),u=a,h=p["".concat(g,".").concat(u)]||p[u]||m[u]||o;return t?r.createElement(h,i(i({ref:n},l),{},{components:t})):r.createElement(h,i({ref:n},l))}));function u(e,n){var t=arguments,a=n&&n.mdxType;if("string"==typeof e||a){var o=t.length,i=new Array(o);i[0]=p;var s={};for(var g in n)hasOwnProperty.call(n,g)&&(s[g]=n[g]);s.originalType=e,s.mdxType="string"==typeof e?e:a,i[1]=s;for(var c=2;c<o;c++)i[c]=t[c];return r.createElement.apply(null,i)}return r.createElement.apply(null,t)}p.displayName="MDXCreateElement"},6840:(e,n,t)=>{t.r(n),t.d(n,{assets:()=>g,contentTitle:()=>i,default:()=>m,frontMatter:()=>o,metadata:()=>s,toc:()=>c});var r=t(7462),a=(t(7294),t(3905));const o={sidebar_position:10},i="Getting sequence lengths",s={unversionedId:"programming/programming_with_gene_annotations/Getting_sequence_lengths",id:"programming/programming_with_gene_annotations/Getting_sequence_lengths",title:"Getting sequence lengths",description:"Up to table of contents",source:"@site/docs/programming/programming_with_gene_annotations/Getting_sequence_lengths.md",sourceDirName:"programming/programming_with_gene_annotations",slug:"/programming/programming_with_gene_annotations/Getting_sequence_lengths",permalink:"/whg-training-resources/programming/programming_with_gene_annotations/Getting_sequence_lengths",draft:!1,editUrl:"https://github.com/whg-training/whg-training-resources/edit/main/docs/programming/programming_with_gene_annotations/Getting_sequence_lengths.md",tags:[],version:"current",sidebarPosition:10,frontMatter:{sidebar_position:10},sidebar:"tutorialSidebar",previous:{title:"Extremely big, small and complex genes",permalink:"/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_3"},next:{title:"How much of the genome is in genes?",permalink:"/whg-training-resources/programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes"}},g={},c=[],l={toc:c};function m(e){let{components:n,...o}=e;return(0,a.kt)("wrapper",(0,r.Z)({},l,o,{components:n,mdxType:"MDXLayout"}),(0,a.kt)("h1",{id:"getting-sequence-lengths"},"Getting sequence lengths"),(0,a.kt)("p",null,(0,a.kt)("a",{parentName:"p",href:"/whg-training-resources/programming/programming_with_gene_annotations/"},"Up to table of contents")),(0,a.kt)("p",null,"To answer questions like 'how much of the genome is in genes' we need to know what the length of each genome is.\nHappily, if you look at the ",(0,a.kt)("a",{parentName:"p",href:"http://ftp.ensembl.org/pub/current_gff3/"},"GFF files on Ensembl")," the sequences from the\nrelevant reference are listed in the metadata. For example, for humans they look like this:"),(0,a.kt)("pre",null,(0,a.kt)("code",{parentName:"pre"},"##gff-version 3\n##sequence-region   1 1 248956422\n##sequence-region   10 1 133797422\n##sequence-region   11 1 135086622\n##sequence-region   12 1 133275309\n...\n##sequence-region   7 1 159345973\n##sequence-region   8 1 145138636\n##sequence-region   9 1 138394717\n##sequence-region   MT 1 16569\n##sequence-region   X 1 156040895\n##sequence-region   Y 2752083 56887902\n")),(0,a.kt)("p",null,"The three pieces of data on each row are: the sequence name (first column), the sequence starting position, and the\nending position. (Incidentally, the reason the Y chromosome starts at a base different than 1 is because it omits the\n",(0,a.kt)("a",{parentName:"p",href:"https://www.ncbi.nlm.nih.gov/grc/human?filters=chr:Y#current-regions"},"pseudoautosomal regions"),", which are often\ntreated for analysis purposes as being on the X chromosome)."),(0,a.kt)("p",null,"If you feel like it, please try the following:"),(0,a.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,a.kt)("div",{parentName:"div",className:"admonition-heading"},(0,a.kt)("h5",{parentName:"div"},(0,a.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,a.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,a.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Challenge")),(0,a.kt)("div",{parentName:"div",className:"admonition-content"},(0,a.kt)("p",{parentName:"div"},"Write a function ",(0,a.kt)("inlineCode",{parentName:"p"},"parse_sequences_from_gff_metadata()")," that"),(0,a.kt)("ul",{parentName:"div"},(0,a.kt)("li",{parentName:"ul"},"takes the gff filename as argument"),(0,a.kt)("li",{parentName:"ul"},"return a pandas dataframe with the above information.  It should have three columns, the ",(0,a.kt)("inlineCode",{parentName:"li"},"sequence"),", the ",(0,a.kt)("inlineCode",{parentName:"li"},"start"),", and the ",(0,a.kt)("inlineCode",{parentName:"li"},"end"),".")),(0,a.kt)("p",{parentName:"div"},"Use this function in ",(0,a.kt)("inlineCode",{parentName:"p"},"gff_to_sqlite.py")," so that it records sequence lengths into the output\ndatabase as well."))),(0,a.kt)("p",null,"Here's a test case for your new function:"),(0,a.kt)("pre",null,(0,a.kt)("code",{parentName:"pre"},"import unittest\n\nclass TestGff(unittest.TestCase):\n    def test_parse_sequences_from_gff_metadata():\n       import io\n       test_data = \"\"\"##gff-version 3\n##sequence-region   1 1 248956422\n##sequence-region   10 1 133797422\n1       GRCh38  chromosome      1       248956422       .       .       .       ID=chromosome:1;Alias=CM000663.2,chr1,NC_000001.11\n\"\"\"\n\n       sequences = parse_sequences_from_gff_metadata( io.StringIO( test_data ))\n\n       self.assertEqual( sequences['sequence'][0], '1' )\n       self.assertEqual( sequences['sequence'][0], '10' )\n       self.assertEqual( sequences['start'][1], 1 )\n       self.assertEqual( sequences['start'][0], 1 )\n       self.assertEqual( sequences['end'][0], 248956422 )\n       self.assertEqual( sequences['end'][1], 133797422 )\n")),(0,a.kt)("p",null,"If you want some hints or just don't fancy writing this piece of sequence parsing code - feel free\nto look at ",(0,a.kt)("a",{parentName:"p",href:"https://github.com/whg-training/whg-training-resources/blob/main/docs/programming/programming_with_gene_annotations/solutions/part3/gff/sequences.py"},"my solution"),"\n(or use a ",(0,a.kt)("a",{target:"_blank",href:t(3252).Z},"direct link"),")."),(0,a.kt)("p",null,"We are now ready to ",(0,a.kt)("a",{parentName:"p",href:"/whg-training-resources/programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes"},"figure out how much of the genome is in genes"),"."))}m.isMDXComponent=!0},3252:(e,n,t)=>{t.d(n,{Z:()=>r});const r=t.p+"assets/files/sequences-5fdf52df6d57081e3cb7da4e7117458a.py"}}]);