"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[4128],{3905:(e,t,n)=>{n.d(t,{Zo:()=>d,kt:()=>u});var a=n(7294);function i(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function o(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var a=Object.getOwnPropertySymbols(e);t&&(a=a.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,a)}return n}function s(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?o(Object(n),!0).forEach((function(t){i(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):o(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function l(e,t){if(null==e)return{};var n,a,i=function(e,t){if(null==e)return{};var n,a,i={},o=Object.keys(e);for(a=0;a<o.length;a++)n=o[a],t.indexOf(n)>=0||(i[n]=e[n]);return i}(e,t);if(Object.getOwnPropertySymbols){var o=Object.getOwnPropertySymbols(e);for(a=0;a<o.length;a++)n=o[a],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(i[n]=e[n])}return i}var r=a.createContext({}),p=function(e){var t=a.useContext(r),n=t;return e&&(n="function"==typeof e?e(t):s(s({},t),e)),n},d=function(e){var t=p(e.components);return a.createElement(r.Provider,{value:t},e.children)},c={inlineCode:"code",wrapper:function(e){var t=e.children;return a.createElement(a.Fragment,{},t)}},m=a.forwardRef((function(e,t){var n=e.components,i=e.mdxType,o=e.originalType,r=e.parentName,d=l(e,["components","mdxType","originalType","parentName"]),m=p(n),u=i,h=m["".concat(r,".").concat(u)]||m[u]||c[u]||o;return n?a.createElement(h,s(s({ref:t},d),{},{components:n})):a.createElement(h,s({ref:t},d))}));function u(e,t){var n=arguments,i=t&&t.mdxType;if("string"==typeof e||i){var o=n.length,s=new Array(o);s[0]=m;var l={};for(var r in t)hasOwnProperty.call(t,r)&&(l[r]=t[r]);l.originalType=e,l.mdxType="string"==typeof e?e:i,s[1]=l;for(var p=2;p<o;p++)s[p]=n[p];return a.createElement.apply(null,s)}return a.createElement.apply(null,n)}m.displayName="MDXCreateElement"},315:(e,t,n)=>{n.r(t),n.d(t,{assets:()=>r,contentTitle:()=>s,default:()=>c,frontMatter:()=>o,metadata:()=>l,toc:()=>p});var a=n(7462),i=(n(7294),n(3905));const o={sidebar_position:1},s="A first GWAS analysis",l={unversionedId:"genome_wide_association_studies/genome_wide_association_analysis/gwas_analysis_plink",id:"genome_wide_association_studies/genome_wide_association_analysis/gwas_analysis_plink",title:"A first GWAS analysis",description:"Authors: Luke Jostins, Gavin Band",source:"@site/docs/genome_wide_association_studies/genome_wide_association_analysis/gwas_analysis_plink.md",sourceDirName:"genome_wide_association_studies/genome_wide_association_analysis",slug:"/genome_wide_association_studies/genome_wide_association_analysis/gwas_analysis_plink",permalink:"/whg-training-resources/genome_wide_association_studies/genome_wide_association_analysis/gwas_analysis_plink",draft:!1,editUrl:"https://github.com/whg-training/whg-training-resources/edit/main/docs/genome_wide_association_studies/genome_wide_association_analysis/gwas_analysis_plink.md",tags:[],version:"current",sidebarPosition:1,frontMatter:{sidebar_position:1},sidebar:"tutorialSidebar",previous:{title:"Genome-wide association study practical",permalink:"/whg-training-resources/genome_wide_association_studies/genome_wide_association_analysis/"},next:{title:"GWAS quality control",permalink:"/whg-training-resources/genome_wide_association_studies/genome_wide_association_analysis/gwas_quality_control"}},r={},p=[{value:"Introduction",id:"introduction",level:2},{value:"Getting set up",id:"getting-set-up",level:2},{value:"Moving to the right folder",id:"moving-to-the-right-folder",level:3},{value:"Downloading the data",id:"downloading-the-data",level:3},{value:"Downloading plink",id:"downloading-plink",level:3},{value:"Setting up R",id:"setting-up-r",level:3},{value:"Getting started",id:"getting-started",level:2},{value:"Warm-up: computing allele frequencies",id:"warm-up-computing-allele-frequencies",level:3},{value:"Running a simple GWAS",id:"running-a-simple-gwas",level:3},{value:"Forest plotting",id:"forest-plotting",level:2},{value:"Renaming output files",id:"renaming-output-files",level:2},{value:"Reference Information",id:"reference-information",level:2}],d={toc:p};function c(e){let{components:t,...n}=e;return(0,i.kt)("wrapper",(0,a.Z)({},d,n,{components:t,mdxType:"MDXLayout"}),(0,i.kt)("h1",{id:"a-first-gwas-analysis"},"A first GWAS analysis"),(0,i.kt)("p",null,"Authors: Luke Jostins, Gavin Band"),(0,i.kt)("h2",{id:"introduction"},"Introduction"),(0,i.kt)("p",null,"This practical will introduce you to analyzing Genome Wide Association Study (GWAS) datasets using a program called PLINK, which\nis a freely available GWAS analysis toolkit. PLINK has many functions including those related to data organization, formatting,\nquality control, association testing, population stratification and much more. Details about PLINK and its documentation are\navailable at the reference section at the end of these practicals."),(0,i.kt)("h2",{id:"getting-set-up"},"Getting set up"),(0,i.kt)("p",null,"In this practical we will run command-line programs to perform the analyses, and R to interpret them. If you haven't installed\nthese already, visit the ",(0,i.kt)("a",{parentName:"p",href:"/prerequisites"},"prerequisites section")," for a guide."),(0,i.kt)("p",null,"To get set up, you need"),(0,i.kt)("ul",null,(0,i.kt)("li",{parentName:"ul"},"the datasets from the practical folder."),(0,i.kt)("li",{parentName:"ul"},"the command-line programs we'll use")),(0,i.kt)("h3",{id:"moving-to-the-right-folder"},"Moving to the right folder"),(0,i.kt)("p",null,"Before starting please check your terminal is in the right folder.  A good choice is your home folder:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"cd ~\n")),(0,i.kt)("p",null,(0,i.kt)("strong",{parentName:"p"},"However")," if you are using Ubuntu for Windows, you should instead change to your main Desktop folder (this is so that the\nresults work well with RStudio):"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"cd /mnt/c/Users/<your user name>/Desktop/\n")),(0,i.kt)("h3",{id:"downloading-the-data"},"Downloading the data"),(0,i.kt)("p",null,"To get the data, download and extract the tarball from\n",(0,i.kt)("a",{parentName:"p",href:"https://www.well.ox.ac.uk/~gav/projects/whg-training-resources/data/genome_wide_association_studies/genome_wide_association_analysis/"},"this folder"),":"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"curl -O https://www.well.ox.ac.uk/~gav/projects/whg-training-resources/data/genome_wide_association_studies/genome_wide_association_analysis/gwas_practical.tgz\n")),(0,i.kt)("p",null,"This might take a minute or two.  Once the download has completed, extract the tarball and cd into the newly-created directory:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"tar -xzf gwas_practical.tgz\ncd gwas_practical\n")),(0,i.kt)("p",null,"If you explore this directory, you should see there are a number of files and a ",(0,i.kt)("inlineCode",{parentName:"p"},"resources")," and ",(0,i.kt)("inlineCode",{parentName:"p"},"scripts")," folder:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"$ ls\nchr19-example.pca      chr19-example.vcf.gz  resources  snp-example.samples\nchr19-example.samples  plink_versions        scripts    snp-example.vcf\n")),(0,i.kt)("h3",{id:"downloading-plink"},"Downloading plink"),(0,i.kt)("p",null,"Next let's get the plink program we need.  There are two ways to get this:"),(0,i.kt)("ul",null,(0,i.kt)("li",{parentName:"ul"},"The ",(0,i.kt)("inlineCode",{parentName:"li"},"plink_versions/")," folder contains Mac OS and Linux versions of the plink executable. To get that, move the appopriate\nversion into this folder:")),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"mv plink_versions/plink_linux/plink ./\n")),(0,i.kt)("p",null,"or"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"mv plink_versions/plink_mac/plink ./\n")),(0,i.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Note")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("p",{parentName:"div"},"If you are using Ubuntu for Windows, the appropriate version is the 'Linux 64 bit' version."))),(0,i.kt)("p",null,"Alternatively, you can download the appropriate version ",(0,i.kt)("a",{parentName:"p",href:"https://www.cog-genomics.org/plink/"},"from the plink website"),". Extract\nthe archive and move the ",(0,i.kt)("inlineCode",{parentName:"p"},"plink")," executable into this folder."),(0,i.kt)("p",null,"When all is set up you should now be able to run plink like this:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"./plink \u2013-version\n")),(0,i.kt)("p",null,"Plink will tell you its version, and when it was last updated.  "),(0,i.kt)("h3",{id:"setting-up-r"},"Setting up R"),(0,i.kt)("p",null,"Make sure your RStudio session is also pointing at the ",(0,i.kt)("inlineCode",{parentName:"p"},"gwas_practical")," folder (use ",(0,i.kt)("inlineCode",{parentName:"p"},"setwd()")," or the ",(0,i.kt)("inlineCode",{parentName:"p"},"Session->Set Working\nDirectory")," menu to do this."),(0,i.kt)("h2",{id:"getting-started"},"Getting started"),(0,i.kt)("h3",{id:"warm-up-computing-allele-frequencies"},"Warm-up: computing allele frequencies"),(0,i.kt)("p",null,"Plink is a tool for analysis of genome-wide genotype data. A basic first task is to compute allele\nfrequencies. Let's try that out now by running plink like this:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"./plink --freq  --vcf snp-example.vcf\n")),(0,i.kt)("p",null,"The ",(0,i.kt)("inlineCode",{parentName:"p"},"--freq")," command asks PLINK to calculate allele frequencies. ",(0,i.kt)("inlineCode",{parentName:"p"},"--vcf snp-example.vcf")," tells\nPLINK to read in data from the VCF file ",(0,i.kt)("inlineCode",{parentName:"p"},"snp-example.vcf"),"."),(0,i.kt)("p",null,"If you type ",(0,i.kt)("inlineCode",{parentName:"p"},"ls")," you'll see that PLINK has written two output files in the directory we are working in: 1. a log file called\n",(0,i.kt)("inlineCode",{parentName:"p"},"plink.log")," that summarizes what it has done (this is the same as the output that appeared in the terminal window), and 2.\n",(0,i.kt)("inlineCode",{parentName:"p"},"plink.frq"),", which contains the allele frequency information. "),(0,i.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Questions")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("p",{parentName:"div"},"Use a text editor, ",(0,i.kt)("inlineCode",{parentName:"p"},"less"),", or load them into R to look at all the input and output files."),(0,i.kt)("p",{parentName:"div"},(0,i.kt)("strong",{parentName:"p"},"Q"),". How many SNPs were in this dataset? What are their names? What are their allele frequencies?"),(0,i.kt)("p",{parentName:"div"},(0,i.kt)("strong",{parentName:"p"},"Q"),". How many samples are included in total?"),(0,i.kt)("p",{parentName:"div"},(0,i.kt)("strong",{parentName:"p"},"Advanced Question"),". The column names can be frankly somewhat cryptic. Can you find out what\n",(0,i.kt)("inlineCode",{parentName:"p"},"NCHROBS")," means from the documentation? (You may find you need to look at the documentation for the\noriginal version of plink (",(0,i.kt)("a",{parentName:"p",href:"http://zzz.bwh.harvard.edu/plink/summary.shtml"},"http://zzz.bwh.harvard.edu/plink/summary.shtml"),"). On the other hand,\nlooking at the documentation on ",(0,i.kt)("a",{parentName:"p",href:"https://www.cog-genomics.org/plink/1.9/"},"https://www.cog-genomics.org/plink/1.9/"),", are there other or better\n",(0,i.kt)("inlineCode",{parentName:"p"},"--freq"),"-like commands you can run? What do they output? Which version do you like?"))),(0,i.kt)("h3",{id:"running-a-simple-gwas"},"Running a simple GWAS"),(0,i.kt)("p",null,"A basic GWAS study begins by running regression (usually linear regression for quantitative traits,\nor logistic regression for case/control traits like the one we are using) between each genetic\nvariant in the genome and the phenotype of interest. We can run a first tiny GWAS by running:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"./plink --vcf snp-example.vcf --logistic --pheno snp-example.samples --allow-no-sex\n")),(0,i.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Note")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("p",{parentName:"div"},"The ",(0,i.kt)("inlineCode",{parentName:"p"},"--allow-no-sex")," option is needed when working with VCF files in plink. This works around plink's default behaviour which\nremoves samples without a sex assignment - VCF files don't record sample sex, so this is necessary."))),(0,i.kt)("p",null,"PLINK has again generated two files in the directory we are working in: ",(0,i.kt)("inlineCode",{parentName:"p"},"plink.log")," and ",(0,i.kt)("inlineCode",{parentName:"p"},"plink.assoc.logistic"),". The log file\nsimply captures the status information that PLINK reports with each run. The other gives the logistic regression output."),(0,i.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Question")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("p",{parentName:"div"},"How many cases and controls are included in this data?"))),(0,i.kt)("p",null,"If you look in the ",(0,i.kt)("inlineCode",{parentName:"p"},"plink.assoc.logistic")," file we will see that this SNP, rs8135996, is associated with our phenotype."),(0,i.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Questions")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("p",{parentName:"div"},(0,i.kt)("strong",{parentName:"p"},"Q"),". What is the p-value of association?  What is the odds ratio?"),(0,i.kt)("p",{parentName:"div"},(0,i.kt)("strong",{parentName:"p"},"Q"),". And what does the odds ratio mean anyway? If this is a disease trait, is it good news to\nhave the 'G' allele? Or to have the 'A' allele? And how bad is it to have the risk allele?"),(0,i.kt)("p",{parentName:"div"},(0,i.kt)("strong",{parentName:"p"},"Advanced Question"),". What does ",(0,i.kt)("inlineCode",{parentName:"p"},"ADD")," mean in the output? Can you figure out how to run a\nnon-additive test? Which mode of inheritance has the greatest evidence?"))),(0,i.kt)("h2",{id:"forest-plotting"},"Forest plotting"),(0,i.kt)("p",null,"If you followed other parts of our training on statistical modelling, we argued that what you want is to summarise the\nlikelihood by its ",(0,i.kt)("em",{parentName:"p"},"maximum likelihood estimate")," and its ",(0,i.kt)("em",{parentName:"p"},"standard error")," - thus approximating the likelihood function by a\ngaussian function. For logistic regression the appropriate values are the ",(0,i.kt)("em",{parentName:"p"},"log odds ratio")," and its standard error."),(0,i.kt)("p",null,"By default plink hasn't given us the standard error so let's compute it now.\nFirst let's get plink to give us the logistic regression estimate - the ",(0,i.kt)("em",{parentName:"p"},"log odds ratio")," instead of the odds ratio.\nReading the ",(0,i.kt)("a",{parentName:"p",href:"https://www.cog-genomics.org/plink/1.9/assoc#linear"},"documentation for\n--logistic"),' you\'ll see there is an option called "beta" to the --logistic\ncommand that does this.  '),(0,i.kt)("blockquote",null,(0,i.kt)("p",{parentName:"blockquote"},"\"For logistic regressions, the 'beta' modifier causes regression coefficients instead of odds ratios to be reported.\"")),(0,i.kt)("p",null,"Finally there is also a ",(0,i.kt)("inlineCode",{parentName:"p"},"--ci")," option that computes a confidence interval.  Let's apply those now:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"./plink --vcf snp-example.vcf --logistic beta --ci 0.95 --pheno snp-example.samples --allow-no-sex\n")),(0,i.kt)("p",null,"Load this output file into R now:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre",className:"language-R"},'# in R\nX = read.table( "plink.assoc.logistic", hea=T, as.is=T )\n')),(0,i.kt)("p",null,"If you look at the result, you'll see it has output a ",(0,i.kt)("inlineCode",{parentName:"p"},"BETA")," column instad of the original ",(0,i.kt)("inlineCode",{parentName:"p"},"OR"),", and also has ",(0,i.kt)("inlineCode",{parentName:"p"},"SE")," (standard\nerror) and confidence interval columns.  Let's plot that now:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre",className:"language-R"},'plot.betas <- function( betas, ses ) {\n    plot(\n        betas, 1:length(betas),\n        pch = 19, # make filled dots\n        main = "rs8135996 forest plot",\n        xlab = "log odds ratio",\n        ylab = "",\n        xlim = c( -0.5, 0 ),\n        yaxt = \'n\' # Turn off y axis ticks\n    )\n    segments(\n        x0 = betas - 1.96 * ses,\n        x1 = betas + 1.96 * ses,\n        y0 = 1, y1 = 1,\n        col = \'grey\'\n    )\n    abline( v = 0 ) # draw a solid black line at 0\n    grid()\n}\n\nplot.betas( X$BETA, X$SE )\n')),(0,i.kt)("p",null,"Congratulations! This is your first GWAS forest plot (albeit with only one row)."),(0,i.kt)("h2",{id:"renaming-output-files"},"Renaming output files"),(0,i.kt)("p",null,"The final step in this introduction is to learn to rename PLINK\u2019s output files, since we\u2019ll be generating lots of them in the\npracticals.  The ",(0,i.kt)("inlineCode",{parentName:"p"},"--out")," option is used to rename the output:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre",className:"language-sh"},"./plink \\\n--vcf snp-example.vcf \\\n--logistic beta \\\n--beta \\\n--ci 0.95 \\\n--pheno snp-example.samples \\\n--allow-no-sex \\\n--out getting-started\n")),(0,i.kt)("p",null,"Now the output files will be named getting-started.log and getting-started.assoc.logistic. This is the basic pattern to working\nwith PLINK: specifying input files and analyses, along with an output name to save results."),(0,i.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,i.kt)("div",{parentName:"div",className:"admonition-heading"},(0,i.kt)("h5",{parentName:"div"},(0,i.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,i.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,i.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Note")),(0,i.kt)("div",{parentName:"div",className:"admonition-content"},(0,i.kt)("p",{parentName:"div"},"The backslashes in the above command are shell line continuation characters - they let us put the command on multiple lines.\nSome of these commands will get pretty long, so it's useful to put them on multiple lines like this!"))),(0,i.kt)("p",null,"With this background in using plink - it's time to go on to ",(0,i.kt)("a",{parentName:"p",href:"/whg-training-resources/genome_wide_association_studies/genome_wide_association_analysis/gwas_quality_control"},"analysing a simulated GWAS study"),"."),(0,i.kt)("h2",{id:"reference-information"},"Reference Information"),(0,i.kt)("p",null,"You can download PLINK, and find more information at the website:"),(0,i.kt)("p",null,(0,i.kt)("a",{parentName:"p",href:"https://www.cog-genomics.org/plink2"},"https://www.cog-genomics.org/plink2")),(0,i.kt)("p",null,"There is detailed documentation about all the options available, file formats and examples of commands. "),(0,i.kt)("p",null,"A detailed tutorial (similar to work we have done here) is available at:"),(0,i.kt)("p",null,(0,i.kt)("a",{parentName:"p",href:"http://pngu.mgh.harvard.edu/~purcell/plink/tutorial.shtml"},"http://pngu.mgh.harvard.edu/~purcell/plink/tutorial.shtml")))}c.isMDXComponent=!0}}]);