"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[53],{1109:i=>{i.exports=JSON.parse('{"pluginId":"default","version":"current","label":"Next","banner":null,"badge":false,"className":"docs-version-current","isLast":true,"docsSidebars":{"tutorialSidebar":[{"type":"link","label":"Overview of the tutorials","href":"/whg-training-resources/overview","docId":"overview"},{"type":"category","label":"Getting your environment setup","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Setting up a UNIX terminal","href":"/whg-training-resources/prerequisites/UNIX","docId":"prerequisites/UNIX"},{"type":"link","label":"Setting up Conda","href":"/whg-training-resources/prerequisites/CONDA","docId":"prerequisites/CONDA"},{"type":"link","label":"R and RStudio","href":"/whg-training-resources/prerequisites/R","docId":"prerequisites/R"},{"type":"link","label":"Getting a text editor","href":"/whg-training-resources/prerequisites/editor","docId":"prerequisites/editor"},{"type":"link","label":"The R `tidyverse`","href":"/whg-training-resources/prerequisites/tidyverse","docId":"prerequisites/tidyverse"},{"type":"link","label":"Python","href":"/whg-training-resources/prerequisites/python","docId":"prerequisites/python"},{"type":"link","label":"Installing pandas","href":"/whg-training-resources/prerequisites/pandas","docId":"prerequisites/pandas"},{"type":"link","label":"Julia","href":"/whg-training-resources/prerequisites/julia","docId":"prerequisites/julia"},{"type":"link","label":"Using R, python and julia with JupyterLab","href":"/whg-training-resources/prerequisites/Jupyterlab","docId":"prerequisites/Jupyterlab"},{"type":"category","label":"Other languages","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"C++","href":"/whg-training-resources/prerequisites/Other_languages/C++","docId":"prerequisites/Other_languages/C++"},{"type":"link","label":"Javascript","href":"/whg-training-resources/prerequisites/Other_languages/javascript","docId":"prerequisites/Other_languages/javascript"}],"href":"/whg-training-resources/prerequisites/Other_languages/"}],"href":"/whg-training-resources/prerequisites/"},{"type":"category","label":"Bioinformatics","collapsible":true,"collapsed":true,"items":[{"type":"category","label":"Exploring gene annotations on the command-line","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Gene annotations tutorial","href":"/whg-training-resources/bioinformatics/exploring_gene_annotations_in_bash/exploring_genes_in_bash","docId":"bioinformatics/exploring_gene_annotations_in_bash/exploring_genes_in_bash"}],"href":"/whg-training-resources/bioinformatics/exploring_gene_annotations_in_bash/"},{"type":"category","label":"Quick tips and tricks","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"List contents of an FTP site","href":"/whg-training-resources/bioinformatics/tips_and_tricks/recursive_ftp","docId":"bioinformatics/tips_and_tricks/recursive_ftp"}],"href":"/whg-training-resources/bioinformatics/tips_and_tricks/"}],"href":"/whg-training-resources/bioinformatics/"},{"type":"category","label":"Programming","collapsible":true,"collapsed":true,"items":[{"type":"category","label":"Introduction to R","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"R: Introduction to Basic Features","href":"/whg-training-resources/programming/introduction_to_R/R_introduction_to_basic_features","docId":"programming/introduction_to_R/R_introduction_to_basic_features"},{"type":"link","label":"Using R to load and manipulate data","href":"/whg-training-resources/programming/introduction_to_R/R_basic_features_practical","docId":"programming/introduction_to_R/R_basic_features_practical"},{"type":"link","label":"Introduction to R","href":"/whg-training-resources/programming/introduction_to_R/R_introduction_to_basic_features_slides","docId":"programming/introduction_to_R/R_introduction_to_basic_features_slides"}],"href":"/whg-training-resources/programming/introduction_to_R/"},{"type":"category","label":"Programming with gene annotations","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Introduction","href":"/whg-training-resources/programming/programming_with_gene_annotations/Introduction","docId":"programming/programming_with_gene_annotations/Introduction"},{"type":"link","label":"What gene annotation data looks like","href":"/whg-training-resources/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like","docId":"programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like"},{"type":"link","label":"Processing GFF files","href":"/whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code","docId":"programming/programming_with_gene_annotations/Getting_started_writing_some_code"},{"type":"link","label":"Making a module","href":"/whg-training-resources/programming/programming_with_gene_annotations/making_a_module","docId":"programming/programming_with_gene_annotations/making_a_module"},{"type":"link","label":"An aside on refactoring","href":"/whg-training-resources/programming/programming_with_gene_annotations/Refactoring_makes_code_better","docId":"programming/programming_with_gene_annotations/Refactoring_makes_code_better"},{"type":"link","label":"Writing a useful conversion program.","href":"/whg-training-resources/programming/programming_with_gene_annotations/Converting_gff_to_sqlite","docId":"programming/programming_with_gene_annotations/Converting_gff_to_sqlite"},{"type":"link","label":"Investigating human genes","href":"/whg-training-resources/programming/programming_with_gene_annotations/exploring_gencode","docId":"programming/programming_with_gene_annotations/exploring_gencode"},{"type":"link","label":"Counting genes I","href":"/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_1","docId":"programming/programming_with_gene_annotations/Counting_genes_1"},{"type":"link","label":"Memory issues and how to solve them","href":"/whg-training-resources/programming/programming_with_gene_annotations/Memory_issues_and_how_to_solve_them","docId":"programming/programming_with_gene_annotations/Memory_issues_and_how_to_solve_them"},{"type":"link","label":"Counting genes II: types of protein-coding genes","href":"/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_2","docId":"programming/programming_with_gene_annotations/Counting_genes_2"},{"type":"link","label":"Extremely big, small and complex genes","href":"/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_3","docId":"programming/programming_with_gene_annotations/Counting_genes_3"},{"type":"link","label":"Getting sequence lengths","href":"/whg-training-resources/programming/programming_with_gene_annotations/Getting_sequence_lengths","docId":"programming/programming_with_gene_annotations/Getting_sequence_lengths"},{"type":"link","label":"How much of the genome is in genes?","href":"/whg-training-resources/programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes","docId":"programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes"},{"type":"link","label":"Scaling up for a real analysis","href":"/whg-training-resources/programming/programming_with_gene_annotations/Scaling_up","docId":"programming/programming_with_gene_annotations/Scaling_up"},{"type":"link","label":"What next?","href":"/whg-training-resources/programming/programming_with_gene_annotations/Where_next","docId":"programming/programming_with_gene_annotations/Where_next"},{"type":"link","label":"Closing thoughts","href":"/whg-training-resources/programming/programming_with_gene_annotations/Closing_thoughts","docId":"programming/programming_with_gene_annotations/Closing_thoughts"},{"type":"link","label":"Appendix: a list of Ensembl GFF files.","href":"/whg-training-resources/programming/programming_with_gene_annotations/List_of_ensembl_files","docId":"programming/programming_with_gene_annotations/List_of_ensembl_files"},{"type":"link","label":"Appendix: visualisation","href":"/whg-training-resources/programming/programming_with_gene_annotations/Visualisation","docId":"programming/programming_with_gene_annotations/Visualisation"}],"href":"/whg-training-resources/programming/programming_with_gene_annotations/"}],"href":"/whg-training-resources/programming/"},{"type":"category","label":"Data visualisation","collapsible":true,"collapsed":true,"items":[{"type":"category","label":"Visualising RNA-seq data in R","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Visualising RNA-seq using R","href":"/whg-training-resources/data_visualisation/visualising_rna-seq_data/Data-visualisation-practical_GTEx","docId":"data_visualisation/visualising_rna-seq_data/Data-visualisation-practical_GTEx"}],"href":"/whg-training-resources/data_visualisation/visualising_rna-seq_data/"},{"type":"category","label":"Visualisation with ggplot2","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Data visualisation with tidyverse","href":"/whg-training-resources/data_visualisation/data_visualisation_with_ggplot2/dataviz","docId":"data_visualisation/data_visualisation_with_ggplot2/dataviz"}],"href":"/whg-training-resources/data_visualisation/data_visualisation_with_ggplot2/"}],"href":"/whg-training-resources/data_visualisation/"},{"type":"category","label":"Next-generation sequencing","collapsible":true,"collapsed":true,"items":[{"type":"category","label":"Introduction to next-generation sequencing","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Getting the prerequisites","href":"/whg-training-resources/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Prerequisites","docId":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Prerequisites"},{"type":"link","label":"Practical outline","href":"/whg-training-resources/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Pipeline_outline","docId":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Pipeline_outline"},{"type":"link","label":"What\'s in the fastq files?","href":"/whg-training-resources/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Inspecting_the_fastqs","docId":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Inspecting_the_fastqs"},{"type":"link","label":"Inspecting sequence data quality","href":"/whg-training-resources/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Quality_control","docId":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Quality_control"},{"type":"link","label":"Paired-end sequencing in theory and practice","href":"/whg-training-resources/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Short_read_theory","docId":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Short_read_theory"},{"type":"link","label":"Aligning reads","href":"/whg-training-resources/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Aligning_reads","docId":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Aligning_reads"},{"type":"link","label":"Viewing alignments","href":"/whg-training-resources/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Viewing_alignments","docId":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Viewing_alignments"},{"type":"link","label":"Challenge questions","href":"/whg-training-resources/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Challenge_questions","docId":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Challenge_questions"},{"type":"link","label":"Appendix - how accurate is Illumina sequencing?","href":"/whg-training-resources/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/De_novo_error_rate_estimation","docId":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/De_novo_error_rate_estimation"},{"type":"link","label":"Appendix - understanding sequence duplication levels","href":"/whg-training-resources/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/duplicates","docId":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/duplicates"},{"type":"link","label":"Appendix - read trimming","href":"/whg-training-resources/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Read_trimming","docId":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Read_trimming"},{"type":"link","label":"Useful links","href":"/whg-training-resources/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Useful_links","docId":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Useful_links"}],"href":"/whg-training-resources/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/"},{"type":"category","label":"Building an NGS data processing pipeline","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Introduction","href":"/whg-training-resources/next_generation_sequencing/building_an_ngs_pipeline/Introduction","docId":"next_generation_sequencing/building_an_ngs_pipeline/Introduction"},{"type":"link","label":"The pipeline","href":"/whg-training-resources/next_generation_sequencing/building_an_ngs_pipeline/pipeline","docId":"next_generation_sequencing/building_an_ngs_pipeline/pipeline"},{"type":"link","label":"Tips and tricks","href":"/whg-training-resources/next_generation_sequencing/building_an_ngs_pipeline/tips_and_tricks","docId":"next_generation_sequencing/building_an_ngs_pipeline/tips_and_tricks"}],"href":"/whg-training-resources/next_generation_sequencing/building_an_ngs_pipeline/"},{"type":"category","label":"Variant calling, phasing and imputation","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Getting the data","href":"/whg-training-resources/next_generation_sequencing/variant_calling_and_imputation/Prerequisites","docId":"next_generation_sequencing/variant_calling_and_imputation/Prerequisites"},{"type":"link","label":"Variant calling with bcftools","href":"/whg-training-resources/next_generation_sequencing/variant_calling_and_imputation/Variant_calling","docId":"next_generation_sequencing/variant_calling_and_imputation/Variant_calling"},{"type":"link","label":"Variant quality control","href":"/whg-training-resources/next_generation_sequencing/variant_calling_and_imputation/Variant_quality_control","docId":"next_generation_sequencing/variant_calling_and_imputation/Variant_quality_control"},{"type":"link","label":"Phasing the variant calls","href":"/whg-training-resources/next_generation_sequencing/variant_calling_and_imputation/Phasing","docId":"next_generation_sequencing/variant_calling_and_imputation/Phasing"},{"type":"link","label":"Imputing a set of microarray data","href":"/whg-training-resources/next_generation_sequencing/variant_calling_and_imputation/Imputation","docId":"next_generation_sequencing/variant_calling_and_imputation/Imputation"},{"type":"link","label":"Challenge questions on imputation","href":"/whg-training-resources/next_generation_sequencing/variant_calling_and_imputation/Challenge_questions","docId":"next_generation_sequencing/variant_calling_and_imputation/Challenge_questions"}],"href":"/whg-training-resources/next_generation_sequencing/variant_calling_and_imputation/"},{"type":"category","label":"Visualising alignments","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Exploring Reads in IGV","href":"/whg-training-resources/next_generation_sequencing/IGV/Exploring_reads_in_IGV","docId":"next_generation_sequencing/IGV/Exploring_reads_in_IGV"}],"href":"/whg-training-resources/next_generation_sequencing/IGV/"}],"href":"/whg-training-resources/next_generation_sequencing/"},{"type":"category","label":"Statistical modelling","collapsible":true,"collapsed":true,"items":[{"type":"category","label":"Statistical modelling I - regression","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Getting started: linear regression","href":"/whg-training-resources/statistical_modelling/regression_modelling/linear_regression_1","docId":"statistical_modelling/regression_modelling/linear_regression_1"},{"type":"link","label":"Logistic regression","href":"/whg-training-resources/statistical_modelling/regression_modelling/logistic_regression_1","docId":"statistical_modelling/regression_modelling/logistic_regression_1"},{"type":"link","label":"Regression as a parameteric statistical model","href":"/whg-training-resources/statistical_modelling/regression_modelling/regression_as_a_model","docId":"statistical_modelling/regression_modelling/regression_as_a_model"},{"type":"link","label":"Checking estimates using simulation","href":"/whg-training-resources/statistical_modelling/regression_modelling/simulating","docId":"statistical_modelling/regression_modelling/simulating"},{"type":"link","label":"Causal thinking","href":"/whg-training-resources/statistical_modelling/regression_modelling/causal_graphs","docId":"statistical_modelling/regression_modelling/causal_graphs"},{"type":"link","label":"Linear regression cheatsheet","href":"/whg-training-resources/statistical_modelling/regression_modelling/linear_regression_cheatsheet","docId":"statistical_modelling/regression_modelling/linear_regression_cheatsheet"}],"href":"/whg-training-resources/statistical_modelling/regression_modelling/"},{"type":"category","label":"Glycophorin CNV calling tutorial","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Warmup: examining read coverage in a region","href":"/whg-training-resources/statistical_modelling/Hidden_markov_models/glycophorin_cnv_warmup","docId":"statistical_modelling/Hidden_markov_models/glycophorin_cnv_warmup"},{"type":"link","label":"Modelling CNVs using a Hidden Markov Model","href":"/whg-training-resources/statistical_modelling/Hidden_markov_models/glycophorin_cnv_hmm","docId":"statistical_modelling/Hidden_markov_models/glycophorin_cnv_hmm"}],"href":"/whg-training-resources/statistical_modelling/Hidden_markov_models/"},{"type":"category","label":"notes","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Asymptotics and  P-values","href":"/whg-training-resources/statistical_modelling/notes/computing_pvalues","docId":"statistical_modelling/notes/computing_pvalues"}]}],"href":"/whg-training-resources/statistical_modelling/"},{"type":"category","label":"Population genetics","collapsible":true,"collapsed":true,"items":[{"type":"category","label":"Simulating populations","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Genetic drift","href":"/whg-training-resources/population_genetics/simulation/implementing_a_Wright-Fisher_model","docId":"population_genetics/simulation/implementing_a_Wright-Fisher_model"},{"type":"link","label":"Some population-genetic statistics","href":"/whg-training-resources/population_genetics/simulation/statistics","docId":"population_genetics/simulation/statistics"}],"href":"/whg-training-resources/population_genetics/simulation/"},{"type":"category","label":"Principal components analysis","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Introduction","href":"/whg-training-resources/population_genetics/principal_components_analysis/Introduction","docId":"population_genetics/principal_components_analysis/Introduction"},{"type":"link","label":"Getting set up","href":"/whg-training-resources/population_genetics/principal_components_analysis/getting_setup","docId":"population_genetics/principal_components_analysis/getting_setup"},{"type":"link","label":"Overview of the practical","href":"/whg-training-resources/population_genetics/principal_components_analysis/overview","docId":"population_genetics/principal_components_analysis/overview"},{"type":"link","label":"Removing highly-correlated SNPs","href":"/whg-training-resources/population_genetics/principal_components_analysis/ld_pruning","docId":"population_genetics/principal_components_analysis/ld_pruning"},{"type":"link","label":"Removing closely-related samples","href":"/whg-training-resources/population_genetics/principal_components_analysis/relatedness_pruning","docId":"population_genetics/principal_components_analysis/relatedness_pruning"},{"type":"link","label":"Computing principal components","href":"/whg-training-resources/population_genetics/principal_components_analysis/computing_PCs","docId":"population_genetics/principal_components_analysis/computing_PCs"},{"type":"link","label":"Plotting samples against a global reference panel","href":"/whg-training-resources/population_genetics/principal_components_analysis/global_analysis","docId":"population_genetics/principal_components_analysis/global_analysis"},{"type":"link","label":"An aside on the maths","href":"/whg-training-resources/population_genetics/principal_components_analysis/the_maths","docId":"population_genetics/principal_components_analysis/the_maths"}],"href":"/whg-training-resources/population_genetics/principal_components_analysis/"}],"href":"/whg-training-resources/population_genetics/"},{"type":"category","label":"Genome-wide association analysis","collapsible":true,"collapsed":true,"items":[{"type":"category","label":"Genome-wide association study practical","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"A first GWAS analysis","href":"/whg-training-resources/genome_wide_association_studies/genome_wide_association_analysis/gwas_analysis_plink","docId":"genome_wide_association_studies/genome_wide_association_analysis/gwas_analysis_plink"},{"type":"link","label":"GWAS quality control","href":"/whg-training-resources/genome_wide_association_studies/genome_wide_association_analysis/gwas_quality_control","docId":"genome_wide_association_studies/genome_wide_association_analysis/gwas_quality_control"},{"type":"link","label":"Association Practical","href":"/whg-training-resources/genome_wide_association_studies/genome_wide_association_analysis/gwas_association","docId":"genome_wide_association_studies/genome_wide_association_analysis/gwas_association"},{"type":"link","label":"Controlling for population structure","href":"/whg-training-resources/genome_wide_association_studies/genome_wide_association_analysis/population_structure","docId":"genome_wide_association_studies/genome_wide_association_analysis/population_structure"},{"type":"link","label":"Making a regional association plot","href":"/whg-training-resources/genome_wide_association_studies/genome_wide_association_analysis/hitplots","docId":"genome_wide_association_studies/genome_wide_association_analysis/hitplots"}],"href":"/whg-training-resources/genome_wide_association_studies/genome_wide_association_analysis/"},{"type":"category","label":"Meta-analysis and fine-mapping tutorial","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Introduction","href":"/whg-training-resources/genome_wide_association_studies/meta-analysis_and_fine-mapping/Introduction","docId":"genome_wide_association_studies/meta-analysis_and_fine-mapping/Introduction"},{"type":"link","label":"Meta-analysing two studies.","href":"/whg-training-resources/genome_wide_association_studies/meta-analysis_and_fine-mapping/Meta-analysis","docId":"genome_wide_association_studies/meta-analysis_and_fine-mapping/Meta-analysis"},{"type":"link","label":"Using `FINEMAP` to fine-map associations","href":"/whg-training-resources/genome_wide_association_studies/meta-analysis_and_fine-mapping/Fine-mapping","docId":"genome_wide_association_studies/meta-analysis_and_fine-mapping/Fine-mapping"}],"href":"/whg-training-resources/genome_wide_association_studies/meta-analysis_and_fine-mapping/"}],"href":"/whg-training-resources/genome_wide_association_studies/"},{"type":"link","label":"LICENSE","href":"/whg-training-resources/LICENSE","docId":"LICENSE"},{"type":"link","label":"contributors","href":"/whg-training-resources/contributors","docId":"contributors"}]},"docs":{"bioinformatics/exploring_gene_annotations_in_bash/exploring_genes_in_bash":{"id":"bioinformatics/exploring_gene_annotations_in_bash/exploring_genes_in_bash","title":"Gene annotations tutorial","description":"Author: Gavin Band","sidebar":"tutorialSidebar"},"bioinformatics/exploring_gene_annotations_in_bash/README":{"id":"bioinformatics/exploring_gene_annotations_in_bash/README","title":"Exploring gene annotations on the command-line","description":"Welcome! To get started, go here.","sidebar":"tutorialSidebar"},"bioinformatics/README":{"id":"bioinformatics/README","title":"Bioinformatics","description":"Welcome! Click below to visit the tutorials:","sidebar":"tutorialSidebar"},"bioinformatics/tips_and_tricks/README":{"id":"bioinformatics/tips_and_tricks/README","title":"Quick tips and tricks","description":"Welcome! see the table below","sidebar":"tutorialSidebar"},"bioinformatics/tips_and_tricks/recursive_ftp":{"id":"bioinformatics/tips_and_tricks/recursive_ftp","title":"List contents of an FTP site","description":"The tool lftp can be used to quickly list contents of an FTP site. lftp calls itself a \'sophisticated file","sidebar":"tutorialSidebar"},"Clinical variant interpretation/README":{"id":"Clinical variant interpretation/README","title":"Clinical variant interpretation","description":"Contributors: Nicky Whiffin"},"contributors":{"id":"contributors","title":"contributors","description":"* Azim Ansari","sidebar":"tutorialSidebar"},"data_visualisation/data_visualisation_with_ggplot2/dataviz":{"id":"data_visualisation/data_visualisation_with_ggplot2/dataviz","title":"Data visualisation with tidyverse","description":"In this tutorial we will explore data visualisation using the R packages","sidebar":"tutorialSidebar"},"data_visualisation/data_visualisation_with_ggplot2/README":{"id":"data_visualisation/data_visualisation_with_ggplot2/README","title":"Visualisation with ggplot2","description":"Welcome!  To get started, go here.","sidebar":"tutorialSidebar"},"data_visualisation/README":{"id":"data_visualisation/README","title":"Data visualisation","description":"Welcome!  To get started, choose your tutorial from the list below:","sidebar":"tutorialSidebar"},"data_visualisation/visualising_rna-seq_data/Data-visualisation-practical_GTEx":{"id":"data_visualisation/visualising_rna-seq_data/Data-visualisation-practical_GTEx","title":"Visualising RNA-seq using R","description":"This tutorial will show you some of the elementary skills you need to begin visualising RNA-seq","sidebar":"tutorialSidebar"},"data_visualisation/visualising_rna-seq_data/README":{"id":"data_visualisation/visualising_rna-seq_data/README","title":"Visualising RNA-seq data in R","description":"Welcome!  To get started, go here.","sidebar":"tutorialSidebar"},"genome_wide_association_studies/genome_wide_association_analysis/gwas_analysis_plink":{"id":"genome_wide_association_studies/genome_wide_association_analysis/gwas_analysis_plink","title":"A first GWAS analysis","description":"Authors: Luke Jostins, Gavin Band","sidebar":"tutorialSidebar"},"genome_wide_association_studies/genome_wide_association_analysis/gwas_association":{"id":"genome_wide_association_studies/genome_wide_association_analysis/gwas_association","title":"Association Practical","description":"Now we will test for association between the SNPs in our dataset and the disease outcome.","sidebar":"tutorialSidebar"},"genome_wide_association_studies/genome_wide_association_analysis/gwas_quality_control":{"id":"genome_wide_association_studies/genome_wide_association_analysis/gwas_quality_control","title":"GWAS quality control","description":"We will now work with a set of data files containing many SNPs from chromosome 19 genotyped on controls and cases. Data from a","sidebar":"tutorialSidebar"},"genome_wide_association_studies/genome_wide_association_analysis/hitplots":{"id":"genome_wide_association_studies/genome_wide_association_analysis/hitplots","title":"Making a regional association plot","description":"In this part of the practical, we will hone our R plotting skills to make beautiful regional association plots (\u2018hitplots\u2019) of","sidebar":"tutorialSidebar"},"genome_wide_association_studies/genome_wide_association_analysis/population_structure":{"id":"genome_wide_association_studies/genome_wide_association_analysis/population_structure","title":"Controlling for population structure","description":"We have seen how applying appropriate quality control filters to our data eliminated many false positives, but a systematic","sidebar":"tutorialSidebar"},"genome_wide_association_studies/genome_wide_association_analysis/README":{"id":"genome_wide_association_studies/genome_wide_association_analysis/README","title":"Genome-wide association study practical","description":"Welcome! In this practical we will run a basic genome-wide association analysis using some","sidebar":"tutorialSidebar"},"genome_wide_association_studies/meta-analysis_and_fine-mapping/Fine-mapping":{"id":"genome_wide_association_studies/meta-analysis_and_fine-mapping/Fine-mapping","title":"Using `FINEMAP` to fine-map associations","description":"Up to the table of contents / Back to the meta-analysis section","sidebar":"tutorialSidebar"},"genome_wide_association_studies/meta-analysis_and_fine-mapping/Introduction":{"id":"genome_wide_association_studies/meta-analysis_and_fine-mapping/Introduction","title":"Introduction","description":"Up to the table of contents / Forward to the meta-analysis section","sidebar":"tutorialSidebar"},"genome_wide_association_studies/meta-analysis_and_fine-mapping/Meta-analysis":{"id":"genome_wide_association_studies/meta-analysis_and_fine-mapping/Meta-analysis","title":"Meta-analysing two studies.","description":"Up to the table of contents / Back to the Introduction / Forward to the fine-mapping section","sidebar":"tutorialSidebar"},"genome_wide_association_studies/meta-analysis_and_fine-mapping/README":{"id":"genome_wide_association_studies/meta-analysis_and_fine-mapping/README","title":"Meta-analysis and fine-mapping tutorial","description":"Contributors: Luke Jostins, Gavin Band","sidebar":"tutorialSidebar"},"genome_wide_association_studies/README":{"id":"genome_wide_association_studies/README","title":"Genome-wide association analysis","description":"In this folder are:","sidebar":"tutorialSidebar"},"genome_wide_association_studies/slides/README":{"id":"genome_wide_association_studies/slides/README","title":"README","description":"Slides from these sessions can be found here:"},"LICENSE":{"id":"LICENSE","title":"LICENSE","description":"MIT License","sidebar":"tutorialSidebar"},"next_generation_sequencing/building_an_ngs_pipeline/Introduction":{"id":"next_generation_sequencing/building_an_ngs_pipeline/Introduction","title":"Introduction","description":"In this tutorial you will bring a set of FASTQ files","sidebar":"tutorialSidebar"},"next_generation_sequencing/building_an_ngs_pipeline/pipeline":{"id":"next_generation_sequencing/building_an_ngs_pipeline/pipeline","title":"The pipeline","description":"Overview","sidebar":"tutorialSidebar"},"next_generation_sequencing/building_an_ngs_pipeline/README":{"id":"next_generation_sequencing/building_an_ngs_pipeline/README","title":"Building an NGS data processing pipeline","description":"Welcome!  Go to the introduction to get started.","sidebar":"tutorialSidebar"},"next_generation_sequencing/building_an_ngs_pipeline/tips_and_tricks":{"id":"next_generation_sequencing/building_an_ngs_pipeline/tips_and_tricks","title":"Tips and tricks","description":"Here is some guidance to help you write your pipeline. Click the links to jump to the relevant","sidebar":"tutorialSidebar"},"next_generation_sequencing/Getting started with snakemake/README":{"id":"next_generation_sequencing/Getting started with snakemake/README","title":"Getting started with snakemake","description":"This is a quick snakemake example that uses bwa mem to align a few reads."},"next_generation_sequencing/IGV/Exploring_reads_in_IGV":{"id":"next_generation_sequencing/IGV/Exploring_reads_in_IGV","title":"Exploring Reads in IGV","description":"Contributors: Annie Forster, Gavin Band","sidebar":"tutorialSidebar"},"next_generation_sequencing/IGV/README":{"id":"next_generation_sequencing/IGV/README","title":"Visualising alignments","description":"Welcome! Go the the IGV practical to get started.","sidebar":"tutorialSidebar"},"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Aligning_reads":{"id":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Aligning_reads","title":"Aligning reads","description":"Most analyses of short-read sequence data are based on the basic paradigm made up of two steps:","sidebar":"tutorialSidebar"},"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Challenge_questions":{"id":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Challenge_questions","title":"Challenge questions","description":"We have three challenges.:","sidebar":"tutorialSidebar"},"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/De_novo_error_rate_estimation":{"id":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/De_novo_error_rate_estimation","title":"Appendix - how accurate is Illumina sequencing?","description":"Note. Only do this section if you have plenty of time left, and/or have already read the","sidebar":"tutorialSidebar"},"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/duplicates":{"id":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/duplicates","title":"Appendix - understanding sequence duplication levels","description":"Duplicates arise naturally in sequencing from random fragments that just happen to have","sidebar":"tutorialSidebar"},"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Inspecting_the_fastqs":{"id":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Inspecting_the_fastqs","title":"What\'s in the fastq files?","description":"Using UNIX to inspect the fastqs","sidebar":"tutorialSidebar"},"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Pipeline_outline":{"id":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Pipeline_outline","title":"Practical outline","description":"Overview","sidebar":"tutorialSidebar"},"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Prerequisites":{"id":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Prerequisites","title":"Getting the prerequisites","description":"To run this tutorial you will need two things: some software and some data.  This page will get you set up.","sidebar":"tutorialSidebar"},"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Quality_control":{"id":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Quality_control","title":"Inspecting sequence data quality","description":"There are a set of analyses that can be carried out directly on a FASTQ file to assess the quality","sidebar":"tutorialSidebar"},"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Read_trimming":{"id":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Read_trimming","title":"Appendix - read trimming","description":"Read trimming refers to the process of removing trailing parts of the reads before analysis for various reasons including","sidebar":"tutorialSidebar"},"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/README":{"id":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/README","title":"Introduction to next-generation sequencing","description":"In this tutorial we will demonstrate a basic pipeline for analysing paired-end short-read genomic","sidebar":"tutorialSidebar"},"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Short_read_theory":{"id":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Short_read_theory","title":"Paired-end sequencing in theory and practice","description":"What paired-end sequence data should lool like","sidebar":"tutorialSidebar"},"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Useful_links":{"id":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Useful_links","title":"Useful links","description":"* Illumina\'s page on base quality scores.","sidebar":"tutorialSidebar"},"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Viewing_alignments":{"id":"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Viewing_alignments","title":"Viewing alignments","description":"If you got this far, you should have a file called QG0033-C.bam containing aligned,","sidebar":"tutorialSidebar"},"next_generation_sequencing/README":{"id":"next_generation_sequencing/README","title":"Next-generation sequencing","description":"* Go to the introduction to next-generation sequence (NGS) data analysis tutorial.","sidebar":"tutorialSidebar"},"next_generation_sequencing/variant_calling_and_imputation/Challenge_questions":{"id":"next_generation_sequencing/variant_calling_and_imputation/Challenge_questions","title":"Challenge questions on imputation","description":"Imputation $r^2$","sidebar":"tutorialSidebar"},"next_generation_sequencing/variant_calling_and_imputation/Imputation":{"id":"next_generation_sequencing/variant_calling_and_imputation/Imputation","title":"Imputing a set of microarray data","description":"Ok, so we\'ve called genetic variants, we\'ve [QCd","sidebar":"tutorialSidebar"},"next_generation_sequencing/variant_calling_and_imputation/Phasing":{"id":"next_generation_sequencing/variant_calling_and_imputation/Phasing","title":"Phasing the variant calls","description":"If you got this far you should have a robust, well-QCd set of variants. But we\'re not done! Humans","sidebar":"tutorialSidebar"},"next_generation_sequencing/variant_calling_and_imputation/Prerequisites":{"id":"next_generation_sequencing/variant_calling_and_imputation/Prerequisites","title":"Getting the data","description":"To get the data download and unzip this tarball:","sidebar":"tutorialSidebar"},"next_generation_sequencing/variant_calling_and_imputation/README":{"id":"next_generation_sequencing/variant_calling_and_imputation/README","title":"Variant calling, phasing and imputation","description":"This morning we focussed on quality control, aligning, and inspecting some sequence data for a single sample.","sidebar":"tutorialSidebar"},"next_generation_sequencing/variant_calling_and_imputation/Variant_calling":{"id":"next_generation_sequencing/variant_calling_and_imputation/Variant_calling","title":"Variant calling with bcftools","description":"First let\'s see how to use a simple pipeline to identify genetic variants","sidebar":"tutorialSidebar"},"next_generation_sequencing/variant_calling_and_imputation/Variant_quality_control":{"id":"next_generation_sequencing/variant_calling_and_imputation/Variant_quality_control","title":"Variant quality control","description":"For variant quality control we\'ll start with the output of bcftools call and use the various INFO","sidebar":"tutorialSidebar"},"overview":{"id":"overview","title":"Overview of the tutorials","description":"This folder contains course materials (practicals, notes and datasets) that were developed for the taught statistics component of the","sidebar":"tutorialSidebar"},"population_genetics/principal_components_analysis/computing_PCs":{"id":"population_genetics/principal_components_analysis/computing_PCs","title":"Computing principal components","description":"Up to the table of contents / Back to the relatedness pruning step / Forward to the global analysis step.","sidebar":"tutorialSidebar"},"population_genetics/principal_components_analysis/getting_setup":{"id":"population_genetics/principal_components_analysis/getting_setup","title":"Getting set up","description":"Back to the table of contents - Forward to the overview","sidebar":"tutorialSidebar"},"population_genetics/principal_components_analysis/global_analysis":{"id":"population_genetics/principal_components_analysis/global_analysis","title":"Plotting samples against a global reference panel","description":"Up to the table of contents / Back to the PC computation step","sidebar":"tutorialSidebar"},"population_genetics/principal_components_analysis/Introduction":{"id":"population_genetics/principal_components_analysis/Introduction","title":"Introduction","description":"In this practical we will do principal component analysis (PCA), which is one of the fundamental tools of","sidebar":"tutorialSidebar"},"population_genetics/principal_components_analysis/ld_pruning":{"id":"population_genetics/principal_components_analysis/ld_pruning","title":"Removing highly-correlated SNPs","description":"Up to the table of contents / Back to the previous page / Forward to the next page","sidebar":"tutorialSidebar"},"population_genetics/principal_components_analysis/overview":{"id":"population_genetics/principal_components_analysis/overview","title":"Overview of the practical","description":"Up to the table of contents / Back to the setup page / Forward to the page on LD pruning","sidebar":"tutorialSidebar"},"population_genetics/principal_components_analysis/README":{"id":"population_genetics/principal_components_analysis/README","title":"Principal components analysis","description":"Welcome!  When you\'re ready to start, go to the introduction.","sidebar":"tutorialSidebar"},"population_genetics/principal_components_analysis/relatedness_pruning":{"id":"population_genetics/principal_components_analysis/relatedness_pruning","title":"Removing closely-related samples","description":"Up to the table of contents - Back to the page on LD pruning - Forward to compute the PCs","sidebar":"tutorialSidebar"},"population_genetics/principal_components_analysis/the_maths":{"id":"population_genetics/principal_components_analysis/the_maths","title":"An aside on the maths","description":"The maths of PCA (in the way usually applied) works like this. Suppose $X$ is a big matrix of genotypes, with $L$ rows","sidebar":"tutorialSidebar"},"population_genetics/README":{"id":"population_genetics/README","title":"Population genetics","description":"In this folder are:","sidebar":"tutorialSidebar"},"population_genetics/simulation/implementing_a_Wright-Fisher_model":{"id":"population_genetics/simulation/implementing_a_Wright-Fisher_model","title":"Genetic drift","description":"The Wright-Fisher model is about the simplest population-genetic model there is.  It consists of:","sidebar":"tutorialSidebar"},"population_genetics/simulation/README":{"id":"population_genetics/simulation/README","title":"Simulating populations","description":"In this practical we will implement a simple Wright-Fisher population genetics model, and couple","sidebar":"tutorialSidebar"},"population_genetics/simulation/statistics":{"id":"population_genetics/simulation/statistics","title":"Some population-genetic statistics","description":"If you followed the basic simulation section you\'ll know that random inheritance","sidebar":"tutorialSidebar"},"prerequisites/CONDA":{"id":"prerequisites/CONDA","title":"Setting up Conda","description":"One of the easiest ways to set up your environment that works across platforms is to use conda.","sidebar":"tutorialSidebar"},"prerequisites/editor":{"id":"prerequisites/editor","title":"Getting a text editor","description":"Whatever you do, you\'re going to want a good text editor.  There are many choices -","sidebar":"tutorialSidebar"},"prerequisites/julia":{"id":"prerequisites/julia","title":"Julia","description":"Julia is an interesting language that is pretty good for many data analysis and computational","sidebar":"tutorialSidebar"},"prerequisites/Jupyterlab":{"id":"prerequisites/Jupyterlab","title":"Using R, python and julia with JupyterLab","description":"The plain command-line interpreters for R, python and julia are useful, but for data analysis work you may want","sidebar":"tutorialSidebar"},"prerequisites/Other_languages/C++":{"id":"prerequisites/Other_languages/C++","title":"C++","description":"This page is to be written.","sidebar":"tutorialSidebar"},"prerequisites/Other_languages/javascript":{"id":"prerequisites/Other_languages/javascript","title":"Javascript","description":"You don\'t need to install javascript - you have it already in your browser. To have a go at using","sidebar":"tutorialSidebar"},"prerequisites/Other_languages/README":{"id":"prerequisites/Other_languages/README","title":"Other languages","description":"There are lots of other programming languages out there - useful for different tasks. Here are some:","sidebar":"tutorialSidebar"},"prerequisites/pandas":{"id":"prerequisites/pandas","title":"Installing pandas","description":"","sidebar":"tutorialSidebar"},"prerequisites/python":{"id":"prerequisites/python","title":"Python","description":"You shouldn\'t need to install python - both your base system, and your conda environment should have a version of python installed already.  Check by typing","sidebar":"tutorialSidebar"},"prerequisites/R":{"id":"prerequisites/R","title":"R and RStudio","description":"A great deal of data analysis work is done in R - you\'ll need this","sidebar":"tutorialSidebar"},"prerequisites/README":{"id":"prerequisites/README","title":"Getting your environment setup","description":"The tutorials found here involve running bioinformatics tools and writing computer code. You will","sidebar":"tutorialSidebar"},"prerequisites/tidyverse":{"id":"prerequisites/tidyverse","title":"The R `tidyverse`","description":"Base R is ok but it can be a bit... clunky. For many of the tutorials we wil use a set of","sidebar":"tutorialSidebar"},"prerequisites/UNIX":{"id":"prerequisites/UNIX","title":"Setting up a UNIX terminal","description":"You will be doing a lot of work in the UNIX terminal - this is a command-line interface.","sidebar":"tutorialSidebar"},"programming/introduction_to_R/R_basic_features_practical":{"id":"programming/introduction_to_R/R_basic_features_practical","title":"Using R to load and manipulate data","description":"Tutors: Helen Lockstone and Ben Wright Bioinformatics Core Wellcome","sidebar":"tutorialSidebar"},"programming/introduction_to_R/R_introduction_to_basic_features":{"id":"programming/introduction_to_R/R_introduction_to_basic_features","title":"R: Introduction to Basic Features","description":"Author: Helen Lockstone","sidebar":"tutorialSidebar"},"programming/introduction_to_R/R_introduction_to_basic_features_slides":{"id":"programming/introduction_to_R/R_introduction_to_basic_features_slides","title":"Introduction to R","description":"Authors: Helen Lockstone and Matthieu Miossec Bioinformatics Core","sidebar":"tutorialSidebar"},"programming/introduction_to_R/README":{"id":"programming/introduction_to_R/README","title":"Introduction to R","description":"Welcome! To start the tutorial, go here.","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Closing_thoughts":{"id":"programming/programming_with_gene_annotations/Closing_thoughts","title":"Closing thoughts","description":"Go up to the table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Converting_gff_to_sqlite":{"id":"programming/programming_with_gene_annotations/Converting_gff_to_sqlite","title":"Writing a useful conversion program.","description":"Your module is already very useful! To demonstrate this, let\'s convert the GFF file into a different format - a","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Counting_genes_1":{"id":"programming/programming_with_gene_annotations/Counting_genes_1","title":"Counting genes I","description":"Some first statistics","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Counting_genes_2":{"id":"programming/programming_with_gene_annotations/Counting_genes_2","title":"Counting genes II: types of protein-coding genes","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Counting_genes_3":{"id":"programming/programming_with_gene_annotations/Counting_genes_3","title":"Extremely big, small and complex genes","description":"What is the largest gene, the gene with the most exons, or the gene with the most transcripts in each species?","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/exploring_gencode":{"id":"programming/programming_with_gene_annotations/exploring_gencode","title":"Investigating human genes","description":"If our module is any good, we should be able to do all the things we were [able to do on","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Getting_sequence_lengths":{"id":"programming/programming_with_gene_annotations/Getting_sequence_lengths","title":"Getting sequence lengths","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Getting_started_writing_some_code":{"id":"programming/programming_with_gene_annotations/Getting_started_writing_some_code","title":"Processing GFF files","description":"Up to table of contents / Back to the previous page / Go to the next page","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes":{"id":"programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes","title":"How much of the genome is in genes?","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Introduction":{"id":"programming/programming_with_gene_annotations/Introduction","title":"Introduction","description":"Up to table of contents / Go to next page.","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/List_of_ensembl_files":{"id":"programming/programming_with_gene_annotations/List_of_ensembl_files","title":"Appendix: a list of Ensembl GFF files.","description":"A list of gff files from ensembl can be found in solutions/ensemblgfflisting.txt.","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/making_a_module":{"id":"programming/programming_with_gene_annotations/making_a_module","title":"Making a module","description":"If you\'ve got this far, you\'ve written a function parsegff3to_dataframe() that [passes the","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Memory_issues_and_how_to_solve_them":{"id":"programming/programming_with_gene_annotations/Memory_issues_and_how_to_solve_them","title":"Memory issues and how to solve them","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/README":{"id":"programming/programming_with_gene_annotations/README","title":"Programming with gene annotations","description":"Welcome! This tutorial has two aims. On one hand it is about programming, specifically the type of","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Refactoring_makes_code_better":{"id":"programming/programming_with_gene_annotations/Refactoring_makes_code_better","title":"An aside on refactoring","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Scaling_up":{"id":"programming/programming_with_gene_annotations/Scaling_up","title":"Scaling up for a real analysis","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Visualisation":{"id":"programming/programming_with_gene_annotations/Visualisation","title":"Appendix: visualisation","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like":{"id":"programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like","title":"What gene annotation data looks like","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Where_next":{"id":"programming/programming_with_gene_annotations/Where_next","title":"What next?","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/README":{"id":"programming/README","title":"Programming","description":"Welcome!  To get started, choose your tutorial from hte list below:","sidebar":"tutorialSidebar"},"statistical_modelling/bayesian_statistics/figures/README":{"id":"statistical_modelling/bayesian_statistics/figures/README","title":"README","description":"Directory for images."},"statistical_modelling/bayesian_statistics/README":{"id":"statistical_modelling/bayesian_statistics/README","title":"Bayesian statistics","description":"Welcome!  This folder contains tutorials on Bayesian modelling and Markov Chain Monte Carlo (MCMC)."},"statistical_modelling/Hidden_markov_models/glycophorin_cnv_hmm":{"id":"statistical_modelling/Hidden_markov_models/glycophorin_cnv_hmm","title":"Modelling CNVs using a Hidden Markov Model","description":"!! Note !!  To run this part of the practical, you first need to have computed a set of log-likelihoods for","sidebar":"tutorialSidebar"},"statistical_modelling/Hidden_markov_models/glycophorin_cnv_warmup":{"id":"statistical_modelling/Hidden_markov_models/glycophorin_cnv_warmup","title":"Warmup: examining read coverage in a region","description":"Welcome! The practical studies a region of the human genome on chromosome 4 that is known to","sidebar":"tutorialSidebar"},"statistical_modelling/Hidden_markov_models/README":{"id":"statistical_modelling/Hidden_markov_models/README","title":"Glycophorin CNV calling tutorial","description":"Welcome! In this tutorial you will build a hidden markov model to model sequence read coverage","sidebar":"tutorialSidebar"},"statistical_modelling/notes/computing_pvalues":{"id":"statistical_modelling/notes/computing_pvalues","title":"Asymptotics and  P-values","description":"Much of classical statistics relies on \'asymptotics\', which means statistical behaviour when data","sidebar":"tutorialSidebar"},"statistical_modelling/README":{"id":"statistical_modelling/README","title":"Statistical modelling","description":"Welcome!  This folder contains tutorials on statistical modelling.","sidebar":"tutorialSidebar"},"statistical_modelling/regression_modelling/causal_graphs":{"id":"statistical_modelling/regression_modelling/causal_graphs","title":"Causal thinking","description":"In the linear regression example we saw some pretty odd behaviour. A model fit of genotype on expression","sidebar":"tutorialSidebar"},"statistical_modelling/regression_modelling/linear_regression_1":{"id":"statistical_modelling/regression_modelling/linear_regression_1","title":"Getting started: linear regression","description":"Prerequisites","sidebar":"tutorialSidebar"},"statistical_modelling/regression_modelling/linear_regression_cheatsheet":{"id":"statistical_modelling/regression_modelling/linear_regression_cheatsheet","title":"Linear regression cheatsheet","description":"See below for the probability cheatsheet, or download directly.","sidebar":"tutorialSidebar"},"statistical_modelling/regression_modelling/logistic_regression_1":{"id":"statistical_modelling/regression_modelling/logistic_regression_1","title":"Logistic regression","description":"Before moving on let\'s try a different type of regression: logistic regression. Unlike linear regression, which models a","sidebar":"tutorialSidebar"},"statistical_modelling/regression_modelling/README":{"id":"statistical_modelling/regression_modelling/README","title":"Statistical modelling I - regression","description":"Welcome!  This folder contains tutorials on modelling using regression.","sidebar":"tutorialSidebar"},"statistical_modelling/regression_modelling/regression_as_a_model":{"id":"statistical_modelling/regression_modelling/regression_as_a_model","title":"Regression as a parameteric statistical model","description":"So far we\'ve fit some linear and logistic regression models and seen how they behave on some real datasets.","sidebar":"tutorialSidebar"},"statistical_modelling/regression_modelling/simulating":{"id":"statistical_modelling/regression_modelling/simulating","title":"Checking estimates using simulation","description":"In the linear regression tutorial we saw results like this:","sidebar":"tutorialSidebar"}}}')}}]);