"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[2066],{3905:(e,t,n)=>{n.d(t,{Zo:()=>m,kt:()=>g});var a=n(7294);function o(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function r(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var a=Object.getOwnPropertySymbols(e);t&&(a=a.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,a)}return n}function i(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?r(Object(n),!0).forEach((function(t){o(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):r(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function s(e,t){if(null==e)return{};var n,a,o=function(e,t){if(null==e)return{};var n,a,o={},r=Object.keys(e);for(a=0;a<r.length;a++)n=r[a],t.indexOf(n)>=0||(o[n]=e[n]);return o}(e,t);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);for(a=0;a<r.length;a++)n=r[a],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(o[n]=e[n])}return o}var l=a.createContext({}),p=function(e){var t=a.useContext(l),n=t;return e&&(n="function"==typeof e?e(t):i(i({},t),e)),n},m=function(e){var t=p(e.components);return a.createElement(l.Provider,{value:t},e.children)},d={inlineCode:"code",wrapper:function(e){var t=e.children;return a.createElement(a.Fragment,{},t)}},c=a.forwardRef((function(e,t){var n=e.components,o=e.mdxType,r=e.originalType,l=e.parentName,m=s(e,["components","mdxType","originalType","parentName"]),c=p(n),g=o,h=c["".concat(l,".").concat(g)]||c[g]||d[g]||r;return n?a.createElement(h,i(i({ref:t},m),{},{components:n})):a.createElement(h,i({ref:t},m))}));function g(e,t){var n=arguments,o=t&&t.mdxType;if("string"==typeof e||o){var r=n.length,i=new Array(r);i[0]=c;var s={};for(var l in t)hasOwnProperty.call(t,l)&&(s[l]=t[l]);s.originalType=e,s.mdxType="string"==typeof e?e:o,i[1]=s;for(var p=2;p<r;p++)i[p]=n[p];return a.createElement.apply(null,i)}return a.createElement.apply(null,n)}c.displayName="MDXCreateElement"},5107:(e,t,n)=>{n.r(t),n.d(t,{assets:()=>l,contentTitle:()=>i,default:()=>d,frontMatter:()=>r,metadata:()=>s,toc:()=>p});var a=n(7462),o=(n(7294),n(3905));const r={sidebar_position:2},i="What gene annotation data looks like",s={unversionedId:"programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like",id:"programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like",title:"What gene annotation data looks like",description:"Up to table of contents",source:"@site/docs/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like.md",sourceDirName:"programming/programming_with_gene_annotations",slug:"/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like",permalink:"/whg-training-resources/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like",draft:!1,editUrl:"https://github.com/whg-training/whg-training-resources/edit/main/docs/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like.md",tags:[],version:"current",sidebarPosition:2,frontMatter:{sidebar_position:2},sidebar:"tutorialSidebar",previous:{title:"Introduction",permalink:"/whg-training-resources/programming/programming_with_gene_annotations/Introduction"},next:{title:"Processing GFF files",permalink:"/whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code"}},l={},p=[{value:"The general feature format",id:"the-general-feature-format",level:2},{value:"Making some test files",id:"making-some-test-files",level:3},{value:"Aside: how are gene annotations created anyway?",id:"aside-how-are-gene-annotations-created-anyway",level:3},{value:"Getting started coding\u200b",id:"getting-started-coding",level:2}],m={toc:p};function d(e){let{components:t,...n}=e;return(0,o.kt)("wrapper",(0,a.Z)({},m,n,{components:t,mdxType:"MDXLayout"}),(0,o.kt)("h1",{id:"what-gene-annotation-data-looks-like"},"What gene annotation data looks like"),(0,o.kt)("p",null,(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/programming/programming_with_gene_annotations/"},"Up to table of contents")),(0,o.kt)("h2",{id:"the-general-feature-format"},"The general feature format"),(0,o.kt)("p",null,"For this tutorial we will work with data in a format called 'General Feature Format' or ",(0,o.kt)("strong",{parentName:"p"},"GFF3"),". You can\n",(0,o.kt)("a",{parentName:"p",href:"https://m.ensembl.org/info/website/upload/gff3.html"},"read its specification here"),". Along with the closely-related GTF format, it is the most commonly-used format for storing gene annotations."),(0,o.kt)("p",null,"Before starting you will need to make sure you have a good understanding of what's in the data. To get started,\ncreate a new directory to work in:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-sh"},"mkdir genes_programming; cd genes_programming\n")),(0,o.kt)("p",null,"Now start by downloading three files:"),(0,o.kt)("ol",null,(0,o.kt)("li",{parentName:"ol"},(0,o.kt)("p",{parentName:"li"},"The human gene annotations ",(0,o.kt)("a",{parentName:"p",href:"https://www.gencodegenes.org"},"from the GENCODE project"))),(0,o.kt)("li",{parentName:"ol"},(0,o.kt)("p",{parentName:"li"},"Another version of the human gene annotations ",(0,o.kt)("a",{parentName:"p",href:"http://ftp.ensembl.org/pub/current_gff3/homo_sapiens/"},"from\nEnsembl"),". (",(0,o.kt)("strong",{parentName:"p"},"Note:")," the file\n",(0,o.kt)("inlineCode",{parentName:"p"},"Homo_sapiens.GRCh38.107.chr.gff3.gz")," will do for now - this contains genes that Ensembl could assign to one\nof the main chromosomes.)")),(0,o.kt)("li",{parentName:"ol"},(0,o.kt)("p",{parentName:"li"},"The ",(0,o.kt)("em",{parentName:"p"},"Plasmodium falciparum")," (malaria parasites) ",(0,o.kt)("a",{parentName:"p",href:"https://plasmodb.org/plasmo/app/downloads/Current_Release/Pfalciparum3D7/gff/data/"},"from\nPlasmodb"),"."))),(0,o.kt)("p",null,"A backup copy of these files can be found\n",(0,o.kt)("a",{parentName:"p",href:"https://www.well.ox.ac.uk/~gav/projects/whg-training-resources/data/programming/programming_with_gene_annotations/"},"in this folder"),"."),(0,o.kt)("p",null,"Use your ",(0,o.kt)("a",{parentName:"p",href:"/bioinformatics/exploring_gene_annotations_in_bash"},"command-line skills")," to download these files to the ",(0,o.kt)("inlineCode",{parentName:"p"},"genes_programming")," folder and\nhave a look at them.  (Feel free to download data for more species as well.)"),(0,o.kt)("p",null,"For example, the GENCODE file looks something like this:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre"},"##gff-version 3\n#description: evidence-based annotation of the human genome (GRCh38), version 38 (Ensembl 104)\n#provider: GENCODE\n#contact: gencode-help@ebi.ac.uk\n#format: gff3\n#date: 2021-03-12\n##sequence-region chr1 1 248956422\nchr1    HAVANA  gene    11869   14409   .   +   .   ID=ENSG00000223972.5;gene_id=ENSG00000223972.5;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;level=2;hgnc_id=HGNC:37102;havana_gene=OTTHUMG00000000961.2\nchr1    HAVANA  transcript  11869   14409   .   +   .   ID=ENST00000456328.2;Parent=ENSG00000223972.5;gene_id=ENSG00000223972.5;transcript_id=ENST00000456328.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=processed_transcript;transcript_name=DDX11L1-202;level=2;transcript_support_level=1;hgnc_id=HGNC:37102;tag=basic;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000362751.1\nchr1    HAVANA  exon    11869   12227   .   +   .   ID=exon:ENST00000456328.2:1;Parent=ENST00000456328.2;gene_id=ENSG00000223972.5;transcript_id=ENST00000456328.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=processed_transcript;transcript_name=DDX11L1-202;exon_number=1;exon_id=ENSE00002234944.1;level=2;transcript_support_level=1;hgnc_id=HGNC:37102;tag=basic;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000362751.1\nchr1    HAVANA  exon    12613   12721   .   +   .   ID=exon:ENST00000456328.2:2;Parent=ENST00000456328.2;gene_id=ENSG00000223972.5;transcript_id=ENST00000456328.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=processed_transcript;transcript_name=DDX11L1-202;exon_number=2;exon_id=ENSE00003582793.1;level=2;transcript_support_level=1;hgnc_id=HGNC:37102;tag=basic;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000362751.1\n...\n")),(0,o.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,o.kt)("div",{parentName:"div",className:"admonition-heading"},(0,o.kt)("h5",{parentName:"div"},(0,o.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,o.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,o.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Question")),(0,o.kt)("div",{parentName:"div",className:"admonition-content"},(0,o.kt)("p",{parentName:"div"},"GFF files from different projects differ slightly in what's in them. Can you spot differences in how the three\nfiles above are encoded?"))),(0,o.kt)("div",{className:"admonition admonition-note alert alert--secondary"},(0,o.kt)("div",{parentName:"div",className:"admonition-heading"},(0,o.kt)("h5",{parentName:"div"},(0,o.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,o.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"14",height:"16",viewBox:"0 0 14 16"},(0,o.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.3 5.69a.942.942 0 0 1-.28-.7c0-.28.09-.52.28-.7.19-.18.42-.28.7-.28.28 0 .52.09.7.28.18.19.28.42.28.7 0 .28-.09.52-.28.7a1 1 0 0 1-.7.3c-.28 0-.52-.11-.7-.3zM8 7.99c-.02-.25-.11-.48-.31-.69-.2-.19-.42-.3-.69-.31H6c-.27.02-.48.13-.69.31-.2.2-.3.44-.31.69h1v3c.02.27.11.5.31.69.2.2.42.31.69.31h1c.27 0 .48-.11.69-.31.2-.19.3-.42.31-.69H8V7.98v.01zM7 2.3c-3.14 0-5.7 2.54-5.7 5.68 0 3.14 2.56 5.7 5.7 5.7s5.7-2.55 5.7-5.7c0-3.15-2.56-5.69-5.7-5.69v.01zM7 .98c3.86 0 7 3.14 7 7s-3.14 7-7 7-7-3.12-7-7 3.14-7 7-7z"}))),"Note")),(0,o.kt)("div",{parentName:"div",className:"admonition-content"},(0,o.kt)("p",{parentName:"div"},"Each project has its own version of the file format specification as well.  For example see:"),(0,o.kt)("ul",{parentName:"div"},(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},"the ",(0,o.kt)("a",{parentName:"p",href:"https://www.ensembl.org/info/website/upload/gff.html"},"Ensembl page on GFF format"))),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},"the ",(0,o.kt)("a",{parentName:"p",href:"https://www.gencodegenes.org/pages/data_format.html"},"GENCODE file format page")," (this actually described\n",(0,o.kt)("inlineCode",{parentName:"p"},"GTF")," format, but it's almost identical and has the same basic encoding.  We'll use ",(0,o.kt)("inlineCode",{parentName:"p"},"GFF3")," here.)")),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},"The PlasmoDB files are similar to the Ensembl ones, but if you look closely you'll see the file has some\ndifference to both those above."))))),(0,o.kt)("h3",{id:"making-some-test-files"},"Making some test files"),(0,o.kt)("p",null,"These files are pretty big. For code development purposes it is useful to make some smaller datasets that don't\nuse up so much space.  Make these now by taking the first 1,000 lines of each file:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre"},"cat gencode.v41.annotation.gff3 | head -n 1000 > gencode.v41.annotation.head.gff3\n")),(0,o.kt)("p",null,"and so on."),(0,o.kt)("div",{className:"admonition admonition-tip alert alert--success"},(0,o.kt)("div",{parentName:"div",className:"admonition-heading"},(0,o.kt)("h5",{parentName:"div"},(0,o.kt)("span",{parentName:"h5",className:"admonition-icon"},(0,o.kt)("svg",{parentName:"span",xmlns:"http://www.w3.org/2000/svg",width:"12",height:"16",viewBox:"0 0 12 16"},(0,o.kt)("path",{parentName:"svg",fillRule:"evenodd",d:"M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"}))),"Note")),(0,o.kt)("div",{parentName:"div",className:"admonition-content"},(0,o.kt)("p",{parentName:"div"},"If your data file ends in ",(0,o.kt)("inlineCode",{parentName:"p"},".gz"),", it is gzip-compressed. You will need to use ",(0,o.kt)("inlineCode",{parentName:"p"},"gunzip -c")," instead of ",(0,o.kt)("inlineCode",{parentName:"p"},"cat"),". This\nwill decompress the data and pipe it into ",(0,o.kt)("inlineCode",{parentName:"p"},"head"),"."))),(0,o.kt)("h3",{id:"aside-how-are-gene-annotations-created-anyway"},"Aside: how are gene annotations created anyway?"),(0,o.kt)("p",null,"Gene annotation pipelines are pretty complicated - they involve prediction of genes from genome sequence,\nalignment of known transcript sequences and short-read RNA, transfer of annotations from other genomes, and\nother steps.  "),(0,o.kt)("p",null,"The Ensembl gene annotations are created using a particular pipeline that is described in detail\n",(0,o.kt)("a",{parentName:"p",href:"https://www.ensembl.org/info/genome/genebuild/index.html"},"here"),". This includes a manual curation step,\n'HAVANA', which is applied to the best-annotated genomes including humans, mice, zebrafish and rats."),(0,o.kt)("p",null,"GENCODE is based on the Ensembl annotation ",(0,o.kt)("a",{parentName:"p",href:"https://www.gencodegenes.org/pages/faq.html"},"as described here"),"\nbut with extra annotation added."),(0,o.kt)("p",null,"The NCBI is a second project that creates gene annotations (NCBI RefSeq). Their pipeline for Eukaryotes is\ndescribed ",(0,o.kt)("a",{parentName:"p",href:"https://www.ncbi.nlm.nih.gov/genome/annotation_euk/process/"},"here"),". The process looks like this:"),(0,o.kt)("p",null,(0,o.kt)("img",{parentName:"p",src:"https://www.ncbi.nlm.nih.gov/core/assets/genome/images/Pipeline_RFAM.png",alt:"diagram of gene annotation pipeline"})),(0,o.kt)("p",null,"Simple!"),(0,o.kt)("p",null,"An important thing to notice is that these annotations are made from severe data sources - including\ncomputational predictions from the genome assemblies themselves, but also alignments of known RNA and protein\nsequences, and a great deal of automatic and for some genomes manual curation. They shouldn't be regarded as\nthe whole truth about genes, but as a current best guess - and for some genomes like humans they are fairly\ncomplete."),(0,o.kt)("h2",{id:"getting-started-coding"},"Getting started coding\u200b"),(0,o.kt)("p",null,"In this tutorial we're going to write some code to process these data files. When you're ready, ",(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code"},"get started\nwriting some code"),"."))}d.isMDXComponent=!0}}]);