"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[8496],{3905:(e,n,t)=>{t.d(n,{Zo:()=>l,kt:()=>p});var r=t(7294);function i(e,n,t){return n in e?Object.defineProperty(e,n,{value:t,enumerable:!0,configurable:!0,writable:!0}):e[n]=t,e}function o(e,n){var t=Object.keys(e);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);n&&(r=r.filter((function(n){return Object.getOwnPropertyDescriptor(e,n).enumerable}))),t.push.apply(t,r)}return t}function a(e){for(var n=1;n<arguments.length;n++){var t=null!=arguments[n]?arguments[n]:{};n%2?o(Object(t),!0).forEach((function(n){i(e,n,t[n])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(t)):o(Object(t)).forEach((function(n){Object.defineProperty(e,n,Object.getOwnPropertyDescriptor(t,n))}))}return e}function s(e,n){if(null==e)return{};var t,r,i=function(e,n){if(null==e)return{};var t,r,i={},o=Object.keys(e);for(r=0;r<o.length;r++)t=o[r],n.indexOf(t)>=0||(i[t]=e[t]);return i}(e,n);if(Object.getOwnPropertySymbols){var o=Object.getOwnPropertySymbols(e);for(r=0;r<o.length;r++)t=o[r],n.indexOf(t)>=0||Object.prototype.propertyIsEnumerable.call(e,t)&&(i[t]=e[t])}return i}var c=r.createContext({}),u=function(e){var n=r.useContext(c),t=n;return e&&(t="function"==typeof e?e(n):a(a({},n),e)),t},l=function(e){var n=u(e.components);return r.createElement(c.Provider,{value:n},e.children)},g={inlineCode:"code",wrapper:function(e){var n=e.children;return r.createElement(r.Fragment,{},n)}},_=r.forwardRef((function(e,n){var t=e.components,i=e.mdxType,o=e.originalType,c=e.parentName,l=s(e,["components","mdxType","originalType","parentName"]),_=u(t),p=i,d=_["".concat(c,".").concat(p)]||_[p]||g[p]||o;return t?r.createElement(d,a(a({ref:n},l),{},{components:t})):r.createElement(d,a({ref:n},l))}));function p(e,n){var t=arguments,i=n&&n.mdxType;if("string"==typeof e||i){var o=t.length,a=new Array(o);a[0]=_;var s={};for(var c in n)hasOwnProperty.call(n,c)&&(s[c]=n[c]);s.originalType=e,s.mdxType="string"==typeof e?e:i,a[1]=s;for(var u=2;u<o;u++)a[u]=t[u];return r.createElement.apply(null,a)}return r.createElement.apply(null,t)}_.displayName="MDXCreateElement"},8146:(e,n,t)=>{t.r(n),t.d(n,{assets:()=>c,contentTitle:()=>a,default:()=>g,frontMatter:()=>o,metadata:()=>s,toc:()=>u});var r=t(7462),i=(t(7294),t(3905));const o={sidebar_position:1},a="Introduction to next-generation sequencing",s={unversionedId:"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/README",id:"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/README",title:"Introduction to next-generation sequencing",description:"In this tutorial we will demonstrate a basic pipeline for analysing paired-end short-read genomic",source:"@site/docs/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/README.md",sourceDirName:"next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis",slug:"/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/",permalink:"/whg-training-resources/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/",draft:!1,editUrl:"https://github.com/whg-training/whg-training-resources/edit/main/docs/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/README.md",tags:[],version:"current",sidebarPosition:1,frontMatter:{sidebar_position:1},sidebar:"tutorialSidebar",previous:{title:"Next-generation sequencing",permalink:"/whg-training-resources/next_generation_sequencing/"},next:{title:"Getting the prerequisites",permalink:"/whg-training-resources/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Prerequisites"}},c={},u=[{value:"Contributors",id:"contributors",level:2}],l={toc:u};function g(e){let{components:n,...t}=e;return(0,i.kt)("wrapper",(0,r.Z)({},l,t,{components:n,mdxType:"MDXLayout"}),(0,i.kt)("h1",{id:"introduction-to-next-generation-sequencing"},"Introduction to next-generation sequencing"),(0,i.kt)("p",null,"In this tutorial we will demonstrate a basic pipeline for analysing paired-end short-read genomic\nsequencing data. We will start with raw data in a FASTQ file, inspect quality control metrics,\nalign the data, and then use it to look for genetic variation."),(0,i.kt)("p",null,"To get started, go ",(0,i.kt)("a",{parentName:"p",href:"/whg-training-resources/next_generation_sequencing/introduction_to_next_generation_sequencing_data_analysis/Prerequisites"},"to and get the prerequisites"),"."),(0,i.kt)("h2",{id:"contributors"},"Contributors"),(0,i.kt)("ul",null,(0,i.kt)("li",{parentName:"ul"},"Gavin Band"),(0,i.kt)("li",{parentName:"ul"},"Mattheiu Miossec"),(0,i.kt)("li",{parentName:"ul"},"Annie Forster")))}g.isMDXComponent=!0}}]);