<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-programming/introduction_to_R/R_basic_features_practical">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.20">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.24/dist/katex.min.css" integrity="sha384-odtC+0UGzzFL/6PNoE8rX/SPcQDXBJ+uRepguP4QkPCm2LBxH3FA3y+fKSiJ+AmM" crossorigin="anonymous"><title data-rh="true">Using R to load and manipulate data | WHG training resources</title><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://well.ox.ac.uk//whg-training-resources/programming/introduction_to_R/R_basic_features_practical/"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Using R to load and manipulate data | WHG training resources"><meta data-rh="true" name="description" content="Tutors: Helen Lockstone and Ben Wright Bioinformatics Core Wellcome"><meta data-rh="true" property="og:description" content="Tutors: Helen Lockstone and Ben Wright Bioinformatics Core Wellcome"><link data-rh="true" rel="icon" href="/whg-training-resources/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://well.ox.ac.uk//whg-training-resources/programming/introduction_to_R/R_basic_features_practical/"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/programming/introduction_to_R/R_basic_features_practical/" hreflang="en"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/programming/introduction_to_R/R_basic_features_practical/" hreflang="x-default"><link rel="stylesheet" href="/whg-training-resources/assets/css/styles.b827ae01.css">
<link rel="preload" href="/whg-training-resources/assets/js/runtime~main.50846bb3.js" as="script">
<link rel="preload" href="/whg-training-resources/assets/js/main.1b628545.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region"><a href="#" class="skipToContent_ZgBM">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/whg-training-resources/"><div class="navbar__logo"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--light_TfLj"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--dark_oUvU"></div><b class="navbar__title text--truncate">Training Resources</b></a><a class="navbar__item navbar__link navbar__link--active" href="/whg-training-resources/overview/">Click here for the tutorials</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a><div class="toggle_S7eR colorModeToggle_vKtC"><button class="clean-btn toggleButton_rCf9 toggleButtonDisabled_Pu9x" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_v35p"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_nQuB"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_dLyj"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docsWrapper_mKqt"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_RiI4" type="button"></button><div class="docPage_ualW"><aside class="theme-doc-sidebar-container docSidebarContainer_UQUJ"><div class="sidebar_RiAD"><nav class="menu thin-scrollbar menu_izAj"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/overview/">Overview of the tutorials</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/prerequisites/">Getting your environment setup</a><button aria-label="Toggle the collapsible sidebar category &#x27;Getting your environment setup&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/bioinformatics/">Bioinformatics</a><button aria-label="Toggle the collapsible sidebar category &#x27;Bioinformatics&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" href="/whg-training-resources/programming/">Programming</a><button aria-label="Toggle the collapsible sidebar category &#x27;Programming&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/programming/introduction_to_R/">Introduction to R</a><button aria-label="Toggle the collapsible sidebar category &#x27;Introduction to R&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/introduction_to_R/R_introduction_to_basic_features/">R: Introduction to Basic Features</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/whg-training-resources/programming/introduction_to_R/R_basic_features_practical/">Using R to load and manipulate data</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/introduction_to_R/R_introduction_to_basic_features_slides/">Introduction to R</a></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/">Programming with gene annotations</a><button aria-label="Toggle the collapsible sidebar category &#x27;Programming with gene annotations&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/data_visualisation/">Data visualisation</a><button aria-label="Toggle the collapsible sidebar category &#x27;Data visualisation&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/next_generation_sequencing/">Next-generation sequencing</a><button aria-label="Toggle the collapsible sidebar category &#x27;Next-generation sequencing&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/statistical_modelling/">Statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;Statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/population_genetics/">Population genetics</a><button aria-label="Toggle the collapsible sidebar category &#x27;Population genetics&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/genome_wide_association_studies/">Genome-wide association analysis</a><button aria-label="Toggle the collapsible sidebar category &#x27;Genome-wide association analysis&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/LICENSE/">LICENSE</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/contributors/">contributors</a></li></ul></nav></div></aside><main class="docMainContainer_uL0j"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_DM6M"><div class="docItemContainer_vinB"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Xlws" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/whg-training-resources/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_kU5B"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/programming/"><span itemprop="name">Programming</span></a><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/programming/introduction_to_R/"><span itemprop="name">Introduction to R</span></a><meta itemprop="position" content="2"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">Using R to load and manipulate data</span><meta itemprop="position" content="3"></li></ul></nav><div class="tocCollapsible_bZGK theme-doc-toc-mobile tocMobile_TmEX"><button type="button" class="clean-btn tocCollapsibleButton_l22C">On this page</button></div><div class="theme-doc-markdown markdown"><h1>Using R to load and manipulate data</h1><p>Tutors: Helen Lockstone and Ben Wright Bioinformatics Core Wellcome
Centre for Human Genetics University of Oxford</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="practical-tutorial">Practical Tutorial<a class="hash-link" href="#practical-tutorial" title="Direct link to heading">​</a></h2><p>Earlier we introduced the R software environment, some key features of
the R programming language and how to start using it. We will now do
some practical exercises working with example data to perform typical
tasks. First, some important acknowledgements:</p><p><em>The tutorial that follows is in part adapted from the <a href="https://software-carpentry.org/lessons/" target="_blank" rel="noopener noreferrer">Software
Carpentry Foundation</a>
Programming with R, specifically the Analysing Patient Data
<a href="http://swcarpentry.github.io/r-novice-inflammation/01-starting-with-data/index.html" target="_blank" rel="noopener noreferrer">tutorial</a>.
The Software Carpentry material is available for re-use under a
<a href="https://creativecommons.org/licenses/by/4.0/" target="_blank" rel="noopener noreferrer">Creative Commons License</a>
and I am grateful to the original authors.</em></p><p><em>This document and a variety of extensions to the tutorial material were
written and developed by Helen Lockstone, with contributions from Ben
Wright. All the material for today&#x27;s course is available
<a href="https://www.well.ox.ac.uk/training/R_Basic_Features" target="_blank" rel="noopener noreferrer">here</a></em></p><p>You can open a copy of this tutorial guide on the machine you are using
to copy and paste any particularly long commands. Generally, manually
type in the shorter commands to get used to the R environment and
structure of commands – most are very short. Commands are shown within
light-grey boxes; remember that lines starting with a <!-- -->#<!-- --> are comments
and do not need to be run.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="setup">Setup<a class="hash-link" href="#setup" title="Direct link to heading">​</a></h3><p>We set up our R session for this practical before the break. You can
check the current working directory with the following command:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">getwd()</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The output (directory path) should match the location you are working in
today, and this folder should also contain the files named
&quot;inflammation<!-- -->_<!-- -->data.csv&quot; and &quot;sample.csv&quot;.</p><p>If this is the case, you should have everything you need to run this
tutorial. If not, please let us know and we will get you started.</p><p>If you see an error message at any point, first check the command
matches that in the tutorial exactly and that you haven&#x27;t accidentally
missed an earlier command out. Pay particular attention to lower/upper
case letters, underscores, dashes or dots in function or object names,
and that brackets and quotes are correctly paired. If you can&#x27;t spot the
problem or have a question at any point, please don&#x27;t hesitate to ask.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="helpful-tips">Helpful Tips<a class="hash-link" href="#helpful-tips" title="Direct link to heading">​</a></h3><ul><li><p>Enter your commands in the top-left panel of RStudio (a text editor)
as this means they can be saved to keep a record of what you have
done. To run a command written in this panel, make sure the cursor
is located somewhere in the line of code and click the Run icon with
the green arrow. The command is automatically copied into the lower
console panel and executed by R.</p></li><li><p>To save your work, click on the disk icon in the same toolbar as the
&#x27;run&#x27; button. Giving a filename with a .R extension, such as
&#x27;R<!-- -->_<!-- -->course<!-- -->_<!-- -->code.R&#x27; saves it as an R script file - this can be
opened like any text file but the .R extension is useful to identify
your files that contain scripts. This is the usually the easiest way
to work; if you need to close your session and return to it another
time, it is easy to run the code again. If it were appropriate, all
of the code stored in a script can be executed in R from start to
finish with the command:</p></li></ul><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">source(&quot;script_name.R&quot;)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><ul><li><p>The <!-- -->#<!-- --> symbol is the comment character in R – lines in a script
starting with a <!-- -->#<!-- --> (or more commonly <!-- -->#<!-- -->#) can include comments about
what the code is doing. It is strongly recommended to comment your
code as much as possible because it will help others understand what
it is doing, including yourself if you revisit it sometime after
originally working on it. A <!-- -->#<!-- --> can also be used after a command to
note any result or information relating to that command – everything
after the <!-- -->#<!-- --> will be ignored by R but serve as useful information to
the programmer. You can also use long lines of <!-- -->#<!-- --> symbols to break
your code into sections.</p></li><li><p>In the console panel, you can use the up/down arrow keys to scroll
through previous commands to re-run or edit them easily if needed.
RStudio also has features to help auto-complete names of functions
and objects, and pairing brackets and quotes.</p></li></ul><h2 class="anchor anchorWithStickyNavbar_mojV" id="reading-in-data-from-a-file">Reading in data from a file<a class="hash-link" href="#reading-in-data-from-a-file" title="Direct link to heading">​</a></h2><p>The first thing we need to do is load or read in the data from our files
so it is accessible in the current R session. There are a few possible
ways to do this but we will use the function <strong>read.csv</strong> because our
files are saved in &#x27;comma-separated values&#x27; or csv format. To find out
details of how to use this function you can search RStudio&#x27;s Help menu
(bottom right panel) or type:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">help(read.csv)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The help page shows us the arguments for this function and their default
values where applicable. For example, we see sep=&quot;,&quot;, which means the
fields in a row will be separated on commas, and header=TRUE, which
means it is expecting the first row of the file to contain names for
each column. The first file we will work with is the one named
&#x27;inflammation<!-- -->_<!-- -->data.csv&#x27;. If we inspect this file in Excel or a text
editor (by opening it directly from its location on your computer) we
see there are no column names, just a large set of numerical values.
Therefore, we need to explicitly include the header argument in our
command, changing it to FALSE to over-ride the default behaviour of the
function. This is an example of how arguments modify a function&#x27;s
precise behaviour, rather than requiring two separate functions to exist
for files with/without header rows.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">inf.data &lt;- read.csv(&quot;inflammation_data.csv&quot;, header=FALSE)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><em>Note: if you see an error message similar to &#x27;No such file or
directory&#x27; when trying to read in a file, it is likely that either (i)
the file is not located in the current working directory; (ii) there is
one (or more) typos in the filename.</em></p><p>It is worth breaking this command apart to refresh on some of the
terminology used this morning, as it can be hard at first to
differentiate object names (decided by us) from R functions (pre-defined
in the language) when looking at R code or commands.</p><p>We have given a name for a new object, <strong>inf.data</strong>, in which to store
the contents of the file &#x27;inflammation<!-- -->_<!-- -->data.csv&#x27;. Our object name is
descriptive without being too long, shortening &#x27;inflammation&#x27; to &#x27;inf&#x27;
for our convenience. We are using the in-built R function <strong>read.csv</strong>,
and provide two arguments:</p><ul><li>the name of the file to read in</li><li>header=FALSE, indicating to R that our file does not have a header
row</li></ul><p>There are many other arguments to the <strong>read.csv</strong> function to further
refine its behaviour but these are either optional or the default
settings are fine for most situations. By running this command, R
creates the object <strong>inf.data</strong> and information about it appears in the
top-right panel of RStudio. It is detailed as 60 obs. (observations) of
40 variables and if you hover the mouse pointer over the name,
<strong>inf.data</strong>, it indicates that the object is a data frame. Finally by
clicking the spreadsheet icon to the righthand side, the contents of the
object are loaded in the top-left panel in a new window, titled by the
name of object. This shows we have successfully loaded the data. The
columns have been automatically named by R as V1 through to V40, as
column names have to start with a letter. The rows are simply numbered -
the only restriction to row names is that they must be unique.</p><p>What do you think will happen if you run the command above without
including the header argument? Try it by saving the contents into a new
object called <strong>test</strong> and compare this to <strong>inf.data</strong>:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">test &lt;- read.csv(&quot;inflammation_data.csv&quot;) </span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Another object named <strong>test</strong> now appears. We can use the <strong>head</strong>
command to inspect the first 6 rows of each object. For display
purposes, we&#x27;ll also only include the first 6 columns:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">head(test [, 1:6])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">head(inf.data[ , 1:6])</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Discuss with a neighbour what you observe.</p><p><em>This is a good example of how easily something unwanted can happen in R
and the importance of checking your objects contain what you intend them
to. Any mistakes can simply be corrected by re-running the command e.g.
with the appropriate header argument, and overwriting any previous
version of the object.</em></p><h2 class="anchor anchorWithStickyNavbar_mojV" id="two-dimensional-data-structures">Two-dimensional data structures<a class="hash-link" href="#two-dimensional-data-structures" title="Direct link to heading">​</a></h2><p>Our original file contained rows and columns of data, and R has suitable
2-dimensional data structures to store such data: matrices and
dataframes. These can both be thought of as tables of data, analogous to
an Excel spreadsheet. Matrices require all columns to be of the same
type, while data frames can have columns of different data types. Given
that experimental data is often a mixture of numeric values (e.g.
measurements) and associated descriptive information, data frames are a
very commonly used data structure in R.</p><p><em>NB while it is possible to hold mixed data types in a matrix object as
well, R will use its internal hierarchy of data types to choose one that
is applicable to all columns – often this means numerical columns get
converted to character strings, and certain functions may not perform as
expected if this is not noticed, or produce an error message.</em></p><p>In this case, a data frame object has been created:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">class(inf.data)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>We can check how each column of data has been treated by R e.g. for the
first column:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">class(inf.data[,1])</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><em>In fact, as all the columns contain data of the same type, R could
equally well store this data as a matrix object. A data frame has been
created because the functions <strong>read.csv</strong> and <strong>read.table</strong> are
specifically designed to deal with mixed column classes and produce data
frames by default. Another function <strong>scan</strong> can be used to read in
matrices, especially large ones.</em></p><p>It is worth noting that some functions operate on matrix objects, and so
converting between classes is sometimes needed.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">inf.data &lt;- as.matrix(inf.data)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">class(inf.data)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>We will continue with the matrix form of this object for now, and load a
mixed dataset later. Most operations on either kind of 2-dimensional
object are the same. For example, we can find the dimensions of a matrix
or a data frame with the <strong>dim</strong> function.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">dim(inf.data)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The output of <strong>dim</strong> is printed to the screen and shows the number of
rows the object contains, followed by the number of columns (the
convention is always rows, then columns but this can be hard to remember
at first as there is no indication).</p><p>If unsure, the functions <strong>nrow</strong> and <strong>ncol</strong> will return the number of
rows or columns respectively; these take as their argument the name of
the object:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">nrow(inf.data)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ncol(inf.data)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This particular file suffers from the lack of any labels to annotate
what data is recorded in the rows and columns. The Software Carpentry
tutorial provides the following information: <em>“We are studying
inflammation in patients who have been given a new treatment for
arthritis. Each row holds the observations for just one patient. Each
column holds the inflammation measured in a day, so we have a set of
values in successive days.&quot;</em></p><p>Our object has 60 rows and 40 columns, so we infer from the information
above that there are 60 patients, and 40 days.</p><p>Again we see a way for mistakes to easily creep into data analysis –
here we have to rely on information given to us second-hand to know what
is what. What if that information were wrong? Are there any checks we
can make ourselves to be sure patients are in rows? We are not told how
many patients were included so simply checking the number of rows won&#x27;t
help. And what if there were 50 patients and measurements taken over 50
days?</p><p>Even with careful scrutiny it would be hard to know how the data are
presented (patients in rows or columns) from the data alone. We could
perhaps make some plots to help us, or we might spot the zero values in
the first column. Scrolling down the object display in the top-left
panel or displaying the first column in the console confirms they are
all zeroes, and the values in each row tend to rise across the first few
columns. We may be reassured by this that the patients are indeed in the
rows, since we might expect inflammation to rise over time, and an
individual recording 0 on every single day might be unlikely (though not
impossible).</p><p>It would be prudent to add some row and column names to reduce the
chance of making a mistake later when dealing with this data:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">rownames(inf.data) &lt;- paste(&quot;Patient&quot;, 1:60, sep=&quot;_&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">colnames(inf.data) &lt;-  paste(&quot;Day&quot;, 1:40, sep=&quot;_&quot;)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This introduces the very useful and versatile function <strong>paste</strong>. Note
that adding row and column names does not change the size of the data
object, but we can see them displayed by reloading the object. They are
similar to the alphabetical columns and numbered rows in an Excel
spreadsheet.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="accessing-data">Accessing Data<a class="hash-link" href="#accessing-data" title="Direct link to heading">​</a></h2><p>Earlier we looked at accessing elements of a one-dimensional vector
object. For matrices and dataframes, a similar approach with square
brackets is used:</p><p><em>object<!-- -->_<!-- -->name<!-- -->[<!-- -->rows, cols<!-- -->]</em></p><p>By specifying the rows and columns of interest, an object can be subset
in a variety of ways to inspect or extract different parts of it.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">inf.data[1,1] # this pulls out the data value in the first row of the first column</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">inf.data[30, 20] # any single entry can be extracted by specifying the row and column</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>How might you select the data in the first 5 rows for the first 5
columns? Add your command for this below.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">## If we need to select non-contiguous portions of the object, we&#x27;ll need the help of the c() function:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">inf.data[c(1, 3, 5), c(10, 20)]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## If you want to display all columns for selected row(s), leave blank space after the comma:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">inf.data[5, ] # All columns for row 5</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## Or blank space before the comma to select all rows for given column(s):</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">inf.data[ , 1:5] # all rows, columns 1 through 5</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>We added column names to our object earlier, and columns can also be
accessed by name (in quotes):</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">inf.data[&quot;Patient_1&quot;, ]</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Suppose you want to determine the maximum inflammation for patient 5
across days three to seven. To do this you would extract the relevant
subset from the data frame and calculate the maximum value. Which of the
following lines of R code gives the correct answer?</p><ol><li>max(inf.data<!-- -->[<!-- -->5, <!-- -->]<!-- -->)</li><li>max(inf.data<!-- -->[<!-- -->3:7, 5<!-- -->]<!-- -->)</li><li>max(inf.data<!-- -->[<!-- -->5, 3:7<!-- -->]<!-- -->)</li><li>max(inf.data<!-- -->[<!-- -->5, 3, 7<!-- -->]<!-- -->)</li></ol><h2 class="anchor anchorWithStickyNavbar_mojV" id="analysing-data">Analysing Data<a class="hash-link" href="#analysing-data" title="Direct link to heading">​</a></h2><p>We can perform many simple analyses of the data by applying functions
such as max, min, mean, or summary to our data object. We might want to
determine the maximum value per patient or the average value per day.
The following examples illustrate how this can be done extremely
efficiently in R, starting with an approach that is the opposite (and
definitely not recommended!).</p><p>Suppose we want to find the maximum inflammation score for each patient
across the 40 days of measurements. Let&#x27;s start by calculating it for
patient 1.</p><p>Extracting the data for patient 1 (i.e. the first row) is the first
obvious step, and perhaps we decide it makes sense to store the values
for this patient in a new object:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">inf.patient1 &lt;- inf.data[1, ] </span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>We can then calculate the maximum value for Patient 1:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">max(inf.patient1)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Although this seems reasonable enough, there are several issues:</p><ul><li>We&#x27;ve created an additional object to store data that is simply a
duplicate of what is already contained in our original object</li><li>It doesn&#x27;t scale well to do this for all 60 patients</li><li>The result is output to the console and therefore hard to do
anything further with</li></ul><p>If we did continue with this approach, there would be 60 new objects
(all with very similar names), a high probability of having made a
typing mistake somewhere (perhaps overwriting one patient&#x27;s data with
another), and a large set of results that we&#x27;d have to manually write
down or transfer to an Excel spreadsheet - all of which is very messy
and prone to error.</p><p>We can easily dispense with the intermediate step of creating a new
object:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain"># max inflammation for patient 1</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">max(inf.data[1, ])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># or equivalently</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">max(inf.data[&quot;Patient_1&quot;, ])</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>These commands are the same as extracting the data for patient 1 as we
did earlier, but instead of printing to the screen or storing in a new
object, the command is used directly as an argument to the function
<strong>max</strong> by enclosing in the ().</p><p><em>Commands can be nested in this way to achieve multiple steps in a
single line of code; too many commands in one line though can make it
harder to work out what the code is doing, as well as increase the
chance of the code not doing as intended – the location of brackets
becomes vital.</em></p><p>We&#x27;d really like a way to this for all 60 patients without duplicating
the code 60 times. Loops are one option (not discussed here) but the
<strong>apply</strong> function is the most efficient approach:</p><p><strong>apply</strong> allows us to repeat a function on all of the rows (MARGIN = 1)
or columns (MARGIN = 2) of a data frame simultaneously:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">max_inf_patient &lt;- apply(inf.data, MARGIN=1, max)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Similarly, we could compute the average inflammation per day with a
single line of code:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">avg_inf_day &lt;- apply(inf.data, MARGIN=2, mean)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Comparing these two commands will help understand the <strong>apply</strong>
function, which is not intuitive but highly efficient as we have seen.
The arguments to apply are:</p><ul><li>The data object</li><li>MARGIN, indicating whether to apply over rows (1) or columns (2)</li><li>The name of (another) function to be applied</li></ul><p>We wanted to find the maximum inflammation score for each patient, so we
looked across the rows and used the <strong>max</strong> function. To modify the
command to find the average inflammation per day, we switched the MARGIN
argument to 2 for columns, and gave the final argument as <strong>mean</strong>.</p><p><em>While the MARGIN argument is explicitly assigned above, R is equally
happy to infer from the shortened command apply(inf.data, 1, max) that
the 1 should be assigned to the second defined argument of apply. You
can also write your own bespoke functions as required and use apply to
run them over an object.</em></p><p>We have also solved the final issue with our initial approach by storing
the results in suitably-named objects for further work.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">length(max_inf_patient) </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">head(max_inf_patient)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Another useful function is <strong>summary</strong>. This returns the minimum value,
first quartile, median, mean, third quartile and the maximum value, all
very useful information to make an initial inspection of your data.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">summary(inf.data[, 1:4]) # for each of the first 4 days</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><h2 class="anchor anchorWithStickyNavbar_mojV" id="plotting-data">Plotting Data<a class="hash-link" href="#plotting-data" title="Direct link to heading">​</a></h2><p>Visualising data is a vital part of statistical analysis, and R&#x27;s
plotting capabilities are a key reason for its popularity. There is a
related course R: Visualisation that you can take if interested to learn
more. Here, we introduce ways to make a few simple plots. Let&#x27;s take a
look at the average inflammation over time. Recall that we already
calculated these values above and saved them in an object named
<strong>avg<!-- -->_<!-- -->inf<!-- -->_<!-- -->day</strong>. Plotting the values is done with the function
<strong>plot</strong>:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">plot(avg_inf_day)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## Default labels and settings are used but we can refine our plot with some additional arguments:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot(avg_inf_day, main=&quot;Inflammation Scores Over Time&quot;, xlab=&quot;Day&quot;, ylab=&quot;Average_inflammation_score&quot;) # adding title and axis labels</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot(avg_inf_day, main=&quot;Inflammation Scores Over Time&quot;, xlab=&quot;Day&quot;, ylab=&quot;Average_inflammation_score&quot;, pch=4, col=&quot;red&quot;) # changing the plotting symbols and colour</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## Similarly, we could plot the data per patient:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot(max_inf_patient)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## Here, we might decide to use a boxplot instead:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">boxplot(max_inf_patient, main=&quot;Maximum Inflammation Scores&quot;, ylab=&quot;Max_inf_score&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">legend(&quot;topright&quot;, legend=&quot;n=60 patients&quot;, cex=0.8) # adding a legend</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>When we are happy with our plots, they can be saved to a file.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain"> pdf(&quot;Inflammation_plots.pdf&quot;, onefile=T)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot(avg_inf_day, main=&quot;Inflammation Scores Over Time&quot;, xlab=&quot;Day&quot;, ylab=&quot;Average_inflammation_score&quot;, pch=4, col=&quot;red&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">boxplot(max_inf_patient, main=&quot;Maximum Inflammation Scores&quot;, ylab=&quot;Max_inf_score&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">legend(&quot;topright&quot;, legend=&quot;n=60 patients&quot;, cex=0.8)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dev.off()</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This will be saved to the current working directory by default so if we
check the folder, a new file named &#x27;Inflammation<!-- -->_<!-- -->plots.pdf&#x27; should have
been created. The onefile=T argument instructs R to append additional
plots to the same file and the <strong>dev.off()</strong> command at the end closes
the file connection. You can also export plots directly to a pdf file
from the RStudio plot panel.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="data-handling">Data Handling<a class="hash-link" href="#data-handling" title="Direct link to heading">​</a></h2><p>We&#x27;ll next read in data from another file to illustrate a few more
features of data frames and how to work with them in R. In this case the
file does contain a header row and the default arguments for read.csv
are appropriate for this file so we only need provide the filename:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">data2 &lt;- read.csv(&quot;sample.csv&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">head(data2)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This displays the first 6 rows, and we can see immediately that we have
a range of different types of data in each column. Let&#x27;s see how R has
treated it (you can paste the following 4 lines as one block).</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">for(i in 1:ncol(data2))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">{</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    print(class(data2[,i]))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Here, we&#x27;ve used a <strong>for</strong> loop to iterate over each column in the
object <strong>data2</strong>, and print to screen the class of each column. The
output tells us that columns 1:3 are treated as factors, column 5 as
numeric and the remaining columns as integer values. We haven&#x27;t yet
mentioned factors and will only briefly discuss them here but they are
very important for statistical analysis in R. They are one-dimensional,
like vectors, and are particularly useful for categorical data.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">length(data2$Group)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">data2$Group</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Each of the 100 entries in the Group column are printed to the screen,
and at the end is the additional information: <em>Levels: Control
Treatment1 Treatment2</em></p><p>These are the unique set of entries in this column, known as the levels
of the factor. You may have come across factors before in the context of
experimental design or ANOVA - in this case the experiment might test
the effect of 2 treatments (Treatment1 and Treatment2) on blood
pressure, perhaps to see if it reduces compared to a control group.
Other information about the patients, such as their age and gender may
be useful to include in the analysis, especially if they are not matched
across the treatment groups.</p><p>R will treat any columns containing character strings (text) as a factor
by default with <strong>read.csv</strong> or <strong>read.table</strong>. We don&#x27;t always want to
do this though, and indeed it is usually preferable to switch this
behaviour off, and specifically convert data we do want to treat as
factors later. This is because factors store data differently and so can
sometimes behave differently to vectors. For example, here the first
column of IDs would preferably be a character vector, as could Gender
unless we needed to include it as an additional explanatory factor in
our analysis model.</p><p>The way to switch off this default behaviour is with the argument
&#x27;stringsAsFactors&#x27; – if you check the help page for <strong>read.csv</strong> again,
you&#x27;ll see it listed among the arguments, and it is TRUE by default
(although it&#x27;s not readily apparent that this is the case).</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">data2 &lt;- read.csv(&quot;sample.csv&quot;, stringsAsFactors=FALSE)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Repeating our loop to check the class of each column, we now see that
the first 3 columns are character vectors rather than factors:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">for(i in 1:ncol(data2))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">{</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    print(class(data2[,i]))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>We can specifically convert the Group column to a factor:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">data2$Group &lt;- as.factor(data2$Group)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">class(data2$Group)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>A very useful summary function is <strong>table</strong>:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">table(data2$Group)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">table(data2$Gender)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This alerts us to the fact that data in the Gender column has not been
entered consistently, which we might have already spotted from viewing
the object in RStudio.</p><p>To fix this, we can make sure F and M are used throughout; this involves
determining which rows contain a lowercase f for example, and
substituting with F. Similarly for the lowercase case m:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">data2$Gender &lt;- gsub(&quot;f&quot;, &quot;F&quot;, data2$Gender)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">data2$Gender &lt;- gsub(&quot;m&quot;, &quot;M&quot;, data2$Gender)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## checking we have modified the data as intended</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">table(data2$Gender)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This last section is just a brief foray into data handling and
manipulation in R, which enables all manner of formatting, editing,
updating or data cleaning tasks to be performed. These are frequently
required before embarking on some analysis, and often take longer too!
It is the topic of the next course in the series, R: Data Handling -
please sign up if interested.</p><p>Well, we have reached the end of today&#x27;s course – we hope it was useful
for you and good luck on your R journey! Please see below for further
reading and information.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="further-resources-and-useful-information">Further Resources and Useful Information<a class="hash-link" href="#further-resources-and-useful-information" title="Direct link to heading">​</a></h2><p>We have helped collate a comprehensive <a href="https://help.it.ox.ac.uk/courses/R" target="_blank" rel="noopener noreferrer">set of R
resources</a> with IT Services (with
particular thanks to Dave Baker for creating the website, and
contributions from Samantha Curle and Andre Python), which lists courses
(online and workshops) as well as recommended textbooks, websites etc.</p><p>The Software Carpentry Foundation website contains many tutorials for
learning a variety of programming languages, including R. There is also
a series of domain-specific Data Carpentry courses, which focus on
computational skills needed to handle and analyse data - tutorials are
currently available for Ecology, Genomics, Geospatial Data and Social
Sciences, with others in development.</p><ul><li><a href="https://software-carpentry.org/lessons/" target="_blank" rel="noopener noreferrer">Software Carpentry
lessons</a></li><li><a href="https://datacarpentry.org/lessons/" target="_blank" rel="noopener noreferrer">Data Carpentry lessons</a></li></ul><p>In particular, the tutorials at the following links give further details
and examples on some of the ideas already introduced or extend to other
topics once you feel comfortable interacting with the R environment.</p><ul><li><a href="http://swcarpentry.github.io/r-novice-inflammation/13-supp-data-structures/index.html" target="_blank" rel="noopener noreferrer">Data
structures</a></li><li><a href="http://swcarpentry.github.io/r-novice-inflammation/11-supp-read-write-csv/index.html" target="_blank" rel="noopener noreferrer">Read and write csv
files</a></li><li><a href="http://swcarpentry.github.io/r-novice-inflammation/12-supp-factors/index.html" target="_blank" rel="noopener noreferrer">Factors</a></li><li><a href="http://swcarpentry.github.io/r-novice-inflammation/10-supp-addressing-data/index.html" target="_blank" rel="noopener noreferrer">Addressing
data</a></li><li><a href="http://swcarpentry.github.io/r-novice-inflammation/06-best-practices-R/index.html" target="_blank" rel="noopener noreferrer">Best
practices</a></li><li><a href="https://datacarpentry.org/R-genomics/" target="_blank" rel="noopener noreferrer">R genomics</a></li><li><a href="https://datacarpentry.org/r-intro-geospatial/" target="_blank" rel="noopener noreferrer">Geospatial data</a></li></ul><h2 class="anchor anchorWithStickyNavbar_mojV" id="getting-help">Getting Help<a class="hash-link" href="#getting-help" title="Direct link to heading">​</a></h2><p>As you read the R help pages, you may not find them terribly helpful.
The R help function is most useful for refreshing your memory about
specific functions you have used before. Unfortunately, it is not very
useful for learning the language itself.</p><p>There are several online forums used by R programmers, novice and
expert, to get help and advice from their peers. Searching for your R
problem will often give results from one of these forums.</p><p>One page &#x27;quick reference&#x27; documents: - <a href="http://www.well.ox.ac.uk/training/R_materials/r-cheat-sheet.pdf" target="_blank" rel="noopener noreferrer">R cheat
sheet</a> -
<a href="http://www.well.ox.ac.uk/training/R_materials/R_reference_card.pdf" target="_blank" rel="noopener noreferrer">R reference
card</a>
Full R manual: - <a href="http://www.well.ox.ac.uk/training/R_materials/R-introduction_manual.pdf" target="_blank" rel="noopener noreferrer">R
manual</a></p><h2 class="anchor anchorWithStickyNavbar_mojV" id="installing-packages">Installing Packages<a class="hash-link" href="#installing-packages" title="Direct link to heading">​</a></h2><p>When freshly installed, R has only its basic functions available. This
is still a considerable number of functions and is adequate for a great
many tasks. R&#x27;s functionality is extended by the use of packages, each
of which is a self-contained bundle of additional functions. These are
typically written by people other than the main R developers, but a
centralised repository of these packages (CRAN) is maintained and
accessible from within R. For example, to install the package called
&#x27;limma&#x27;, you would use:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">install.packages(&quot;limma&quot;)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>which downloads the files needed. You may be asked to choose where to
download the files from, or asked to confirm that they will be installed
to a user directory if you do not have admin permissions for your
computer. Some R packages rely on other R packages. R handles all of
those dependencies in the background and will download and install every
needed package as part of installing the requested one. Installing
packages does not make them available straight away. You need to tell R
to make a package available in your current session using the command:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">library(limma)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>You will need to do this each time you restart R. This step is required
so you don&#x27;t waste memory by loading in packages you don&#x27;t need every
session. Note that the help() function only knows about functions that
have been made available using the library() command.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="bioconductor">Bioconductor<a class="hash-link" href="#bioconductor" title="Direct link to heading">​</a></h2><p><a href="https://www.bioconductor.org/" target="_blank" rel="noopener noreferrer">Bioconductor</a> is a separate third-party
repository of R code, specifically geared towards bioinformatics. Many
of the packages in Bioconductor are also available via CRAN using the
usual package installation method. However, Bioconductor has its own
preferred installation mechanism which gets around some of R&#x27;s more
annoying limitations with version incompatibilities. In particular, the
packages in Bioconductor are updated more swiftly following the release
of a new version of R. To get started with Bioconductor, use the
following command:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">source(&quot;https://bioconductor.org/biocLite.R&quot;)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This loads an R script hosted on the web that defines a new function.
You then use this function in a similar way to how you would use the
install.packages() function:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">biocLite(&quot;limma&quot;)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Loading Bioconductor packages this way is typically a longer process but
more reliable.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/whg-training/whg-training-resources/edit/main/docs/programming/introduction_to_R/R_basic_features_practical.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_dcUD" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_foO9"></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><a class="pagination-nav__link pagination-nav__link--prev" href="/whg-training-resources/programming/introduction_to_R/R_introduction_to_basic_features/"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">R: Introduction to Basic Features</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/whg-training-resources/programming/introduction_to_R/R_introduction_to_basic_features_slides/"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Introduction to R</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_cNA8 thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#practical-tutorial" class="table-of-contents__link toc-highlight">Practical Tutorial</a><ul><li><a href="#setup" class="table-of-contents__link toc-highlight">Setup</a></li><li><a href="#helpful-tips" class="table-of-contents__link toc-highlight">Helpful Tips</a></li></ul></li><li><a href="#reading-in-data-from-a-file" class="table-of-contents__link toc-highlight">Reading in data from a file</a></li><li><a href="#two-dimensional-data-structures" class="table-of-contents__link toc-highlight">Two-dimensional data structures</a></li><li><a href="#accessing-data" class="table-of-contents__link toc-highlight">Accessing Data</a></li><li><a href="#analysing-data" class="table-of-contents__link toc-highlight">Analysing Data</a></li><li><a href="#plotting-data" class="table-of-contents__link toc-highlight">Plotting Data</a></li><li><a href="#data-handling" class="table-of-contents__link toc-highlight">Data Handling</a></li><li><a href="#further-resources-and-useful-information" class="table-of-contents__link toc-highlight">Further Resources and Useful Information</a></li><li><a href="#getting-help" class="table-of-contents__link toc-highlight">Getting Help</a></li><li><a href="#installing-packages" class="table-of-contents__link toc-highlight">Installing Packages</a></li><li><a href="#bioconductor" class="table-of-contents__link toc-highlight">Bioconductor</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">License</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/whg-training-resources/LICENSE/">License</a></li></ul></div><div class="col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2022 University of Oxford.  Built with Docusaurus.</div></div></div></footer></div>
<script src="/whg-training-resources/assets/js/runtime~main.50846bb3.js"></script>
<script src="/whg-training-resources/assets/js/main.1b628545.js"></script>
</body>
</html>