<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-programming/programming_with_gene_annotations/Counting_genes_2">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.20">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.24/dist/katex.min.css" integrity="sha384-odtC+0UGzzFL/6PNoE8rX/SPcQDXBJ+uRepguP4QkPCm2LBxH3FA3y+fKSiJ+AmM" crossorigin="anonymous"><title data-rh="true">Counting genes II: types of protein-coding genes | WHG training resources</title><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_2/"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Counting genes II: types of protein-coding genes | WHG training resources"><meta data-rh="true" name="description" content="Up to table of contents"><meta data-rh="true" property="og:description" content="Up to table of contents"><link data-rh="true" rel="icon" href="/whg-training-resources/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_2/"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_2/" hreflang="en"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_2/" hreflang="x-default"><link rel="stylesheet" href="/whg-training-resources/assets/css/styles.b827ae01.css">
<link rel="preload" href="/whg-training-resources/assets/js/runtime~main.50846bb3.js" as="script">
<link rel="preload" href="/whg-training-resources/assets/js/main.1b628545.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region"><a href="#" class="skipToContent_ZgBM">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/whg-training-resources/"><div class="navbar__logo"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--light_TfLj"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--dark_oUvU"></div><b class="navbar__title text--truncate">Training Resources</b></a><a class="navbar__item navbar__link navbar__link--active" href="/whg-training-resources/overview/">Click here for the tutorials</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a><div class="toggle_S7eR colorModeToggle_vKtC"><button class="clean-btn toggleButton_rCf9 toggleButtonDisabled_Pu9x" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_v35p"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_nQuB"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_dLyj"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docsWrapper_mKqt"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_RiI4" type="button"></button><div class="docPage_ualW"><aside class="theme-doc-sidebar-container docSidebarContainer_UQUJ"><div class="sidebar_RiAD"><nav class="menu thin-scrollbar menu_izAj"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/overview/">Overview of the tutorials</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/prerequisites/">Getting your environment setup</a><button aria-label="Toggle the collapsible sidebar category &#x27;Getting your environment setup&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/bioinformatics/">Bioinformatics</a><button aria-label="Toggle the collapsible sidebar category &#x27;Bioinformatics&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" href="/whg-training-resources/programming/">Programming</a><button aria-label="Toggle the collapsible sidebar category &#x27;Programming&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/programming/introduction_to_R/">Introduction to R</a><button aria-label="Toggle the collapsible sidebar category &#x27;Introduction to R&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/">Programming with gene annotations</a><button aria-label="Toggle the collapsible sidebar category &#x27;Programming with gene annotations&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Introduction/">Introduction</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like/">What gene annotation data looks like</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code/">Processing GFF files</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/making_a_module/">Making a module</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Refactoring_makes_code_better/">An aside on refactoring</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Converting_gff_to_sqlite/">Writing a useful conversion program.</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/exploring_gencode/">Investigating human genes</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_1/">Counting genes I</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Memory_issues_and_how_to_solve_them/">Memory issues and how to solve them</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_2/">Counting genes II: types of protein-coding genes</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_3/">Extremely big, small and complex genes</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Getting_sequence_lengths/">Getting sequence lengths</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes/">How much of the genome is in genes?</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Scaling_up/">Scaling up for a real analysis</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Where_next/">What next?</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Closing_thoughts/">Closing thoughts</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/List_of_ensembl_files/">Appendix: a list of Ensembl GFF files.</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Visualisation/">Appendix: visualisation</a></li></ul></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/data_visualisation/">Data visualisation</a><button aria-label="Toggle the collapsible sidebar category &#x27;Data visualisation&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/next_generation_sequencing/">Next-generation sequencing</a><button aria-label="Toggle the collapsible sidebar category &#x27;Next-generation sequencing&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/statistical_modelling/">Statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;Statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/population_genetics/">Population genetics</a><button aria-label="Toggle the collapsible sidebar category &#x27;Population genetics&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/genome_wide_association_studies/">Genome-wide association analysis</a><button aria-label="Toggle the collapsible sidebar category &#x27;Genome-wide association analysis&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/LICENSE/">LICENSE</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/contributors/">contributors</a></li></ul></nav></div></aside><main class="docMainContainer_uL0j"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_DM6M"><div class="docItemContainer_vinB"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Xlws" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/whg-training-resources/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_kU5B"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/programming/"><span itemprop="name">Programming</span></a><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/programming/programming_with_gene_annotations/"><span itemprop="name">Programming with gene annotations</span></a><meta itemprop="position" content="2"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">Counting genes II: types of protein-coding genes</span><meta itemprop="position" content="3"></li></ul></nav><div class="tocCollapsible_bZGK theme-doc-toc-mobile tocMobile_TmEX"><button type="button" class="clean-btn tocCollapsibleButton_l22C">On this page</button></div><div class="theme-doc-markdown markdown"><h1>Counting genes II: types of protein-coding genes</h1><p><a href="/whg-training-resources/programming/programming_with_gene_annotations/">Up to table of contents</a></p><p>If you&#x27;ve followed so far you will have code <code>gff.py</code> that can parse a GFF file, and in the process
will pull out certain fields from the <code>attributes</code> column. This includes the <code>ID</code> attribute, the
<code>Parent</code> attribute that says how records are linked, and also the <code>biotype</code> column and maybe also
the <code>Name</code> (which contains gene names). In the <a href="http://ftp.ensembl.org/pub/current_gff3/" target="_blank" rel="noopener noreferrer">Ensembl
files</a> the <code>biotype</code> is useful because it tells us what
kind of genes they are.</p><p>If you haven&#x27;t already, please download the GFF files for some different species and run them
through your <a href="/whg-training-resources/programming/programming_with_gene_annotations/Converting_gff_to_sqlite/">sqlite conversion program</a> to get them into the
database. I&#x27;m going to assume you followed the suggestion to add a column that distinguishes the
different species - in my code this is called <code>analysis</code> and I&#x27;ll use that below.</p><p>(For reference my version of the code is at
<a target="_blank" href="/whg-training-resources/assets/files/gff_to_sqlite-c776051fcd5741fec1c416efbdd418da.py/">solutions/part2/gff_to_sqlite.py</a>
and
<a target="_blank" href="/whg-training-resources/assets/files/gff-6953e4a4db02bd87c9b8eae2a8a48610.py/">solutions/part2/gff.py</a> - feel free to run that
instead, if needed.)</p><p>This sqlite file is now in a good shape to start really exploring genes. First let&#x27;s look at what <code>biotypes</code> there are.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="what-types-of-protein-coding-gene-are-there">What types of protein-coding gene are there?<a class="hash-link" href="#what-types-of-protein-coding-gene-are-there" title="Direct link to heading">​</a></h2><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">sqlite&gt; .mode column</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sqlite&gt; .header on</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sqlite&gt; .width 50 25 25</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sqlite&gt; SELECT analysis, biotype, COUNT(*) FROM gff_data WHERE type==&#x27;gene&#x27; GROUP BY analysis, biotype ;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>With the current data I have this gives:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">analysis                                            biotype                    COUNT(*)                 </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">--------------------------------------------------  -------------------------  -------------------------</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Acanthochromis_polyacanthus.ASM210954v1.104         IG_J_gene                  2                        </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Acanthochromis_polyacanthus.ASM210954v1.104         IG_V_gene                  4                        </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Acanthochromis_polyacanthus.ASM210954v1.104         TR_J_gene                  5                        </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Acanthochromis_polyacanthus.ASM210954v1.104         protein_coding             24016                    </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Camelus_dromedarius.CamDro2.104.chr.gff3            IG_C_gene                  1                        </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Camelus_dromedarius.CamDro2.104.chr.gff3            IG_V_gene                  13                       </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Camelus_dromedarius.CamDro2.104.chr.gff3            TR_C_gene                  1                        </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Camelus_dromedarius.CamDro2.104.chr.gff3            TR_J_gene                  5                        </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Camelus_dromedarius.CamDro2.104.chr.gff3            TR_V_gene                  3                        </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Camelus_dromedarius.CamDro2.104.chr.gff3            protein_coding             18896                    </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Gallus_gallus.GRCg6a.104                            IG_V_gene                  98                       </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Gallus_gallus.GRCg6a.104                            protein_coding             16568                    </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Homo_sapiens.GRCh38.104                             IG_C_gene                  14                       </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Homo_sapiens.GRCh38.104                             IG_D_gene                  37                       </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Homo_sapiens.GRCh38.104                             IG_J_gene                  18                       </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Homo_sapiens.GRCh38.104                             IG_V_gene                  145                      </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Homo_sapiens.GRCh38.104                             TEC                        1056                     </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Homo_sapiens.GRCh38.104                             TR_C_gene                  6                        </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Homo_sapiens.GRCh38.104                             TR_D_gene                  4                        </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Homo_sapiens.GRCh38.104                             TR_J_gene                  79                       </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Homo_sapiens.GRCh38.104                             TR_V_gene                  106                      </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Homo_sapiens.GRCh38.104                             polymorphic_pseudogene     49                       </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Homo_sapiens.GRCh38.104                             protein_coding             19937                    </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Mus_musculus.GRCm39.104                             IG_C_gene                  13                       </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Mus_musculus.GRCm39.104                             IG_D_gene                  19                       </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Mus_musculus.GRCm39.104                             IG_J_gene                  14                       </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Mus_musculus.GRCm39.104                             IG_LV_gene                 4                        </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Mus_musculus.GRCm39.104                             IG_V_gene                  218                      </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Mus_musculus.GRCm39.104                             TEC                        3238                     </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Mus_musculus.GRCm39.104                             TR_C_gene                  8                        </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Mus_musculus.GRCm39.104                             TR_D_gene                  4                        </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Mus_musculus.GRCm39.104                             TR_J_gene                  70                       </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Mus_musculus.GRCm39.104                             TR_V_gene                  144                      </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Mus_musculus.GRCm39.104                             polymorphic_pseudogene     89                       </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Mus_musculus.GRCm39.104                             protein_coding             21834                    </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Pan_troglodytes.Pan_tro_3.0.104.chr                 IG_C_gene                  11                       </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Pan_troglodytes.Pan_tro_3.0.104.chr                 IG_V_gene                  91                       </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Pan_troglodytes.Pan_tro_3.0.104.chr                 TR_C_gene                  9                        </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Pan_troglodytes.Pan_tro_3.0.104.chr                 TR_V_gene                  66                       </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Pan_troglodytes.Pan_tro_3.0.104.chr                 protein_coding             21879                    </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">PlasmoDB-54_Pfalciparum3D7                          protein_coding_gene        5318                     </span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The various biotypes used by Ensembl are <a href="https://m.ensembl.org/info/genome/genebuild/biotypes.html" target="_blank" rel="noopener noreferrer">documented
here</a>. As the above shows, the vast
majority of genes in the file are protein-coding genes. In each organism, however a few genes are
listed as <code>polymorphic_pseudogene</code>. You can <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3491395/" target="_blank" rel="noopener noreferrer">read more about pseudogenes
here</a>. They are genes that are non-coding in
the reference sequence (due to a disabling mutation), but still code for proteins in some
individuals.</p><p>The <code>IG_</code> and <code>TR_</code> biotypes are also interesting: they are the &#x27;constant&#x27;, &#x27;joining&#x27;, and
&#x27;variable&#x27; gene segments of immunuglobulin and T cell receptors. They do encode proteins, but via a
yet more <a href="https://en.wikipedia.org/wiki/V(D)J_recombination" target="_blank" rel="noopener noreferrer">complex process that involves somatic recombination to assemble the mature genes in B and
T cells</a>. These gene segments also lie in
regions that are <a href="https://doi.org/10.1371/journal.pcbi.1009254" target="_blank" rel="noopener noreferrer">especially complex</a>. </p><p>(What are the <code>TEC</code> genes?)</p><p>However, the vast majority of the genes listed have type <code>protein_coding</code>, and we will focus on
these in the rest of this tutorial.</p><p><em>Note.</em> Here is similar code to count genes across species in python:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">import pandas, sqlite3</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">db = sqlite3.connect( &quot;genes.sqlite&quot; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">data = pandas.read_sql( &quot;SELECT * FROM gff_data WHERE type IN ( &#x27;gene&#x27; )&quot;, db )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">data.groupby( [ &#x27;analysis&#x27;, &#x27;biotype&#x27; ] ).size()</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><h2 class="anchor anchorWithStickyNavbar_mojV" id="what-do-genes-look-like">What do genes look like?<a class="hash-link" href="#what-do-genes-look-like" title="Direct link to heading">​</a></h2><p><strong>Challenge.</strong> Pick a gene (say human ABO or FUT2) and investigate in detail using your file. (If you&#x27;ve followed so
far this will be in the <code>Name</code> column.) How many transcripts does it have? (Hint: find records with <code>Parent</code> equal to
the gene <code>ID</code>. (Are there any other records with the gene as parent, that aren&#x27;t transcripts? What are they?). What is
the <code>transcript_support_level</code> for transcripts, and what does that mean? Pick a transcript and look at its sub-records
(again by matching the <code>Parent`` to the transcript </code>ID`). What types are they apart from exons? Can you see how this
data compares to the representation on <a href="http://www.ensembl.org/" target="_blank" rel="noopener noreferrer">Ensembl</a> or on
the <a href="https://genome.ucsc.edu" target="_blank" rel="noopener noreferrer">UCSC genome browser</a>?</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="what-are-all-those-species-anyway">What are all those species anyway?<a class="hash-link" href="#what-are-all-those-species-anyway" title="Direct link to heading">​</a></h2><p>If like me you&#x27;re a bit unclear about what <a href="http://ftp.ensembl.org/pub/current_gff3/" target="_blank" rel="noopener noreferrer">all those
species</a> are, now might be a good time to go and look at
the <a href="http://www.onezoom.org" target="_blank" rel="noopener noreferrer">OneZoom Tree of Life explorer</a>. This will tell you, for example, that
<em>Acanthochromis polyacanthus</em> is a member of the <a href="http://www.onezoom.org/life/@Ovalentaria=5553750?img=best_any&amp;anim=flight#x1307,y908,w1.5714" target="_blank" rel="noopener noreferrer">Sticky Eggs
Group</a>
that also includes Cichlids, Silversides and Guppies, or that <em>mus musculus</em> is one of 37 species
collectively known as &#x27;house mice&#x27;.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="how-complicated-are-protein-coding-genes">How complicated are protein-coding genes?<a class="hash-link" href="#how-complicated-are-protein-coding-genes" title="Direct link to heading">​</a></h2><p>Let&#x27;s count how many transcripts each gene has, and how many exons.</p><div class="admonition admonition-tip alert alert--success"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="16" viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Challenge</h5></div><div class="admonition-content"><p>Write python code that reads in the appropriate data from <code>gff_data</code> table, and for
each gene counts i. the number of transcripts and ii. the average number of exons (averaged over
the transcripts for that gene).</p></div></div><p><strong>Hints.</strong></p><p>To do this is a bit more involved than previous steps. You need to find a way to join the
transcript records to the genes, and the exon records to the transcripts. Here is one way:</p><ul><li><p>Start by loading just the genes, transcripts, and exons into seperate data frames.</p></li><li><p>Write a function <code>count_exons_per_transcript()</code> that takes the transcripts and exons, and
returns a dataframe of transcripts with a column that reports the number of exons.</p></li><li><p>Then similarly write a function <code>summarise_transcripts_per_gene()</code> that takes the genes and the
above summarised transcripts, and returns a dataframe of transcripts with a column for the number
of transcripts and a column for the average number of exons.</p></li><li><p>There are a couple of ways to implement these functions. One way is to iterate through dataframes
and build python data structures that capture the hierarchy of genes, transcripts, and exons.
E.g. the first function might return something like this:</p></li></ul><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">transcript_summary = [</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      &quot;ID&quot;:&quot;transcript:ENST00000641515&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      &quot;exons&quot;: [</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         &quot;ENSE00003812156&quot;, &quot;ENSE00003813641&quot;, &quot;ENSE00003813949&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      ]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   },</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   ...</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>and the second function something like this:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">gene_summary = {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   &quot;gene:ENSG00000186092&quot;: {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      &quot;number_of_transcripts&quot;: 1,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      &quot;average_number_of_exons&quot;: 3,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      &quot;transcripts&quot;: [</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         # array of objects recording transcripts and their exons</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            &quot;ID&quot;:&quot;transcript:ENST00000641515&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            &quot;exons&quot;: [</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">               &quot;ENSE00003812156&quot;, &quot;ENSE00003813641&quot;, &quot;ENSE00003813949&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            ]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         }</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      ]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   },</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   ...</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>You would then iterate through the second object to compute the summaries.</p><ul><li>A slicker way to implement this is to use &#x27;data joins&#x27; and &#x27;group by&#x27; operations. In pandas the
functions to use are the <a href="https://pandas.pydata.org/docs/user_guide/merging.html" target="_blank" rel="noopener noreferrer"><code>merge()
function</code></a>) and the
<a href="https://pandas.pydata.org/docs/user_guide/groupby.html" target="_blank" rel="noopener noreferrer"><code>groupby()</code></a> function. A simple version
of this is:</li></ul><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">transcripts_and_exons = pandas.merge(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    transcripts,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    exons[ [&quot;ID&quot;, &quot;Parent&quot; ]],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    how = &quot;outer&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    left_on = &quot;ID&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    right_on = &quot;Parent&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">transcript_summary = transcripts_and_exons.groupby( [&#x27;ID_x&#x27;, &#x27;Parent_x&#x27;] ).count()</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This joins the transcripts to the exon IDs, and then groups the result by transcript to count the rows.</p><p>However the above has a few issues. First, gets the names wrong - &quot;ID_x&quot;, &quot;Parent_x&quot; and so on, and
the third column does not even have a name! It is always good to have the right names. The
<code>.rename()</code> option and the &#x27;Named aggregation&#x27; syntax described on the <a href="https://pandas.pydata.org/docs/user_guide/groupby.html" target="_blank" rel="noopener noreferrer"><code>groupby()</code>
page</a> can be used to fix this. Something like this:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">transcripts_and_exons.rename(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   columns = { &#x27;ID_x&#x27;: &#x27;ID&#x27;, &#x27;Parent_x&#x27;: &#x27;Parent&#x27;, &#x27;ID_y&#x27;: &#x27;exon_ID&#x27;, &#x27;Parent_y&#x27;, &#x27;exon_Parent&#x27; },</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   inplace = True</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">def count( x ):</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   return len(x)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">transcript_summary = (</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   transcripts_and_exons</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      .groupby( [&#x27;ID&#x27;, &#x27;Parent&#x27;] )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      .agg(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         number_of_exons = pandas.NamedAgg(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            column = &quot;exon_ID&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            aggfunc = count</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>If you run this you will see it has &quot;ID&quot;, &quot;Parent&quot; and &quot;number_of_exons&quot; columns. </p><p>More seriously, the above code has a possible bug. (You did test it, right?) Specifically it gets
the answer wrong if a transcript has no exons. How did I discover this? <a target="_blank" href="/whg-training-resources/assets/files/test_gff-c303bb1903bbef04e3df3bddba38f7ef.py/">By writing a test</a>.  So that needs to be fixed too.  Good luck! </p><p><strong>Note.</strong> My version of the code can be found in
<a href="https://github.com/whg-training/whg-training-resources/blob/main/docs/programming/programming_with_gene_annotations/solutions/part2" target="_blank" rel="noopener noreferrer">this folder</a> in the <code>gene_summary.py</code> file (or <a href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_2/solutions/part2/gene_summary.py]/">download directly</a>
The file implements both a pandas version (<code>summarise_genes()</code>) and a python datastructure version (<code>summarise_genes_python_version()</code>).</p><p><strong>Note.</strong> which of these approaches do you find easier to understand? The python version of my code is definitely
longer, but neither seems especially simple. However, the <code>count_exons_per_transcript()</code> and
<code>summarise_transcripts_per_gene()</code> functions are pretty similar and so they are good candidates for a refactor - so I
think this code can be improved...</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="a-sqlite-approach">A sqlite approach<a class="hash-link" href="#a-sqlite-approach" title="Direct link to heading">​</a></h2><p>In the rest of this section I&#x27;ll show how you could solve those problems in the sqlite database
itself - again using data joins and group by operations.</p><p>First, it&#x27;s convenient to make some views of the data that just reflect the genes, transcripts and
exons:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">CREATE VIEW gene_view AS SELECT * FROM gff_data WHERE type == &#x27;gene&#x27; ;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">CREATE VIEW transcript_view AS SELECT * FROM gff_data WHERE type == &#x27;mRNA&#x27; ;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">CREATE VIEW exon_view AS SELECT * FROM gff_data WHERE type == &#x27;exon&#x27; ;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Second, to make lookups efficient, we&#x27;ll also need to make sure the <code>ID</code> field is indexed:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">CREATE INDEX gff_data_ID_index ON gff_data( ID ) ;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>(This can take a minute or so as it builds the index).</p><p>Now we can join up the data to make the counts.  First we&#x27;ll count exons in transcripts:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">CREATE VIEW transcript_summary_view AS</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">SELECT T.*, COUNT(E.Parent) AS number_of_exons</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">FROM transcript_view T</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">LEFT JOIN exon_view E ON E.Parent == T.ID</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">GROUP BY T.ID;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>If you look at the result:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">SELECT * FROM transcript_summary_view LIMIT 10 ;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>you&#x27;ll see it takes a little while to run its computation, and it has one row per transcript, with
the number of exons in the last column.</p><p><strong>Note.</strong> As with python in principle we have to be careful what we count above. The <code>COUNT(E.Parent)</code> makes sure that
we count only rows that correspond to an actual exon (rather than a transcript with no exons, which should get a count
of zero. (There shouldn&#x27;t really be any transcripts like this - otherwise it isn&#x27;t really a transcript - but it is
worth making sure. Are there any?)</p><p>Next we can summarise transcripts for each gene:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">CREATE VIEW gene_summary_view AS</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">SELECT</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    gene_view.*,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    COUNT(T.Parent) AS number_of_transcripts,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    CAST( SUM(number_of_exons) AS FLOAT ) / COUNT(T.ID) AS average_number_of_exons</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">FROM gene_view</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">LEFT JOIN transcript_summary_view T ON T.Parent == gene_view.ID</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">GROUP BY gene_view.ID;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This solves our problem:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">SELECT ID, Name, biotype, number_of_transcripts, average_number_of_exons  FROM gene_summary_view LIMIT 10;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><strong>Note.</strong> Using <code>.mode line</code> or <code>.mode column</code> and <code>.header on</code> is useful in sqlite3. You can also
just run these queries and load them into python as described above.</p><p>Unfortunately these queries are taking... quite a while. In fact, we are asking sqlite to do quite
a bit of work here, largely because we&#x27;ve built this on dynamic views of the original <code>gff_data</code>
table - which is very big and includes plenty that is not relevant to our query. To speed up our
queries (in a way that mimics what we would do in python) we can &#x27;bake&#x27; the relevant view into
their own tables first:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">CREATE TABLE genes AS SELECT analysis, ID, Parent, Name, biotype, seqid, start, end, strand FROM gene_view ;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">CREATE TABLE transcript_summary AS SELECT analysis, ID, Parent, Name, biotype, seqid, start, end, strand, number_of_exons FROM transcript_summary_view ;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">CREATE INDEX transcript_summary_Parent_INDEX ON transcript_summary( Parent ) ;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>and then rewrite <code>gene_summary_view</code> to use these tables:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">DROP VIEW gene_summary_view ;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">CREATE VIEW gene_summary_view AS</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">SELECT</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    genes.*,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    COUNT(T.ID) AS number_of_transcripts,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    CAST( SUM(number_of_exons) AS FLOAT ) / COUNT(T.ID) AS average_number_of_exons</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">FROM genes</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">LEFT JOIN transcript_summary T ON T.Parent == genes.ID</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">GROUP BY genes.ID;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>On my system, with the data above this brings query time down to a few seconds.</p><p><strong>Note.</strong> The <code>CAST( .. AS FLOAT )</code> syntax is needed above to make sure sqlite uses floating-point arithmetic for the division.</p><p>Let&#x27;s use this data to <a href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_3/">find some extreme genes</a>.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/whg-training/whg-training-resources/edit/main/docs/programming/programming_with_gene_annotations/Counting_genes_2.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_dcUD" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_foO9"></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><a class="pagination-nav__link pagination-nav__link--prev" href="/whg-training-resources/programming/programming_with_gene_annotations/Memory_issues_and_how_to_solve_them/"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Memory issues and how to solve them</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_3/"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Extremely big, small and complex genes</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_cNA8 thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#what-types-of-protein-coding-gene-are-there" class="table-of-contents__link toc-highlight">What types of protein-coding gene are there?</a></li><li><a href="#what-do-genes-look-like" class="table-of-contents__link toc-highlight">What do genes look like?</a></li><li><a href="#what-are-all-those-species-anyway" class="table-of-contents__link toc-highlight">What are all those species anyway?</a></li><li><a href="#how-complicated-are-protein-coding-genes" class="table-of-contents__link toc-highlight">How complicated are protein-coding genes?</a></li><li><a href="#a-sqlite-approach" class="table-of-contents__link toc-highlight">A sqlite approach</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">License</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/whg-training-resources/LICENSE/">License</a></li></ul></div><div class="col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2022 University of Oxford.  Built with Docusaurus.</div></div></div></footer></div>
<script src="/whg-training-resources/assets/js/runtime~main.50846bb3.js"></script>
<script src="/whg-training-resources/assets/js/main.1b628545.js"></script>
</body>
</html>