<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-programming/programming_with_gene_annotations/Counting_genes_1">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.20">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.24/dist/katex.min.css" integrity="sha384-odtC+0UGzzFL/6PNoE8rX/SPcQDXBJ+uRepguP4QkPCm2LBxH3FA3y+fKSiJ+AmM" crossorigin="anonymous"><title data-rh="true">Counting genes I | WHG training resources</title><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_1/"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Counting genes I | WHG training resources"><meta data-rh="true" name="description" content="Some first statistics"><meta data-rh="true" property="og:description" content="Some first statistics"><link data-rh="true" rel="icon" href="/whg-training-resources/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_1/"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_1/" hreflang="en"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_1/" hreflang="x-default"><link rel="stylesheet" href="/whg-training-resources/assets/css/styles.b827ae01.css">
<link rel="preload" href="/whg-training-resources/assets/js/runtime~main.50846bb3.js" as="script">
<link rel="preload" href="/whg-training-resources/assets/js/main.1b628545.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region"><a href="#" class="skipToContent_ZgBM">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/whg-training-resources/"><div class="navbar__logo"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--light_TfLj"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--dark_oUvU"></div><b class="navbar__title text--truncate">Training Resources</b></a><a class="navbar__item navbar__link navbar__link--active" href="/whg-training-resources/overview/">Click here for the tutorials</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a><div class="toggle_S7eR colorModeToggle_vKtC"><button class="clean-btn toggleButton_rCf9 toggleButtonDisabled_Pu9x" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_v35p"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_nQuB"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_dLyj"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docsWrapper_mKqt"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_RiI4" type="button"></button><div class="docPage_ualW"><aside class="theme-doc-sidebar-container docSidebarContainer_UQUJ"><div class="sidebar_RiAD"><nav class="menu thin-scrollbar menu_izAj"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/overview/">Overview of the tutorials</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/prerequisites/">Getting your environment setup</a><button aria-label="Toggle the collapsible sidebar category &#x27;Getting your environment setup&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/bioinformatics/">Bioinformatics</a><button aria-label="Toggle the collapsible sidebar category &#x27;Bioinformatics&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" href="/whg-training-resources/programming/">Programming</a><button aria-label="Toggle the collapsible sidebar category &#x27;Programming&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/programming/introduction_to_R/">Introduction to R</a><button aria-label="Toggle the collapsible sidebar category &#x27;Introduction to R&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/">Programming with gene annotations</a><button aria-label="Toggle the collapsible sidebar category &#x27;Programming with gene annotations&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Introduction/">Introduction</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like/">What gene annotation data looks like</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code/">Processing GFF files</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/making_a_module/">Making a module</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Refactoring_makes_code_better/">An aside on refactoring</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Converting_gff_to_sqlite/">Writing a useful conversion program.</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/exploring_gencode/">Investigating human genes</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_1/">Counting genes I</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Memory_issues_and_how_to_solve_them/">Memory issues and how to solve them</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_2/">Counting genes II: types of protein-coding genes</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_3/">Extremely big, small and complex genes</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Getting_sequence_lengths/">Getting sequence lengths</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes/">How much of the genome is in genes?</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Scaling_up/">Scaling up for a real analysis</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Where_next/">What next?</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Closing_thoughts/">Closing thoughts</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/List_of_ensembl_files/">Appendix: a list of Ensembl GFF files.</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Visualisation/">Appendix: visualisation</a></li></ul></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/data_visualisation/">Data visualisation</a><button aria-label="Toggle the collapsible sidebar category &#x27;Data visualisation&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/next_generation_sequencing/">Next-generation sequencing</a><button aria-label="Toggle the collapsible sidebar category &#x27;Next-generation sequencing&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/statistical_modelling/">Statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;Statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/population_genetics/">Population genetics</a><button aria-label="Toggle the collapsible sidebar category &#x27;Population genetics&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/genome_wide_association_studies/">Genome-wide association analysis</a><button aria-label="Toggle the collapsible sidebar category &#x27;Genome-wide association analysis&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/LICENSE/">LICENSE</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/contributors/">contributors</a></li></ul></nav></div></aside><main class="docMainContainer_uL0j"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_DM6M"><div class="docItemContainer_vinB"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Xlws" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/whg-training-resources/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_kU5B"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/programming/"><span itemprop="name">Programming</span></a><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/programming/programming_with_gene_annotations/"><span itemprop="name">Programming with gene annotations</span></a><meta itemprop="position" content="2"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">Counting genes I</span><meta itemprop="position" content="3"></li></ul></nav><div class="tocCollapsible_bZGK theme-doc-toc-mobile tocMobile_TmEX"><button type="button" class="clean-btn tocCollapsibleButton_l22C">On this page</button></div><div class="theme-doc-markdown markdown"><h1>Counting genes I</h1><h2 class="anchor anchorWithStickyNavbar_mojV" id="some-first-statistics">Some first statistics<a class="hash-link" href="#some-first-statistics" title="Direct link to heading">​</a></h2><p><a href="/whg-training-resources/programming/programming_with_gene_annotations/">Up to table of contents</a></p><p>For simplicity, from here on in we will work with the GFF3 files from the <a href="http://ftp.ensembl.org/pub/current_gff3/" target="_blank" rel="noopener noreferrer">Ensembl ftp
site</a> - also those from
<a href="https://plasmodb.org/plasmo/app/downloads/Current_Release/" target="_blank" rel="noopener noreferrer">PlasmoDB</a>. These use the terminology <code>mRNA</code> for a
gene transcript, and they also have the genome sequence lengths written in the metadata, making life easy for
us. Before starting, download some of these files now.  A guide to the species can be found in the <a href="https://www.ensembl.org/info/about/species.html" target="_blank" rel="noopener noreferrer">Ensembl species list</a>
and the <a href="https://www.ensembl.org/info/about/speciestree.html" target="_blank" rel="noopener noreferrer">Ensembl species tree</a>.</p><div class="admonition admonition-tip alert alert--success"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="16" viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Note</h5></div><div class="admonition-content"><p>If you want a warmup on pandas, try <a href="/whg-training-resources/programming/programming_with_gene_annotations/exploring_gencode/">exploring the GENCODE annotations</a> first. But we&#x27;ll switch to using
the Ensembl files from now on.</p></div></div><p>We will focus on protein-coding genes, and their transcripts, exons and coding sequence. They have type <code>gene</code>, <code>mRNA</code>, <code>exon</code>
and <code>CDS</code> in the files Ensembl respectively. They come in a basic hierarchy:</p><ul><li><p>Each gene can have multiple transcripts (i.e. multiple expressed forms at mRNA level - e.g. they
might differ in how exons are spliced together, or have different transcription start or end
sites).</p></li><li><p>Each transcript is made up of one or more exons. The introns in between are spliced out. (This
happens during the transcription process.)</p></li><li><p>Of the exon sequence that makes it into a transcript, only a portion is actually translated into
the mature protein during processing by the ribosomes. This is the <em>coding sequence</em> and the rest
is <em>untranslated sequence</em>.</p></li></ul><p>The files we&#x27;re looking are (roughly speaking) humanities&#x27; best guess at what this picture looks
like in each organism.</p><p>We&#x27;re actually in a good shape for many of <a href="/whg-training-resources/programming/programming_with_gene_annotations/Introduction/">our scientific questions</a>. Let&#x27;s start
by counting genes:</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="how-many-genes-are-there">How many genes are there?<a class="hash-link" href="#how-many-genes-are-there" title="Direct link to heading">​</a></h3><p>Let&#x27;s start with counting genes - easy! One way is to use sqlite. For example using the sqlite file
you created in the last step:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">sqlite&gt; SELECT COUNT(*) FROM gff_data WHERE type == `gene` ;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><em>Note.</em> I ran the above command interactively in sqlite3 - you get there by typing <code>sqlite3 genes.sqlite</code> in the shell.
Type <code>.quit</code> to quit.)</p><p>If you followed the suggestions, or used my version of the code, your file will also have an <code>analysis</code> column
that lets you differentiate the records for different species. So you can make these counts for multiple species:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">sqlite&gt; .mode column</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sqlite&gt; .header on</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sqlite&gt; SELECT analysis, COUNT(*) FROM gff_data WHERE type==&#x27;gene&#x27; GROUP BY analysis ;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>In my data, which includes spiny chromis, dromedary camels, red junglefowl, humans, mice, chimpanzees, and
malaria parasites, this gives:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">analysis                                     COUNT(*)  </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">-------------------------------------------  ----------</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Acanthochromis_polyacanthus.ASM210954v1.104  24027     </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Camelus_dromedarius.CamDro2.104.chr.gff3     18919     </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Gallus_gallus.GRCg6a.104                     16666     </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Homo_sapiens.GRCh38.104                      21451     </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Mus_musculus.GRCm39.104                      25655     </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Pan_troglodytes.Pan_tro_3.0.104.chr          22056     </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">PlasmoDB-54_Pfalciparum3D7                   5318      </span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><div class="admonition admonition-tip alert alert--success"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="16" viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Note</h5></div><div class="admonition-content"><p>To make the above query work the <a href="https://plasmodb.org/plasmo/app/downloads/Current_Release/" target="_blank" rel="noopener noreferrer"><em>P.falciparum</em> data from
PlasmoDB</a>, I kludged it by running this sql:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">UPDATE genes</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">SET type = &#x27;gene&#x27;, biotype = &#x27;protein_coding_gene&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">WHERE analysis == &#x27;PlasmoDB-54_Pfalciparum3D7&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">AND type == &#x27;protein_coding_gene&#x27; ;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>I wouldn&#x27;t generally recommend this type of manually-fix-the-data-until-it-works approach
(not least because it would have to be re-done every time we imported new data) but it&#x27;ll do for this tutorial.</p></div></div><p>You can also do this type of thing pretty easily in python or R. Here is the same &#x27;group by&#x27; operation done using
pandas or dplyr:</p><div class="tabs-container tabList_uSqn"><ul role="tablist" aria-orientation="horizontal" class="tabs"><li role="tab" tabindex="0" aria-selected="true" class="tabs__item tabItem_LplD tabs__item--active">Python code</li><li role="tab" tabindex="-1" aria-selected="false" class="tabs__item tabItem_LplD">R code</li></ul><div class="margin-top--md"><div role="tabpanel" class="tabItem_OmH5"><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">import pandas, sqlite3</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">db = sqlite3.connect( &quot;genes.sqlite&quot; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">genes = pandas.read_sql( &quot;SELECT * FROM gff_data WHERE type == &#x27;gene&#x27;&quot;, db )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    genes</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    .groupby( &quot;analysis&quot; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    .size()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div></div><div role="tabpanel" class="tabItem_OmH5" hidden=""><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">library( RSQLite )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library( dplyr )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">db = dbConnect( dbDriver( &quot;SQLite&quot; ), &quot;genes.sqlite&quot; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">genes = dbGetQuery( db, &quot;SELECT * FROM gff_data WHERE type == &#x27;gene&#x27;&quot; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    genes</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    %&gt;% group_by( analysis )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    %&gt;% summarise( count = n() )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div></div></div></div><p>This produces:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">analysis</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Acanthochromis_polyacanthus.ASM210954v1.104    24027</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Camelus_dromedarius.CamDro2.104.chr.gff3       18919</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Gallus_gallus.GRCg6a.104                       16666</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Homo_sapiens.GRCh38.104                        21451</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Mus_musculus.GRCm39.104                        25655</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Pan_troglodytes.Pan_tro_3.0.104.chr            22056</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">PlasmoDB-54_Pfalciparum3D7                      5318</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dtype: int64</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Interestingly, both <a href="https://en.wikipedia.org/wiki/House_mouse" target="_blank" rel="noopener noreferrer">house mice</a> and <a href="https://en.wikipedia.org/wiki/Spiny_chromis" target="_blank" rel="noopener noreferrer">spiny
chromis</a> have more (annotated) genes than humans.
Chimpanzees have a similar number, while <a href="https://en.wikipedia.org/wiki/Red_junglefowl" target="_blank" rel="noopener noreferrer">Red
junglefowl</a> and <a href="https://en.wikipedia.org/wiki/Dromedary" target="_blank" rel="noopener noreferrer">Dromedary
Camels</a> have respectively 17% and 5% fewer (annotated)
protein-coding genes than humans. <em>Plasmodium falciparum</em> has about a quarter of the number of
genes. (But that&#x27;s still pretty impressive because its genome size is less than a hundredth that of
humans.)</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="what-are-all-these-species">What are all these species?<a class="hash-link" href="#what-are-all-these-species" title="Direct link to heading">​</a></h3><p>If you&#x27;re wondering that - it&#x27;s time to look at <a href="http://www.onezoom.org" target="_blank" rel="noopener noreferrer">OneZoom.org</a> or the <a href="https://www.ensembl.org/info/about/speciestree.html" target="_blank" rel="noopener noreferrer">Ensembl species
tree</a>. Did you know that Dolphins, like
Camels, are <a href="http://www.onezoom.org/life/@Cetartiodactyla=622916" target="_blank" rel="noopener noreferrer">cloven-hoofed ungulates</a>?</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="what-kinds-of-gene-are-there">What kinds of gene are there?<a class="hash-link" href="#what-kinds-of-gene-are-there" title="Direct link to heading">​</a></h3><p>The above counts all &quot;genes&quot;.  But what exactly are these?</p><p>One way to figure this out is by comparing to the other records in there. You can find all the top-level record
types (those with no <code>Parent</code>) in the file like this:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">import pandas, sqlite3</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">db = sqlite3.connect( &quot;genes.sqlite&quot; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">types = pandas.read_sql(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    &quot;SELECT analysis, type, COUNT(*) FROM gff_data WHERE Parent IS NULL GROUP BY analysis, type&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    db</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">print( types )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>All the species have <code>pseudogenes</code> and <code>non-coding RNAs</code> as well as genes.</p><p><strong>Question:</strong> What are pseudogenes and non-coding RNAs anyway? Look them up on <a href="http://www.sequenceontology.org/so_wiki/index.php/Category:SO:SOFA" target="_blank" rel="noopener noreferrer">the sequence
ontology</a>. Here is <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3491395/" target="_blank" rel="noopener noreferrer">an
interesting reference on pseudogenes</a> and on
<a href="https://www.frontiersin.org/articles/10.3389/fgene.2015.00002/full" target="_blank" rel="noopener noreferrer">non-coding RNAs</a>.</p><p>(Some of the files also have <code>region</code> and <code>biological_region</code> records - my guess is these are computational
predictions of regions with biological function - e.g. transcript start sites from
<a href="https://www.sanger.ac.uk/tool/eponine/" target="_blank" rel="noopener noreferrer">Eponine</a>.)</p><p>What sets apart <em>genes</em> from <em>pseudogenes</em> and <em>non-coding RNAs</em> is that they code for proteins.
However, it&#x27;s not quite that simple. If you look more closely at the records with <code>type=gene</code>,
you&#x27;ll see they actually have a second <code>type</code> field - it is called <code>biotype</code> and is buried in the
attributes. This <code>biotype</code> tells us some extra information about the gene. To let us look at this
we need to update the code.</p><div class="admonition admonition-tip alert alert--success"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="16" viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Challenge</h5></div><div class="admonition-content"><p>Update your <code>gff.py</code> code to make it easy to get this information out - i.e. by adding the <code>biotype</code> as an extra column. (It
would also be useful to extract the <code>Name</code>, which contains gene names.)</p></div></div><p><strong>Hints:</strong> </p><ul><li><p>You already extracted <code>ID</code> and <code>Parent</code> into dedicated columns - it ought to be easy to get
other attributes as well?</p></li><li><p>If making changes feels difficult in your code - I suggest you might want to <a href="/whg-training-resources/programming/programming_with_gene_annotations/Refactoring_makes_code_better/">try a
refactor</a> first - that is, make changes to your code to make it simpler
without changing its behaviour. Since you have a good test you can make changes quite safely because the tests
will ensure that it still works correctly. Once the code feels nice and simple, it will be much easier to see
how to make changes.</p></li><li><p>Maybe you should just make <code>parse_gff3_to_dataframe()</code> take a list of attributes to extract - that is, let the
user decide what they want?</p></li><li><p><strong>Extra challenge</strong>: Wouldn&#x27;t it be good if we could remove the attributes we&#x27;ve put into columns (<code>ID</code>, <code>Parent</code>, <code>Name</code> and
<code>biotype</code> say) from the <code>attributes</code> column as well? (Note: this will change the behaviour slightly - remember to add a test
for this new functionality.)\</p></li><li><p><strong>Extra super-difficult challenge</strong>: Do this without using ridiculous amounts of memory.</p></li></ul><p><strong>More hints:</strong> My solutions are in
<a href="https://github.com/whg-training/whg-training-resources/blob/main/docs/programming/programming_with_gene_annotations/solutions/part2" target="_blank" rel="noopener noreferrer">this folder</a>
if you want to take a look. The first version, <a target="_blank" href="/whg-training-resources/assets/files/gff_highmem-2cc1e84ea4d7ec0c0528f7709f164f5f.py/">gff_highmem.py</a>, extracts any attribute and looks relatively simple.
But it turns out to <a href="/whg-training-resources/programming/programming_with_gene_annotations/Memory_issues_and_how_to_solve_them/">use up huge amounts of memory</a>. The second version uses a <em>chunking</em> strategy to remove
the attributes, and this helps control the memory.  On my laptop the various versions behave a bit differently:</p><table><thead><tr><th>Version</th><th>time taken to load GENCODE data</th><th>Max memory usage</th></tr></thead><tbody><tr><td>Original</td><td>About 18 seconds</td><td>About 2.4Gb</td></tr><tr><td>First version above</td><td>About 28 seconds</td><td>About 3.8Gb</td></tr><tr><td>Version using chunking</td><td>Also about 29 seconds</td><td>About 2.6Gb</td></tr></tbody></table><p>Now let&#x27;s talk more about <a href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_2/">Investigating protein-coding genes</a>.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/whg-training/whg-training-resources/edit/main/docs/programming/programming_with_gene_annotations/Counting_genes_1.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_dcUD" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_foO9"></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><a class="pagination-nav__link pagination-nav__link--prev" href="/whg-training-resources/programming/programming_with_gene_annotations/exploring_gencode/"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Investigating human genes</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/whg-training-resources/programming/programming_with_gene_annotations/Memory_issues_and_how_to_solve_them/"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Memory issues and how to solve them</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_cNA8 thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#some-first-statistics" class="table-of-contents__link toc-highlight">Some first statistics</a><ul><li><a href="#how-many-genes-are-there" class="table-of-contents__link toc-highlight">How many genes are there?</a></li><li><a href="#what-are-all-these-species" class="table-of-contents__link toc-highlight">What are all these species?</a></li><li><a href="#what-kinds-of-gene-are-there" class="table-of-contents__link toc-highlight">What kinds of gene are there?</a></li></ul></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">License</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/whg-training-resources/LICENSE/">License</a></li></ul></div><div class="col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2022 University of Oxford.  Built with Docusaurus.</div></div></div></footer></div>
<script src="/whg-training-resources/assets/js/runtime~main.50846bb3.js"></script>
<script src="/whg-training-resources/assets/js/main.1b628545.js"></script>
</body>
</html>