<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.20">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.24/dist/katex.min.css" integrity="sha384-odtC+0UGzzFL/6PNoE8rX/SPcQDXBJ+uRepguP4QkPCm2LBxH3FA3y+fKSiJ+AmM" crossorigin="anonymous"><title data-rh="true">How much of the genome is in genes? | WHG training resources</title><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes/"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="How much of the genome is in genes? | WHG training resources"><meta data-rh="true" name="description" content="Up to table of contents"><meta data-rh="true" property="og:description" content="Up to table of contents"><link data-rh="true" rel="icon" href="/whg-training-resources/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes/"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes/" hreflang="en"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes/" hreflang="x-default"><link rel="stylesheet" href="/whg-training-resources/assets/css/styles.b827ae01.css">
<link rel="preload" href="/whg-training-resources/assets/js/runtime~main.50846bb3.js" as="script">
<link rel="preload" href="/whg-training-resources/assets/js/main.1b628545.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region"><a href="#" class="skipToContent_ZgBM">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/whg-training-resources/"><div class="navbar__logo"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--light_TfLj"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--dark_oUvU"></div><b class="navbar__title text--truncate">Training Resources</b></a><a class="navbar__item navbar__link navbar__link--active" href="/whg-training-resources/overview/">Click here for the tutorials</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a><div class="toggle_S7eR colorModeToggle_vKtC"><button class="clean-btn toggleButton_rCf9 toggleButtonDisabled_Pu9x" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_v35p"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_nQuB"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_dLyj"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docsWrapper_mKqt"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_RiI4" type="button"></button><div class="docPage_ualW"><aside class="theme-doc-sidebar-container docSidebarContainer_UQUJ"><div class="sidebar_RiAD"><nav class="menu thin-scrollbar menu_izAj"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/overview/">Overview of the tutorials</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/prerequisites/">Getting your environment setup</a><button aria-label="Toggle the collapsible sidebar category &#x27;Getting your environment setup&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/bioinformatics/">Bioinformatics</a><button aria-label="Toggle the collapsible sidebar category &#x27;Bioinformatics&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" href="/whg-training-resources/programming/">Programming</a><button aria-label="Toggle the collapsible sidebar category &#x27;Programming&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/programming/introduction_to_R/">Introduction to R</a><button aria-label="Toggle the collapsible sidebar category &#x27;Introduction to R&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/">Programming with gene annotations</a><button aria-label="Toggle the collapsible sidebar category &#x27;Programming with gene annotations&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Introduction/">Introduction</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like/">What gene annotation data looks like</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code/">Processing GFF files</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/making_a_module/">Making a module</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Refactoring_makes_code_better/">An aside on refactoring</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Converting_gff_to_sqlite/">Writing a useful conversion program.</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/exploring_gencode/">Investigating human genes</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_1/">Counting genes I</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Memory_issues_and_how_to_solve_them/">Memory issues and how to solve them</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_2/">Counting genes II: types of protein-coding genes</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_3/">Extremely big, small and complex genes</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Getting_sequence_lengths/">Getting sequence lengths</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes/">How much of the genome is in genes?</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Scaling_up/">Scaling up for a real analysis</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Where_next/">What next?</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Closing_thoughts/">Closing thoughts</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/List_of_ensembl_files/">Appendix: a list of Ensembl GFF files.</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Visualisation/">Appendix: visualisation</a></li></ul></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/data_visualisation/">Data visualisation</a><button aria-label="Toggle the collapsible sidebar category &#x27;Data visualisation&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/next_generation_sequencing/">Next-generation sequencing</a><button aria-label="Toggle the collapsible sidebar category &#x27;Next-generation sequencing&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/statistical_modelling/">Statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;Statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/population_genetics/">Population genetics</a><button aria-label="Toggle the collapsible sidebar category &#x27;Population genetics&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/genome_wide_association_studies/">Genome-wide association analysis</a><button aria-label="Toggle the collapsible sidebar category &#x27;Genome-wide association analysis&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/LICENSE/">LICENSE</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/contributors/">contributors</a></li></ul></nav></div></aside><main class="docMainContainer_uL0j"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_DM6M"><div class="docItemContainer_vinB"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Xlws" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/whg-training-resources/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_kU5B"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/programming/"><span itemprop="name">Programming</span></a><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/programming/programming_with_gene_annotations/"><span itemprop="name">Programming with gene annotations</span></a><meta itemprop="position" content="2"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">How much of the genome is in genes?</span><meta itemprop="position" content="3"></li></ul></nav><div class="tocCollapsible_bZGK theme-doc-toc-mobile tocMobile_TmEX"><button type="button" class="clean-btn tocCollapsibleButton_l22C">On this page</button></div><div class="theme-doc-markdown markdown"><h1>How much of the genome is in genes?</h1><p><a href="/whg-training-resources/programming/programming_with_gene_annotations/">Up to table of contents</a></p><p>To figure out how much of the genome is covered by genes, or by exons, we face a problem.
In principle we could just add together the gene lengths.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">genes[&#x27;length&#x27;] = genes[&#x27;end&#x27;] - genes[&#x27;start&#x27;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">genes.groupby(&#x27;analysis&#x27;).agg(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    covered_length = pandas.NamedAgg( column = &#x27;length&#x27;, aggfunc = lambda w: w.sum() )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Then we could compare to the sequence lengths</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">sequence_lengths = pandas.read_sql( &quot;SELECT analysis, SUM(end-start) AS sequence_length FROM sequences GROUP BY analysis&quot;, db )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>So this would say for example that about 44% of the human genome is covered by genes.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="union-of-regions">Union of regions<a class="hash-link" href="#union-of-regions" title="Direct link to heading">​</a></h2><p>Unfortunately this isn&#x27;t so simple - because many genes overlap each other. This happens either
because there genuinely are different genes encoded by the same bit of DNA, or because of
additional annotated &#x27;genes&#x27; that arise due to the computational gene annotation process. (We saw
one earlier - <a href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_3/">a tiny annotated gene inside <em>SLC4A2</em></a>.)</p><p>To handle this we have to compute the regions covered by a bunch of overlapping genes, and for that
we need to be able to compute this overlap.</p><p><strong>Challenge.</strong>. Write a function <code>compute_union_of_regions()</code> that computes the total region covered
by a set of (possibly overlapping ranges). Both the input and output should be a list of pairs of
the form <code>[ start, end ]</code> (where <code>end</code> &gt;= <code>start</code> and the coordinates are all non-negative
integers. The output should contain a set of non-overlapping ranges that cover the same set of
positions as the input ranges. And for testability reasons, let&#x27;s also require the output to be
sorted (by the region start position).</p><p>Here is a test:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">class TestRanges(unittest.TestCase):</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    def test_union_of_ranges( self ):</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        # check some simple cases first</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        self.assertEqual( compute_union_of_regions( [[1,10]] ) == [[1,10]] )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        self.assertEqual( compute_union_of_regions( [] ) == [] )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        self.assertEqual( compute_union_of_regions( [[1,10], [11,11]] ) == [[1,11]] )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        test_data = [</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            [1, 10],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            [19,199],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            [5, 6],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            [9, 15],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            [20, 25]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        ]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        result = compute_union_of_ranges( test_data )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        self.assertEqual( len( result ), 2 )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        self.assertEqual( result[0] = [ 1, 15 ] )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        self.assertEqual( result[1] = [ 19, 15 ] )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><strong>Hints.</strong> </p><ul><li>First <a href="https://docs.python.org/3/howto/sorting.html" target="_blank" rel="noopener noreferrer">sort the list of regions by the start point</a>.
(But be aware that python functions can mutate their arguments). You may want to use the <code>sorted()</code> function rather than sorting
in-place.
</li><li>Now traverse the list of regions, keeping track of the current interval and extending it if
necessary when you encounter overlapping input regions.</li></ul><p><strong>Important note.</strong> The coordinates in the GFF file are defined to <a href="http://www.ensembl.org/info/website/upload/gff.html" target="_blank" rel="noopener noreferrer">follow the 1-based
convention</a>. This means that the genome
coordinates start at 1, and also that regions are defined to be closed - i.e. they contain both
their endpoints. A region like <!-- -->[1,10]<!-- --> therefore contains 10 base positions.</p><p>(If this sounds obvious, I&#x27;m raising it because in other contexts a 0-based, half-open convention is
used instead (in which the region [1,10) would contain only 9 positions, and would miss the 1st
genome location at zero). This is true for [the database that underlies the <a href="https://genome-blog.soe.ucsc.edu/blog/the-ucsc-genome-browser-coordinate-counting-systems/" target="_blank" rel="noopener noreferrer">UCSC Genome
Browser</a>
for example (but not the browser itself, which converts coordinates to 1-based), and is
common in programming generally.)</p><p>My solution is <a target="_blank" href="/whg-training-resources/assets/files/regions-243718f2d84581d7d2ad5611257c6bd6.py/">here</a>.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="so-how-much-of-the-genome-is-in-genes">So how much of the genome is in genes?<a class="hash-link" href="#so-how-much-of-the-genome-is-in-genes" title="Direct link to heading">​</a></h2><p>We&#x27;re ready to answer this! Let&#x27;s use the above to write function that counts the bases covered in each genome:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain"># Add this to your gff.py file</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">def compute_genome_bases_covered( regions, sequences ):</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   &quot;&quot;&quot;Given a set of regions (as a dataframe with analysis, seqid, and start, and end columns),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   and a set of sequences (as a dataframe with analysis, seqid and sequence_length columns, return</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   a dataframe showing the total number and proportion of sequence bases covered by the regions in</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   each analysis.&quot;&quot;&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   def sum_region_lengths( regions ):</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      # We convert from pandas to a list...</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      aslist = regions[[&#x27;start&#x27;, &#x27;end&#x27;]].values.tolist()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      # ...compute the union using your function...</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      union = compute_union_of_regions( aslist )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      # ...and sum to get the result.</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      result = sum( [ (region[1]-region[0]+1) for region in union ] )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      # Finally, we return as a pandas.Series() object</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      # because this lets us name the output variable.</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      return pandas.Series(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         sum( [ (region[1]-region[0]+1) for region in union ] ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         index = [&#x27;bases_covered&#x27;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   def compute_bases_covered_in_one_chromosome( regions ):</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      return (</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         regions</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         .groupby( [ &quot;analysis&quot;, &quot;seqid&quot; ])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         .apply( sum_region_lengths )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   def add_sequence_lengths( coverage, sequences ):</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      return pandas.merge(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         coverage,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         sequences,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         left_on = [ &quot;analysis&quot;, &quot;seqid&quot; ],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         right_on = [ &quot;analysis&quot;, &quot;seqid&quot; ]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   def sum_over_chromosomes( coverage ):</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      result = coverage.groupby( &quot;analysis&quot; ).agg(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         bases_covered = pandas.NamedAgg( column = &quot;bases_covered&quot;, aggfunc = sum ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         sequence_length = pandas.NamedAgg( column = &quot;sequence_length&quot;, aggfunc = sum )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      result[&#x27;proportion&#x27;] = result[&#x27;bases_covered&#x27;] / result[&#x27;sequence_length&#x27;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      return result</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   per_chromosome = add_sequence_lengths(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      compute_bases_covered_in_one_chromosome( regions )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   return sum_over_chromosomes( per_chromosome )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The code above is a bit complex, but in short:</p><ul><li>it uses the <code>compute_union_of_regions()</code> function to compute the sum of lengths of genes <em>in each
chromosome and species</em>.</li><li>it then adds a column with the sequence lengths <a href="/whg-training-resources/programming/programming_with_gene_annotations/Getting_sequence_lengths/">that we computed earlier</a> to the result.</li><li>Finally it summarises over the whole genome for each species and returns the result.</li></ul><p>Let&#x27;s try it out:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">import pandas, sqlite3, gff</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">db = sqlite3.connect( &quot;genes.sqlite&quot; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">genes = pandas.read_sql( &quot;&quot;&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   SELECT analysis, ID, Parent, seqid, start, end, strand, Name, biotype</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   FROM gff_data</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   WHERE type == &#x27;gene&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   AND biotype == &#x27;protein_coding&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">&quot;&quot;&quot;, db</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sequences = pandas.read_sql( &quot;SELECT analysis, seqid, ( end - start ) AS sequence_length FROM sequences&quot;, db )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gff.compute_genome_bases_covered( genes, sequences )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This gives:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">                                            bases_covered  sequence_length  proportion</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">analysis                                                                               </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Acanthochromis_polyacanthus.ASM210954v1.104      415285145        830196730    0.500225</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Camelus_dromedarius.CamDro2.104.chr.gff3         815783547       2052758671    0.397408</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Gallus_gallus.GRCg6a.104                         480522985       1050156563    0.457573</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Homo_sapiens.GRCh38.104                         1298183323       3085194781    0.420778</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Mus_musculus.GRCm39.104                         1046654226       2723431121    0.384315</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Pan_troglodytes.Pan_tro_3.0.104.chr             1101889770       2967125051    0.371366</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">PlasmoDB-54_Pfalciparum3D7                        18174434         23332823    0.778921</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>So only 42% of the human genome is covered by protein-coding genes. This drops a bit in
chimpanzees, while a colossal 78% of the <em>P.falciparum</em> genome is in genes.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="how-much-of-the-genome-is-in-exons">How much of the genome is in exons?<a class="hash-link" href="#how-much-of-the-genome-is-in-exons" title="Direct link to heading">​</a></h3><p>Here we&#x27;d better be a bit careful to only get exons in protein-coding genes. Let&#x27;s do it in python
as follows (as usual this could be done in sql as well, if we wanted):</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">transcripts = pandas.read_sql( &quot;&quot;&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   SELECT analysis, ID, Parent, seqid, start, end</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   FROM gff_data WHERE type == &#x27;mRNA&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">&quot;&quot;&quot;, db )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exons = pandas.read_sql( &quot;&quot;&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   SELECT analysis, ID, Parent, seqid, start, end</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   FROM gff_data WHERE type == &#x27;exon&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">&quot;&quot;&quot;, db )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">print( &quot;%d transcripts, %d exons.\n&quot; % ( transcripts.shape[0], exons.shape[0] ))</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Now find just the transcripts that actually are protein-coding genes:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">transcripts = transcripts[ transcripts[&#x27;Parent&#x27;].isin( genes[&#x27;ID&#x27;])]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exons = exons[ exons[&#x27;Parent&#x27;].isin( transcripts[&#x27;ID&#x27;] ) ]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">print( &quot;%d transcripts, %d exons.\n&quot; % ( transcripts.shape[0], exons.shape[0] ))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gff.compute_genome_bases_covered( exons, sequences )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This gives:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">                                             bases_covered  sequence_length  proportion</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">analysis                                                                               </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Acanthochromis_polyacanthus.ASM210954v1.104       64427980        830196730    0.077606</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Camelus_dromedarius.CamDro2.104.chr.gff3          44585999       2052758671    0.021720</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Gallus_gallus.GRCg6a.104                          41635090       1050156563    0.039647</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Homo_sapiens.GRCh38.104                           85003514       3085194781    0.027552</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Mus_musculus.GRCm39.104                           76703787       2723431121    0.028164</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Pan_troglodytes.Pan_tro_3.0.104.chr               61897466       2967125051    0.020861</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">PlasmoDB-54_Pfalciparum3D7                        16683959         23332823    0.715042</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>For most genomes, even though a large proportion of genome is in genes, hardly any is in exons. But
P.falciparum is different: we already saw it has <a href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_3/">lots of single-exon genes</a>
and sure enough most of the genome is in these exons.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="how-much-of-the-genome-is-in-coding-sequence">How much of the genome is in coding sequence?<a class="hash-link" href="#how-much-of-the-genome-is-in-coding-sequence" title="Direct link to heading">​</a></h3><p>This is easy now:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">cds = pandas.read_sql( &quot;&quot;&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   SELECT analysis, ID, Parent, seqid, start, end, strand, Name, biotype</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   FROM gff_data WHERE type == &#x27;CDS&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">&quot;&quot;&quot;, db)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cds = cds[ cds[&quot;Parent&quot;].isin( transcripts[&#x27;ID&#x27;])]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gff.compute_genome_bases_covered( cds, sequences )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This gives:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">                                             bases_covered  sequence_length  proportion</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">analysis                                                                               </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Acanthochromis_polyacanthus.ASM210954v1.104       37182784        830196730    0.044788</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Camelus_dromedarius.CamDro2.104.chr.gff3          31455083       2052758671    0.015323</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Gallus_gallus.GRCg6a.104                          28472131       1050156563    0.027112</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Homo_sapiens.GRCh38.104                           35837191       3085194781    0.011616</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Mus_musculus.GRCm39.104                           36746474       2723431121    0.013493</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Pan_troglodytes.Pan_tro_3.0.104.chr               35836571       2967125051    0.012078</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">PlasmoDB-54_Pfalciparum3D7                        12357002         23332823    0.529597</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>In humans, only about half of the sequence in protein-coding gene exons actually codes for proteins - genes have a great deal of <em>untranslated sequence</em>.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="conclusions">Conclusions<a class="hash-link" href="#conclusions" title="Direct link to heading">​</a></h3><p>Two really striking things jump out from this to me.  First, in many eukaryotic organisms, hardly any of the
genome is actually coding for protein. </p><p>What is the rest doing? Well, we saw that some of it is in <a href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_1/">pseudogenes or
encodes non-protein-coding RNAs</a>. Quite a lot actually is functional <a href="https://doi.org/10.1038/nature11247" target="_blank" rel="noopener noreferrer">as
described in the ENCODE paper</a> (at least in the limited sense that
proteins bind there, as part of a general process of regulating gene expression). Some of it is viral DNA
that has been retro-transposed into the genome - such as the <a href="https://en.wikipedia.org/wiki/Retrotransposon" target="_blank" rel="noopener noreferrer">LINEs and SINEs in the human
genome</a>.  Some is genuinely doing nothing. However, much
of the function of even well-annotated genomes (like humans) is not known.</p><p>On the other hand, some organisms like malaria seem to be making use of a great deal of their
genome for making proteins. Quite why this is is also not known.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="where-next">Where next?<a class="hash-link" href="#where-next" title="Direct link to heading">​</a></h2><p>Here are some <a href="/whg-training-resources/programming/programming_with_gene_annotations/Where_next/">suggested next steps</a>.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/whg-training/whg-training-resources/edit/main/docs/programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_dcUD" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_foO9"></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><a class="pagination-nav__link pagination-nav__link--prev" href="/whg-training-resources/programming/programming_with_gene_annotations/Getting_sequence_lengths/"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Getting sequence lengths</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/whg-training-resources/programming/programming_with_gene_annotations/Scaling_up/"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Scaling up for a real analysis</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_cNA8 thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#union-of-regions" class="table-of-contents__link toc-highlight">Union of regions</a></li><li><a href="#so-how-much-of-the-genome-is-in-genes" class="table-of-contents__link toc-highlight">So how much of the genome is in genes?</a><ul><li><a href="#how-much-of-the-genome-is-in-exons" class="table-of-contents__link toc-highlight">How much of the genome is in exons?</a></li><li><a href="#how-much-of-the-genome-is-in-coding-sequence" class="table-of-contents__link toc-highlight">How much of the genome is in coding sequence?</a></li><li><a href="#conclusions" class="table-of-contents__link toc-highlight">Conclusions</a></li></ul></li><li><a href="#where-next" class="table-of-contents__link toc-highlight">Where next?</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">License</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/whg-training-resources/LICENSE/">License</a></li></ul></div><div class="col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2022 University of Oxford.  Built with Docusaurus.</div></div></div></footer></div>
<script src="/whg-training-resources/assets/js/runtime~main.50846bb3.js"></script>
<script src="/whg-training-resources/assets/js/main.1b628545.js"></script>
</body>
</html>